{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0f5078b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "305f92f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from collections import defaultdict\n",
    "import multiprocessing\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import timeit\n",
    "\n",
    "import ribopy\n",
    "from ribopy import Ribo\n",
    "from ribopy.core.get_gadgets import get_region_boundaries, get_reference_names\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "sys.path.insert(0, '../snp')\n",
    "from ref_lib.Fasta import FastaEntry, FastaFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "102e1868",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_transcriptome_file = \"../../mouse_itp_reference/transcriptome/varnt_masked_and_filtered_mouse_transcriptome.fa.gz\"\n",
    "human_transcriptome_file = \"../../../itp/human_itp_reference/transcriptome/appris_human_v2_after_filtering.fa.gz\"\n",
    "\n",
    "mouse_ribo_file = \"../../mouse-itp_v5.ribo\"\n",
    "human_ribo_file = \"../../../itp/human-itp_v4.ribo\"\n",
    "\n",
    "LEN_MIN = 29\n",
    "LEN_MAX = 35"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e59d2af4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_sequences = dict()\n",
    "\n",
    "with FastaFile(mouse_transcriptome_file) as mouse_transcriptome:\n",
    "    for entry in mouse_transcriptome:\n",
    "        mouse_sequences[ entry.header ] = entry.sequence\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ebab5939",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_sequences = dict()\n",
    "\n",
    "with FastaFile(human_transcriptome_file) as human_transcriptome:\n",
    "    for entry in human_transcriptome:\n",
    "        human_sequences[ entry.header ] = entry.sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "348e098c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_ribo = Ribo(mouse_ribo_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "0f5deafd",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_ribo = Ribo(human_ribo_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3a903982",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_region_boundaries = get_region_boundaries(mouse_ribo._handle)\n",
    "mouse_gene_names        = get_reference_names(mouse_ribo._handle)\n",
    "\n",
    "mouse_ref_name_and_boundaries = zip(mouse_gene_names, mouse_region_boundaries)\n",
    "\n",
    "mouse_cds_boundaries = dict()\n",
    "\n",
    "for e in list(mouse_ref_name_and_boundaries):\n",
    "    mouse_cds_boundaries[e[0]] = e[1][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "cdfec5e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_region_boundaries = get_region_boundaries(human_ribo._handle)\n",
    "human_gene_names        = get_reference_names(human_ribo._handle)\n",
    "\n",
    "human_ref_name_and_boundaries = zip(human_gene_names, human_region_boundaries)\n",
    "\n",
    "human_cds_boundaries = dict()\n",
    "\n",
    "for e in list(human_ref_name_and_boundaries):\n",
    "    human_cds_boundaries[e[0]] = e[1][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c2b65775",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_ribo = Ribo(mouse_ribo_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d9ca32bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_coverage_at_length(this_ribo, experiment, rpf_length):\n",
    "    return this_ribo.get_coverage(experiment  = experiment,\n",
    "                                  range_lower = rpf_length, \n",
    "                                  range_upper = rpf_length)\n",
    "\n",
    "def make_coverage_dict_of_experiment(this_ribo, experiment, len_min, len_max):\n",
    "    result_dict = defaultdict(dict)\n",
    "    \n",
    "    for i in range(len_min, len_max + 1):\n",
    "        result_dict[i] = get_coverage_at_length(this_ribo, experiment, i)\n",
    "        \n",
    "    return result_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e540dda1",
   "metadata": {},
   "outputs": [],
   "source": [
    "MII_1_coverages = make_coverage_dict_of_experiment(mouse_ribo, \n",
    "                                                   experiment = \"20210301-ITP-MII-25-B\", \n",
    "                                                   len_min    = LEN_MIN, \n",
    "                                                   len_max    = LEN_MAX )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5bf4366d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[29, 30, 31, 32, 33, 34, 35]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(MII_1_coverages.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d8194c8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def init_frame_triplet():\n",
    "    return [0, 0, 0]\n",
    "\n",
    "def count_nucleotides(coverage_dict, \n",
    "                      cds_annotation_dict, \n",
    "                      sequence_dict, \n",
    "                      rpf_len,\n",
    "                      left_span = 2, right_span = 1 ):\n",
    "    \"\"\"\n",
    "    Given a footprint length, this functio creates a table where each row\n",
    "    comes from nucleotide sequences\n",
    "    and the three columns correspond to the frames 0,1,2\n",
    "    \n",
    "    coverage_dict:\n",
    "       gene_identifier -> coverage vector\n",
    "       \n",
    "    cds_annotation_dict:\n",
    "       gene_identifier -> [cds_start, cds_end]\n",
    "       \n",
    "    sequence_dict:\n",
    "       gene_identifier -> gene_sequence\n",
    "       \n",
    "    rpf_len:\n",
    "       read length, ribosome protected footpront length\n",
    "       \n",
    "    left_span:  the last nucleotides before the 3' end of the read \n",
    "    \n",
    "    right_span: the first nucleotides  the 3' end of the read\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    nuc_counts = defaultdict(init_frame_triplet)\n",
    "    \n",
    "    \n",
    "    for gene, coverage in coverage_dict.items():\n",
    "        for position in range( cds_annotation_dict[gene][0], cds_annotation_dict[gene][1] ):\n",
    "            if coverage[position] == 0:\n",
    "                continue\n",
    "                \n",
    "            this_sequence     = sequence_dict[gene][position + rpf_len - left_span: \\\n",
    "                                                    position + rpf_len + right_span]\n",
    "            \n",
    "            if \"N\" in this_sequence:\n",
    "                continue\n",
    "            \n",
    "            relative_position = position - cds_annotation_dict[gene][0]\n",
    "            \n",
    "            this_frame = relative_position % 3\n",
    "            \n",
    "            nuc_counts[this_sequence][this_frame] += coverage[position]\n",
    "        \n",
    "    return nuc_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0b246bf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_nucleotides_parallel( coverage_dict, \n",
    "                                cds_annotation_dict, \n",
    "                                sequence_dict, \n",
    "                                rpf_min, \n",
    "                                rpf_max,\n",
    "                                processes = 4,\n",
    "                                left_span = 2, right_span = 1 ):\n",
    "    \n",
    "    \"\"\"\n",
    "    Wrapper for the function `count_nucleotides`.\n",
    "    \n",
    "    \n",
    "    It takes a range of ribosome protected footprint lengths and runs count_nucletides\n",
    "    function on these individual lengths.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    rpf_range = list(range(rpf_min, rpf_max + 1))\n",
    "    \n",
    "    # Without this global, we have an error at the Pool.map step.\n",
    "    global f\n",
    "    \n",
    "    def f(x):\n",
    "       return count_nucleotides( coverage_dict[x], cds_annotation_dict, sequence_dict, rpf_len = x,\n",
    "                      left_span = left_span, right_span = right_span ) \n",
    "    \n",
    "    with Pool(processes) as p:\n",
    "        mapped_list = p.map(f, rpf_range)\n",
    "        \n",
    "    return(mapped_list)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "82745ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _adjust_individual_frames(frame_dict):\n",
    "    \"\"\"\n",
    "    Picks the maximum from each frame and moves the reads to the maximally picked frame.\n",
    "    \"\"\"\n",
    "    \n",
    "    result = dict()\n",
    "    \n",
    "    for this_seq, triplet in frame_dict.items():\n",
    "        \n",
    "        max_arg = np.argmax(triplet)\n",
    "        result[this_seq] = [0, 0, 0]\n",
    "        result[this_seq][max_arg] = np.sum(triplet)\n",
    "        \n",
    "    return result\n",
    "\n",
    "\n",
    "\n",
    "def adjust_frames(list_of_frame_dict, processes = 4):\n",
    "    \n",
    "    with Pool(processes) as p:\n",
    "        adjusted_frames = p.map(_adjust_individual_frames, list_of_frame_dict)\n",
    "        \n",
    "    return(adjusted_frames)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4ad9c52",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f0ecaa49",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _p_site_adjust(df):\n",
    "    \"\"\"\n",
    "    Bring maximal counts in the frames to the frame 0 for each length.\n",
    "    This is done via a cyclic shift so that the maximal is at frame 0.\n",
    "    \"\"\"\n",
    "    adjusted_tuples = []\n",
    "\n",
    "    for r,v in df.iterrows():\n",
    "        max_index      = np.argmax(v)\n",
    "        adjusted_tuple = [v[ (i + max_index) % 3] for i in range(3)  ]\n",
    "        adjusted_tuples.append(adjusted_tuple)\n",
    "\n",
    "    mydf = pd.DataFrame(adjusted_tuples)\n",
    "    \n",
    "    return mydf\n",
    "    \n",
    "\n",
    "\n",
    "def adjust_p_sites(nucleotide_counts_list):\n",
    "    summed_frames_per_length = []\n",
    "    \n",
    "    \n",
    "    # Sum the values for the frames 0,1,2\n",
    "    # collect them in a dataframe df_s\n",
    "    for nuc_counts in nucleotide_counts_list:\n",
    "        df = pd.DataFrame(nuc_counts)\n",
    "        summed_frames_per_length.append( df.sum(axis=1) )\n",
    "        \n",
    "    summed_frames_df       = pd.DataFrame(summed_frames_per_length)\n",
    "    p_site_adjusted_frames = _p_site_adjust(summed_frames_df)\n",
    "    \n",
    "    return p_site_adjusted_frames\n",
    "\n",
    "\n",
    "\n",
    "def calculate_periodicity_percentage(df):\n",
    "    frame_sums = df.sum()\n",
    "    \n",
    "    return 100 * ( frame_sums / np.sum(frame_sums) )\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793688bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "558622c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20210301-ITP-MII-25-B\n",
      "20210301-ITP-MII-50-A\n",
      "20210301-ITP-MII-50-B\n",
      "20210318-ITP-MII-50-B\n",
      "20210513-ITP-1cell-cross-50-A\n",
      "20210513-ITP-1cell-cross-50-B\n",
      "20210513-ITP-1cell-cross-50-C\n",
      "20210513-ITP-1cell-cross-50-D\n",
      "20210513-ITP-1cell-cross-50-E\n",
      "20210513-ITP-2cell-cross-50-B\n",
      "20210513-ITP-2cell-cross-50-C\n",
      "20210513-ITP-2cell-cross-50-F\n",
      "20210513-ITP-4cell-cross-50-B\n",
      "20210513-ITP-4cell-cross-50-C\n",
      "20210513-ITP-4cell-cross-50-D\n",
      "20210513-ITP-8cell-cross-50-A\n",
      "20210513-ITP-8cell-cross-50-B\n",
      "20210513-ITP-8cell-cross-50-C\n",
      "20210513-ITP-8cell-cross-50-D\n",
      "20210614-ITP-GV-50-A\n",
      "20210614-ITP-GV-50-B\n",
      "20210614-ITP-GV-50-C\n",
      "20210614-ITP-GV-50-E\n",
      "20210614-ITP-GV-50-F\n",
      "20210614-ITP-MII-50-D\n"
     ]
    }
   ],
   "source": [
    "experiments = mouse_ribo.experiments\n",
    "\n",
    "nucleotide_counts_2_1 = dict()\n",
    "nucleotide_counts_0_1 = dict()\n",
    "nucleotide_counts_0_2 = dict()\n",
    "\n",
    "for e in experiments:\n",
    "    print(e)\n",
    "    this_coverage = make_coverage_dict_of_experiment(\n",
    "                            mouse_ribo, \n",
    "                            experiment = e, \n",
    "                            len_min    = LEN_MIN, \n",
    "                            len_max    = LEN_MAX )\n",
    "    \n",
    "    nucleotide_counts_2_1[e] = count_nucleotides_parallel( \n",
    "                                    coverage_dict       = this_coverage, \n",
    "                                    cds_annotation_dict = mouse_cds_boundaries ,\n",
    "                                    sequence_dict       = mouse_sequences, \n",
    "                                    rpf_min = LEN_MIN, rpf_max = LEN_MAX,\n",
    "                                    left_span           = 2, \n",
    "                                    right_span          = 1 )\n",
    "    \n",
    "    nucleotide_counts_0_1[e] = count_nucleotides_parallel( \n",
    "                                    coverage_dict       = this_coverage, \n",
    "                                    cds_annotation_dict = mouse_cds_boundaries ,\n",
    "                                    sequence_dict       = mouse_sequences, \n",
    "                                    rpf_min = LEN_MIN, rpf_max = LEN_MAX,\n",
    "                                    left_span           = 0, \n",
    "                                    right_span          = 1 )   \n",
    "\n",
    "    nucleotide_counts_0_2[e] = count_nucleotides_parallel( \n",
    "                                    coverage_dict       = this_coverage, \n",
    "                                    cds_annotation_dict = mouse_cds_boundaries ,\n",
    "                                    sequence_dict       = mouse_sequences, \n",
    "                                    rpf_min = LEN_MIN, rpf_max = LEN_MAX,\n",
    "                                    left_span           = 0, \n",
    "                                    right_span          = 2 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "0dcf0344",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    66.805899\n",
       "1    16.141993\n",
       "2    17.052108\n",
       "dtype: float64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "one_nuc_adjusted_frames = adjust_frames(nucleotide_counts_0_1[mouse_ribo.experiments[22]]  )\n",
    "one_nuc_y               = adjust_p_sites(one_nuc_adjusted_frames)\n",
    "calculate_periodicity_percentage(one_nuc_y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "d1dbe5d0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mouse_adjusted_frame_percentages = dict()\n",
    "mouse_raw_frame_percentages      = dict()\n",
    "\n",
    "for e in mouse_ribo.experiments:\n",
    "    adjusted_frames            = adjust_frames(nucleotide_counts_0_1[e]  )\n",
    "    p_adj_frames               = adjust_p_sites(adjusted_frames)\n",
    "    p_raw_frames               = adjust_p_sites(nucleotide_counts_0_1[e])\n",
    "    \n",
    "    mouse_raw_frame_percentages[e]      = calculate_periodicity_percentage(p_raw_frames)\n",
    "    mouse_adjusted_frame_percentages[e] = calculate_periodicity_percentage(p_adj_frames)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "d4a2855c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_adjusted_frame_percentages_df = pd.DataFrame(mouse_adjusted_frame_percentages).transpose()\n",
    "mouse_raw_frame_percentages_df = pd.DataFrame(mouse_raw_frame_percentages).transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "00a29942",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20210301-ITP-MII-25-B</th>\n",
       "      <td>62.171025</td>\n",
       "      <td>18.163530</td>\n",
       "      <td>19.665445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210301-ITP-MII-50-A</th>\n",
       "      <td>65.422901</td>\n",
       "      <td>14.426416</td>\n",
       "      <td>20.150683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210301-ITP-MII-50-B</th>\n",
       "      <td>68.801977</td>\n",
       "      <td>14.477091</td>\n",
       "      <td>16.720932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210318-ITP-MII-50-B</th>\n",
       "      <td>61.859729</td>\n",
       "      <td>17.256521</td>\n",
       "      <td>20.883750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-1cell-cross-50-A</th>\n",
       "      <td>67.158603</td>\n",
       "      <td>8.635235</td>\n",
       "      <td>24.206162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-1cell-cross-50-B</th>\n",
       "      <td>62.973949</td>\n",
       "      <td>14.166165</td>\n",
       "      <td>22.859886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-1cell-cross-50-C</th>\n",
       "      <td>68.048846</td>\n",
       "      <td>9.605475</td>\n",
       "      <td>22.345679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-1cell-cross-50-D</th>\n",
       "      <td>62.464880</td>\n",
       "      <td>16.427329</td>\n",
       "      <td>21.107791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-1cell-cross-50-E</th>\n",
       "      <td>61.230001</td>\n",
       "      <td>15.073500</td>\n",
       "      <td>23.696500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-2cell-cross-50-B</th>\n",
       "      <td>72.482944</td>\n",
       "      <td>8.926726</td>\n",
       "      <td>18.590330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-2cell-cross-50-C</th>\n",
       "      <td>71.268274</td>\n",
       "      <td>9.553804</td>\n",
       "      <td>19.177922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-2cell-cross-50-F</th>\n",
       "      <td>66.714646</td>\n",
       "      <td>12.558129</td>\n",
       "      <td>20.727224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-4cell-cross-50-B</th>\n",
       "      <td>69.830495</td>\n",
       "      <td>12.479034</td>\n",
       "      <td>17.690471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-4cell-cross-50-C</th>\n",
       "      <td>64.317542</td>\n",
       "      <td>15.765015</td>\n",
       "      <td>19.917443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-4cell-cross-50-D</th>\n",
       "      <td>69.095260</td>\n",
       "      <td>9.535134</td>\n",
       "      <td>21.369606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-8cell-cross-50-A</th>\n",
       "      <td>68.547785</td>\n",
       "      <td>15.588185</td>\n",
       "      <td>15.864031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-8cell-cross-50-B</th>\n",
       "      <td>68.934287</td>\n",
       "      <td>11.884136</td>\n",
       "      <td>19.181578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-8cell-cross-50-C</th>\n",
       "      <td>72.125797</td>\n",
       "      <td>9.808818</td>\n",
       "      <td>18.065385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-8cell-cross-50-D</th>\n",
       "      <td>71.979479</td>\n",
       "      <td>7.567625</td>\n",
       "      <td>20.452897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-GV-50-A</th>\n",
       "      <td>66.755987</td>\n",
       "      <td>16.470330</td>\n",
       "      <td>16.773683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-GV-50-B</th>\n",
       "      <td>67.256169</td>\n",
       "      <td>19.642639</td>\n",
       "      <td>13.101191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-GV-50-C</th>\n",
       "      <td>68.497611</td>\n",
       "      <td>12.546296</td>\n",
       "      <td>18.956093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-GV-50-E</th>\n",
       "      <td>66.805899</td>\n",
       "      <td>16.141993</td>\n",
       "      <td>17.052108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-GV-50-F</th>\n",
       "      <td>69.798458</td>\n",
       "      <td>12.960757</td>\n",
       "      <td>17.240785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-MII-50-D</th>\n",
       "      <td>65.741882</td>\n",
       "      <td>16.660795</td>\n",
       "      <td>17.597323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       0          1          2\n",
       "20210301-ITP-MII-25-B          62.171025  18.163530  19.665445\n",
       "20210301-ITP-MII-50-A          65.422901  14.426416  20.150683\n",
       "20210301-ITP-MII-50-B          68.801977  14.477091  16.720932\n",
       "20210318-ITP-MII-50-B          61.859729  17.256521  20.883750\n",
       "20210513-ITP-1cell-cross-50-A  67.158603   8.635235  24.206162\n",
       "20210513-ITP-1cell-cross-50-B  62.973949  14.166165  22.859886\n",
       "20210513-ITP-1cell-cross-50-C  68.048846   9.605475  22.345679\n",
       "20210513-ITP-1cell-cross-50-D  62.464880  16.427329  21.107791\n",
       "20210513-ITP-1cell-cross-50-E  61.230001  15.073500  23.696500\n",
       "20210513-ITP-2cell-cross-50-B  72.482944   8.926726  18.590330\n",
       "20210513-ITP-2cell-cross-50-C  71.268274   9.553804  19.177922\n",
       "20210513-ITP-2cell-cross-50-F  66.714646  12.558129  20.727224\n",
       "20210513-ITP-4cell-cross-50-B  69.830495  12.479034  17.690471\n",
       "20210513-ITP-4cell-cross-50-C  64.317542  15.765015  19.917443\n",
       "20210513-ITP-4cell-cross-50-D  69.095260   9.535134  21.369606\n",
       "20210513-ITP-8cell-cross-50-A  68.547785  15.588185  15.864031\n",
       "20210513-ITP-8cell-cross-50-B  68.934287  11.884136  19.181578\n",
       "20210513-ITP-8cell-cross-50-C  72.125797   9.808818  18.065385\n",
       "20210513-ITP-8cell-cross-50-D  71.979479   7.567625  20.452897\n",
       "20210614-ITP-GV-50-A           66.755987  16.470330  16.773683\n",
       "20210614-ITP-GV-50-B           67.256169  19.642639  13.101191\n",
       "20210614-ITP-GV-50-C           68.497611  12.546296  18.956093\n",
       "20210614-ITP-GV-50-E           66.805899  16.141993  17.052108\n",
       "20210614-ITP-GV-50-F           69.798458  12.960757  17.240785\n",
       "20210614-ITP-MII-50-D          65.741882  16.660795  17.597323"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_adjusted_frame_percentages_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "d320cdf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_adjusted_frame_percentages_df.to_csv(\"mouse_adjusted_frame_percentages.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3476cbed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20210301-ITP-MII-25-B</th>\n",
       "      <td>37.501033</td>\n",
       "      <td>32.366963</td>\n",
       "      <td>30.132003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210301-ITP-MII-50-A</th>\n",
       "      <td>37.062525</td>\n",
       "      <td>32.116927</td>\n",
       "      <td>30.820548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210301-ITP-MII-50-B</th>\n",
       "      <td>36.976306</td>\n",
       "      <td>30.456614</td>\n",
       "      <td>32.567080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210318-ITP-MII-50-B</th>\n",
       "      <td>36.382408</td>\n",
       "      <td>31.976048</td>\n",
       "      <td>31.641544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-1cell-cross-50-A</th>\n",
       "      <td>37.201092</td>\n",
       "      <td>30.218190</td>\n",
       "      <td>32.580718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-1cell-cross-50-B</th>\n",
       "      <td>36.296678</td>\n",
       "      <td>30.961797</td>\n",
       "      <td>32.741524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-1cell-cross-50-C</th>\n",
       "      <td>37.155126</td>\n",
       "      <td>32.555019</td>\n",
       "      <td>30.289855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-1cell-cross-50-D</th>\n",
       "      <td>36.710074</td>\n",
       "      <td>32.112117</td>\n",
       "      <td>31.177809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-1cell-cross-50-E</th>\n",
       "      <td>36.871599</td>\n",
       "      <td>31.515471</td>\n",
       "      <td>31.612929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-2cell-cross-50-B</th>\n",
       "      <td>37.670929</td>\n",
       "      <td>30.402552</td>\n",
       "      <td>31.926519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-2cell-cross-50-C</th>\n",
       "      <td>37.002314</td>\n",
       "      <td>32.789383</td>\n",
       "      <td>30.208303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-2cell-cross-50-F</th>\n",
       "      <td>37.029984</td>\n",
       "      <td>30.320209</td>\n",
       "      <td>32.649807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-4cell-cross-50-B</th>\n",
       "      <td>37.174655</td>\n",
       "      <td>31.990844</td>\n",
       "      <td>30.834501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-4cell-cross-50-C</th>\n",
       "      <td>36.885268</td>\n",
       "      <td>32.292720</td>\n",
       "      <td>30.822011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-4cell-cross-50-D</th>\n",
       "      <td>36.772747</td>\n",
       "      <td>30.635120</td>\n",
       "      <td>32.592133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-8cell-cross-50-A</th>\n",
       "      <td>37.661629</td>\n",
       "      <td>29.574450</td>\n",
       "      <td>32.763922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-8cell-cross-50-B</th>\n",
       "      <td>37.046371</td>\n",
       "      <td>32.581748</td>\n",
       "      <td>30.371881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-8cell-cross-50-C</th>\n",
       "      <td>37.698565</td>\n",
       "      <td>31.749650</td>\n",
       "      <td>30.551785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210513-ITP-8cell-cross-50-D</th>\n",
       "      <td>37.147040</td>\n",
       "      <td>30.302151</td>\n",
       "      <td>32.550809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-GV-50-A</th>\n",
       "      <td>37.873502</td>\n",
       "      <td>31.179534</td>\n",
       "      <td>30.946964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-GV-50-B</th>\n",
       "      <td>37.819653</td>\n",
       "      <td>32.536488</td>\n",
       "      <td>29.643859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-GV-50-C</th>\n",
       "      <td>38.298984</td>\n",
       "      <td>31.305257</td>\n",
       "      <td>30.395759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-GV-50-E</th>\n",
       "      <td>38.644994</td>\n",
       "      <td>31.169308</td>\n",
       "      <td>30.185698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-GV-50-F</th>\n",
       "      <td>38.595474</td>\n",
       "      <td>30.735958</td>\n",
       "      <td>30.668568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210614-ITP-MII-50-D</th>\n",
       "      <td>36.788797</td>\n",
       "      <td>30.065762</td>\n",
       "      <td>33.145441</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       0          1          2\n",
       "20210301-ITP-MII-25-B          37.501033  32.366963  30.132003\n",
       "20210301-ITP-MII-50-A          37.062525  32.116927  30.820548\n",
       "20210301-ITP-MII-50-B          36.976306  30.456614  32.567080\n",
       "20210318-ITP-MII-50-B          36.382408  31.976048  31.641544\n",
       "20210513-ITP-1cell-cross-50-A  37.201092  30.218190  32.580718\n",
       "20210513-ITP-1cell-cross-50-B  36.296678  30.961797  32.741524\n",
       "20210513-ITP-1cell-cross-50-C  37.155126  32.555019  30.289855\n",
       "20210513-ITP-1cell-cross-50-D  36.710074  32.112117  31.177809\n",
       "20210513-ITP-1cell-cross-50-E  36.871599  31.515471  31.612929\n",
       "20210513-ITP-2cell-cross-50-B  37.670929  30.402552  31.926519\n",
       "20210513-ITP-2cell-cross-50-C  37.002314  32.789383  30.208303\n",
       "20210513-ITP-2cell-cross-50-F  37.029984  30.320209  32.649807\n",
       "20210513-ITP-4cell-cross-50-B  37.174655  31.990844  30.834501\n",
       "20210513-ITP-4cell-cross-50-C  36.885268  32.292720  30.822011\n",
       "20210513-ITP-4cell-cross-50-D  36.772747  30.635120  32.592133\n",
       "20210513-ITP-8cell-cross-50-A  37.661629  29.574450  32.763922\n",
       "20210513-ITP-8cell-cross-50-B  37.046371  32.581748  30.371881\n",
       "20210513-ITP-8cell-cross-50-C  37.698565  31.749650  30.551785\n",
       "20210513-ITP-8cell-cross-50-D  37.147040  30.302151  32.550809\n",
       "20210614-ITP-GV-50-A           37.873502  31.179534  30.946964\n",
       "20210614-ITP-GV-50-B           37.819653  32.536488  29.643859\n",
       "20210614-ITP-GV-50-C           38.298984  31.305257  30.395759\n",
       "20210614-ITP-GV-50-E           38.644994  31.169308  30.185698\n",
       "20210614-ITP-GV-50-F           38.595474  30.735958  30.668568\n",
       "20210614-ITP-MII-50-D          36.788797  30.065762  33.145441"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_raw_frame_percentages_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "8a1a25cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_raw_frame_percentages_df.to_csv(\"mouse_raw_frame_percentages.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6d3cb4d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e92630ee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "30b90dc8",
   "metadata": {},
   "source": [
    "--------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f580ea86",
   "metadata": {},
   "source": [
    "## HUMAN DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "45cf7f6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('20191203-Kit-10M-Monosome-1',\n",
       " '20191203-Kit-10M-Monosome-2',\n",
       " '20191203-Kit-10M-Monosome-3',\n",
       " '20201104-ITP-100-5mM-50-1',\n",
       " '20201209-ITP-100-5mM-6',\n",
       " '20210131-ITP-100-5mM-50_diluted-1')"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_ribo.experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "e75a5d59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20191203-Kit-10M-Monosome-1\n",
      "20191203-Kit-10M-Monosome-2\n",
      "20191203-Kit-10M-Monosome-3\n",
      "20201104-ITP-100-5mM-50-1\n",
      "20201209-ITP-100-5mM-6\n",
      "20210131-ITP-100-5mM-50_diluted-1\n"
     ]
    }
   ],
   "source": [
    "human_experiments = human_ribo.experiments\n",
    "\n",
    "\n",
    "human_nucleotide_counts_0_1 = dict()\n",
    "\n",
    "\n",
    "for e in human_experiments:\n",
    "    print(e)\n",
    "    this_coverage = make_coverage_dict_of_experiment(\n",
    "                            human_ribo, \n",
    "                            experiment = e, \n",
    "                            len_min    = LEN_MIN, \n",
    "                            len_max    = LEN_MAX )\n",
    "    \n",
    "    \n",
    "    human_nucleotide_counts_0_1[e] = count_nucleotides_parallel( \n",
    "                                    coverage_dict       = this_coverage, \n",
    "                                    cds_annotation_dict = human_cds_boundaries ,\n",
    "                                    sequence_dict       = human_sequences, \n",
    "                                    rpf_min = LEN_MIN, rpf_max = LEN_MAX,\n",
    "                                    left_span           = 0, \n",
    "                                    right_span          = 1 )   \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "f5340671",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_adjusted_frame_percentages = dict()\n",
    "human_raw_frame_percentages      = dict()\n",
    "\n",
    "for e in human_ribo.experiments:\n",
    "    adjusted_frames            = adjust_frames(human_nucleotide_counts_0_1[e]  )\n",
    "    p_adj_frames               = adjust_p_sites(adjusted_frames)\n",
    "    p_raw_frames               = adjust_p_sites(human_nucleotide_counts_0_1[e])\n",
    "    \n",
    "    human_raw_frame_percentages[e]      = calculate_periodicity_percentage(p_raw_frames)\n",
    "    human_adjusted_frame_percentages[e] = calculate_periodicity_percentage(p_adj_frames) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "acb28c26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20191203-Kit-10M-Monosome-1</th>\n",
       "      <td>39.006724</td>\n",
       "      <td>31.846913</td>\n",
       "      <td>29.146363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20191203-Kit-10M-Monosome-2</th>\n",
       "      <td>38.693376</td>\n",
       "      <td>32.024315</td>\n",
       "      <td>29.282310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20191203-Kit-10M-Monosome-3</th>\n",
       "      <td>38.623567</td>\n",
       "      <td>32.480144</td>\n",
       "      <td>28.896289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20201104-ITP-100-5mM-50-1</th>\n",
       "      <td>38.641237</td>\n",
       "      <td>32.225242</td>\n",
       "      <td>29.133521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20201209-ITP-100-5mM-6</th>\n",
       "      <td>37.139046</td>\n",
       "      <td>31.723705</td>\n",
       "      <td>31.137249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210131-ITP-100-5mM-50_diluted-1</th>\n",
       "      <td>37.707792</td>\n",
       "      <td>31.300839</td>\n",
       "      <td>30.991369</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           0          1          2\n",
       "20191203-Kit-10M-Monosome-1        39.006724  31.846913  29.146363\n",
       "20191203-Kit-10M-Monosome-2        38.693376  32.024315  29.282310\n",
       "20191203-Kit-10M-Monosome-3        38.623567  32.480144  28.896289\n",
       "20201104-ITP-100-5mM-50-1          38.641237  32.225242  29.133521\n",
       "20201209-ITP-100-5mM-6             37.139046  31.723705  31.137249\n",
       "20210131-ITP-100-5mM-50_diluted-1  37.707792  31.300839  30.991369"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_raw_frame_percentages_df = pd.DataFrame(human_raw_frame_percentages).transpose()\n",
    "human_raw_frame_percentages_df.to_csv(\"human_raw_frame_percentages.csv\")\n",
    "human_raw_frame_percentages_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "382ae527",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20191203-Kit-10M-Monosome-1</th>\n",
       "      <td>72.230854</td>\n",
       "      <td>11.828664</td>\n",
       "      <td>15.940481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20191203-Kit-10M-Monosome-2</th>\n",
       "      <td>72.891240</td>\n",
       "      <td>11.484843</td>\n",
       "      <td>15.623917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20191203-Kit-10M-Monosome-3</th>\n",
       "      <td>72.742513</td>\n",
       "      <td>11.697062</td>\n",
       "      <td>15.560425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20201104-ITP-100-5mM-50-1</th>\n",
       "      <td>70.082638</td>\n",
       "      <td>9.634706</td>\n",
       "      <td>20.282655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20201209-ITP-100-5mM-6</th>\n",
       "      <td>61.387099</td>\n",
       "      <td>19.283025</td>\n",
       "      <td>19.329876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210131-ITP-100-5mM-50_diluted-1</th>\n",
       "      <td>67.161393</td>\n",
       "      <td>11.709822</td>\n",
       "      <td>21.128785</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           0          1          2\n",
       "20191203-Kit-10M-Monosome-1        72.230854  11.828664  15.940481\n",
       "20191203-Kit-10M-Monosome-2        72.891240  11.484843  15.623917\n",
       "20191203-Kit-10M-Monosome-3        72.742513  11.697062  15.560425\n",
       "20201104-ITP-100-5mM-50-1          70.082638   9.634706  20.282655\n",
       "20201209-ITP-100-5mM-6             61.387099  19.283025  19.329876\n",
       "20210131-ITP-100-5mM-50_diluted-1  67.161393  11.709822  21.128785"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_adjusted_frame_percentages_df = pd.DataFrame(human_adjusted_frame_percentages).transpose()\n",
    "human_adjusted_frame_percentages_df.to_csv(\"human_adjusted_frame_percentages.csv\")\n",
    "human_adjusted_frame_percentages_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5a8787b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dc1892c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "788335fd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
