{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "57d6fd0d",
   "metadata": {},
   "source": [
    "# A closer look at the SNPs of Highlighted Genes\n",
    "\n",
    "We hhiglighted 24 genes that show striking difference in maternal and paternal allele translation and activation in between stages. Here we study the positions of the SNPs, the regions they come from (5' UTR, CDS, or 3' UTR) and the nucleotide sequences around those SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "07dcdfc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import copy\n",
    "\n",
    "import gzip\n",
    "\n",
    "import sys\n",
    "sys.path.insert(0, '../../snp')\n",
    "from ref_lib.Fasta import FastaEntry, FastaFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "089efb2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_to_amio_acid = {\n",
    "    \"TTT\": \"Phe\",\n",
    "    \"TTC\": \"Phe\",\n",
    "    \n",
    "    \"TTA\": \"Leu\",\n",
    "    \"TTG\": \"Leu\",\n",
    "    \"CTT\": \"Leu\",\n",
    "    \"CTC\": \"Leu\",\n",
    "    \"CTA\": \"Leu\",\n",
    "    \"CTG\": \"Leu\",\n",
    "    \n",
    "    \"ATT\": \"Ile\",\n",
    "    \"ATC\": \"Ile\",\n",
    "    \"ATA\": \"Ile\",\n",
    "    \n",
    "    \"ATG\": \"Met\",\n",
    "    \n",
    "    \"GTT\": \"Val\",\n",
    "    \"GTC\": \"Val\",\n",
    "    \"GTA\": \"Val\",\n",
    "    \"GTG\": \"Val\",\n",
    "    \n",
    "    \"TCT\": \"Ser\",\n",
    "    \"TCC\": \"Ser\",\n",
    "    \"TCA\": \"Ser\",\n",
    "    \"TCG\": \"Ser\",\n",
    "    \n",
    "    \"CCT\": \"Pro\",\n",
    "    \"CCC\": \"Pro\",\n",
    "    \"CCA\": \"Pro\",\n",
    "    \"CCG\": \"Pro\",\n",
    "    \n",
    "    \"ACT\": \"Thr\",\n",
    "    \"ACC\": \"Thr\",\n",
    "    \"ACA\": \"Thr\",\n",
    "    \"ACG\": \"Thr\",\n",
    "    \n",
    "    \"GCT\": \"Ala\",\n",
    "    \"GCC\": \"Ala\",\n",
    "    \"GCA\": \"Ala\",\n",
    "    \"GCG\": \"Ala\",\n",
    "    \n",
    "    \"TAT\": \"Tyr\",\n",
    "    \"TAC\": \"Tyr\",\n",
    "    \n",
    "    \"TAA\": \"STOP\",\n",
    "    \"TAG\": \"STOP\",\n",
    "    \n",
    "    \"CAT\": \"His\",\n",
    "    \"CAC\": \"His\",\n",
    "    \n",
    "    \"CAA\": \"Gln\",\n",
    "    \"CAG\": \"Gln\",\n",
    "    \n",
    "    \"AAT\": \"Asn\",\n",
    "    \"AAC\": \"Asn\",\n",
    "    \n",
    "    \"AAA\": \"Lys\",\n",
    "    \"AAG\": \"Lys\",\n",
    "    \n",
    "    \"GAT\": \"Asp\",\n",
    "    \"GAC\": \"Asp\",\n",
    "    \n",
    "    \"GAA\": \"Glu\",\n",
    "    \"GAG\": \"Glu\",\n",
    "    \n",
    "    \"TGT\": \"Cys\",\n",
    "    \"TGC\": \"Cys\",\n",
    "    \n",
    "    \"TGA\": \"STOP\",\n",
    "    \n",
    "    \"TGG\": \"Trp\",\n",
    "    \n",
    "    \"CGT\": \"Arg\",\n",
    "    \"CGC\": \"Arg\",\n",
    "    \"CGA\": \"Arg\",\n",
    "    \"CGG\": \"Arg\",\n",
    "    \n",
    "    \"AGT\": \"Ser\",\n",
    "    \"AGC\": \"Ser\",\n",
    "    \n",
    "    \"AGA\": \"Arg\",\n",
    "    \"AGG\": \"Arg\",\n",
    "    \n",
    "    \"GGT\": \"Gly\",\n",
    "    \"GGC\": \"Gly\",\n",
    "    \"GGA\": \"Gly\",\n",
    "    \"GGG\": \"Gly\"    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "873d3150",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Ser'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_to_amio_acid[\"AGC\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "be1e87bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This comes from our prior porportionality analysis\n",
    "\n",
    "list_of_genes = (\n",
    "  'Nop14',\n",
    "  'Tmppe', \n",
    "  'Slc13a2',\n",
    "  'Ppp2ca',\n",
    "  'Srpk1',\n",
    "  'Cbx3',\n",
    "  'Ncoa3',\n",
    "  'Cdk1',\n",
    "  'Baz1a',\n",
    "  'Dyrk3',\n",
    "  'Lclat1',\n",
    "  'Lyar',\n",
    "  'Umps',\n",
    "  'Tsen2',\n",
    "  'Ccnh',\n",
    "  \n",
    "  'Folr1',\n",
    "  'Pa2g4',\n",
    "  'Zfp296',\n",
    "  'Mrps9',\n",
    "  'Eif3d',\n",
    "  'Nin',\n",
    "  'Ddx21',\n",
    "  'Bcat1',\n",
    "  'Mysm1'\n",
    ")\n",
    "\n",
    "cluster_1_genes = (  'Nop14',\n",
    "                     'Slc13a2'\n",
    "                  )\n",
    "\n",
    "cluster_2_genes = ('Cbx3',\n",
    "                   'Srpk1',\n",
    "                   'Umps',\n",
    "                   'Mysm1',\n",
    "                   'Ppp2ca',\n",
    "                   'Bcat1'\n",
    "                   )\n",
    "\n",
    "cluster_3_genes = ('Folr1',\n",
    "                   'Zfp296',\n",
    "                   'Nin',\n",
    "                   'Ddx21',\n",
    "                   'Eif3d',\n",
    "                   'Pa2g4',\n",
    "                   'Mrps9',\n",
    "                   'Tsen2'\n",
    "                  )\n",
    "\n",
    "cluster_4_genes = ('Cdk1',\n",
    "                   'Baz1a',\n",
    "                   'Lclat1',\n",
    "                   'Ncoa3',\n",
    "                   'Lyar',\n",
    "                   'Dyrk3',\n",
    "                   'Ccnh',\n",
    "                   'Tmppe' \n",
    "                  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "26436202",
   "metadata": {},
   "outputs": [],
   "source": [
    "ribo_detailed_snp_file      = \"./snp_dataframes/riboseq_detailed_snps.csv.gz\"\n",
    "\n",
    "transcriptome_sequence_file = \"../../../mouse_itp_reference/transcriptome/varnt_masked_and_filtered_mouse_transcriptome.fa.gz\"\n",
    "\n",
    "annotation_file             = \"../../../mouse_itp_reference/transcriptome/appris_mouse_v2_filtered_regions.bed\"\n",
    "\n",
    "all_transcriptomic_snps_file = \"transcriptomic_variants.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0e1de5cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_sequences = dict()\n",
    "\n",
    "with FastaFile(transcriptome_sequence_file) as mouse_transcriptome:\n",
    "    for entry in mouse_transcriptome:\n",
    "        this_gene = entry.header.split(\"|\")[5]\n",
    "        mouse_sequences[ this_gene ] = entry.sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e005360b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ef3a68a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## initialize the dataframe\n",
    "## Simply pick all the entries from the transcrioptomic variants file\n",
    "\n",
    "## VERY IMPORTNAT:\n",
    "## VCF is 1-based\n",
    "# So we subtract 1 and convert it to 0-based\n",
    "# wgile reading the VCF file\n",
    "\n",
    "df_fields = {   \"transcript\" : list(), \n",
    "                \"gene\"       : list(),\n",
    "                \"position\"   : list(),  \n",
    "                \"maternal\"   : list(),\n",
    "                \"paternal\"   : list(),\n",
    "                \"chromosome\" : list(),\n",
    "                \"chr_pos\"    : list(),\n",
    "                \"strand\"     : list() }\n",
    "\n",
    "\n",
    "with gzip.open(all_transcriptomic_snps_file, \"rt\") as input_stream:\n",
    "    for line in input_stream:\n",
    "        if line.startswith(\"#\"):\n",
    "            continue\n",
    "\n",
    "        line_contents   = line.strip().split(\"\\t\")\n",
    "        label_contents  = line_contents[0].split(\"|\")\n",
    "\n",
    "        this_gene       = label_contents[5]\n",
    "        this_transcript = label_contents[0].split(\".\")[0]\n",
    "\n",
    "        if this_gene in list_of_genes:\n",
    "            df_fields[\"transcript\"].append( this_transcript )\n",
    "            df_fields[\"gene\"].append(this_gene)\n",
    "            #Convert 1-based coordinate to 0-vased by -1\n",
    "            df_fields[\"position\"].append( int(line_contents[1]) - 1 )\n",
    "            df_fields[\"maternal\"].append(line_contents[3])\n",
    "            df_fields[\"paternal\"].append(line_contents[4])\n",
    "            df_fields[\"chromosome\"].append(line_contents[10])\n",
    "            df_fields[\"chr_pos\"].append(line_contents[11])\n",
    "            df_fields[\"strand\"].append(line_contents[12])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "80bbf56a",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_df = pd.DataFrame(df_fields)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0e838a7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_boundaries = dict()\n",
    "\n",
    "annotation_file = \"../../../mouse_itp_reference/transcriptome/appris_mouse_v2_filtered_regions.bed\"\n",
    "\n",
    "with open(annotation_file, \"r\") as annotation_stream:\n",
    "    for entry in annotation_stream:\n",
    "        contents = entry.split()\n",
    "        \n",
    "        if contents[3] == \"CDS\":\n",
    "            this_gene                 = contents[0].split(\"|\")[5]\n",
    "            cds_boundaries[this_gene] = [int(contents[1]), int(contents[2])]\n",
    "            \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "37f8af41",
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_region(gene, position, cds_boundaries):\n",
    "    this_region = \"CDS\"\n",
    "    \n",
    "    if position < cds_boundaries[gene][0]:\n",
    "        this_region = \"UTR5\"\n",
    "        \n",
    "    if position > cds_boundaries[gene][1]:\n",
    "        this_region = \"UTR3\"\n",
    "    \n",
    "    return this_region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c39e9ec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Now add the region column\n",
    "\n",
    "snp_regions = list()\n",
    "\n",
    "for ind, contents in snp_df.iterrows():\n",
    "    this_region = determine_region( contents[\"gene\"], int(contents[\"position\"] ), cds_boundaries )\n",
    "    snp_regions.append(this_region)\n",
    "\n",
    "snp_df[\"region\"] = snp_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "72973749",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Next add the sequence around the snp\n",
    "\n",
    "\n",
    "\n",
    "left_span  = 4\n",
    "right_span = 4\n",
    "\n",
    "sequences_around_snps = list()\n",
    "\n",
    "for index, entry in  snp_df.iterrows():\n",
    "    # Note that our snp file is 1-based!\n",
    "    # So we need the -1 adjustment\n",
    "    sequence_start = entry[\"position\"] - left_span\n",
    "    sequence_stop  = entry[\"position\"] + right_span +1 \n",
    "    sequence       = mouse_sequences[ entry[\"gene\"] ][sequence_start : sequence_stop]\n",
    "    sequences_around_snps.append(sequence)\n",
    "\n",
    "snp_df[\"sequence\"] = sequences_around_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5651630f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "      <th>position</th>\n",
       "      <th>maternal</th>\n",
       "      <th>paternal</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>chr_pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>region</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000045866</td>\n",
       "      <td>Ddx21</td>\n",
       "      <td>4410</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr10</td>\n",
       "      <td>62580562</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>TTAANTGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000045866</td>\n",
       "      <td>Ddx21</td>\n",
       "      <td>4399</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr10</td>\n",
       "      <td>62580573</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>ATTGNCACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000045866</td>\n",
       "      <td>Ddx21</td>\n",
       "      <td>4063</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10</td>\n",
       "      <td>62580909</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>CTCCNGGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000045866</td>\n",
       "      <td>Ddx21</td>\n",
       "      <td>3913</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr10</td>\n",
       "      <td>62581059</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>TTGGNAGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000045866</td>\n",
       "      <td>Ddx21</td>\n",
       "      <td>3895</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr10</td>\n",
       "      <td>62581077</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>AGGGNTCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>ENSMUST00000111820</td>\n",
       "      <td>Tmppe</td>\n",
       "      <td>2175</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr9</td>\n",
       "      <td>114406531</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>GATTNCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>ENSMUST00000111820</td>\n",
       "      <td>Tmppe</td>\n",
       "      <td>2591</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr9</td>\n",
       "      <td>114406947</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>TCACNCATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>ENSMUST00000111820</td>\n",
       "      <td>Tmppe</td>\n",
       "      <td>2708</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr9</td>\n",
       "      <td>114407064</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>TGGGNCTGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>ENSMUST00000111820</td>\n",
       "      <td>Tmppe</td>\n",
       "      <td>2721</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr9</td>\n",
       "      <td>114407077</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>TTAGNGAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>ENSMUST00000111820</td>\n",
       "      <td>Tmppe</td>\n",
       "      <td>2750</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr9</td>\n",
       "      <td>114407106</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>CCATNGGAA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>528 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             transcript   gene  position maternal paternal chromosome  \\\n",
       "0    ENSMUST00000045866  Ddx21      4410        A        G      chr10   \n",
       "1    ENSMUST00000045866  Ddx21      4399        A        G      chr10   \n",
       "2    ENSMUST00000045866  Ddx21      4063        C        T      chr10   \n",
       "3    ENSMUST00000045866  Ddx21      3913        G        A      chr10   \n",
       "4    ENSMUST00000045866  Ddx21      3895        C        A      chr10   \n",
       "..                  ...    ...       ...      ...      ...        ...   \n",
       "523  ENSMUST00000111820  Tmppe      2175        A        G       chr9   \n",
       "524  ENSMUST00000111820  Tmppe      2591        T        G       chr9   \n",
       "525  ENSMUST00000111820  Tmppe      2708        C        T       chr9   \n",
       "526  ENSMUST00000111820  Tmppe      2721        A        T       chr9   \n",
       "527  ENSMUST00000111820  Tmppe      2750        A        G       chr9   \n",
       "\n",
       "       chr_pos strand region   sequence  \n",
       "0     62580562      -   UTR3  TTAANTGAA  \n",
       "1     62580573      -   UTR3  ATTGNCACA  \n",
       "2     62580909      -   UTR3  CTCCNGGTA  \n",
       "3     62581059      -   UTR3  TTGGNAGGA  \n",
       "4     62581077      -   UTR3  AGGGNTCTT  \n",
       "..         ...    ...    ...        ...  \n",
       "523  114406531      +   UTR3  GATTNCTGT  \n",
       "524  114406947      +   UTR3  TCACNCATA  \n",
       "525  114407064      +   UTR3  TGGGNCTGA  \n",
       "526  114407077      +   UTR3  TTAGNGAGA  \n",
       "527  114407106      +   UTR3  CCATNGGAA  \n",
       "\n",
       "[528 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f0d32334",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We determine the amino acid change for the SNPS in the CDS\n",
    "\n",
    "maternal_sequences = list()\n",
    "paternal_sequences = list()\n",
    "\n",
    "maternal_aa = list()\n",
    "paternal_aa = list()\n",
    "\n",
    "for index, entry in  snp_df.iterrows():\n",
    "    if entry[\"region\"] != \"CDS\":\n",
    "        maternal_sequences.append(\"NA\")\n",
    "        paternal_sequences.append(\"NA\")\n",
    "        maternal_aa.append(\"NA\")\n",
    "        paternal_aa.append(\"NA\")\n",
    "        continue\n",
    "\n",
    "    this_position  = entry[\"position\"]\n",
    "    cds_start      = cds_boundaries[entry[\"gene\"]][0]\n",
    "    this_frame = (this_position - cds_start) % 3\n",
    "\n",
    "    codon_start = this_position - this_frame\n",
    "    raw_triplet = list(mouse_sequences[entry[\"gene\"]][ codon_start: codon_start + 3 ] )\n",
    "\n",
    "    maternal_triplet = copy.deepcopy(raw_triplet)\n",
    "    paternal_triplet = copy.deepcopy(raw_triplet)\n",
    "\n",
    "    maternal_triplet[this_frame] = entry[\"maternal\"]\n",
    "    paternal_triplet[this_frame] = entry[\"paternal\"]\n",
    "\n",
    "    maternal_triplet = \"\".join(maternal_triplet)\n",
    "    paternal_triplet = \"\".join(paternal_triplet)\n",
    "    maternal_sequences.append( maternal_triplet )\n",
    "    paternal_sequences.append( paternal_triplet )\n",
    "    maternal_aa.append(codon_to_amio_acid[maternal_triplet])\n",
    "    paternal_aa.append(codon_to_amio_acid[paternal_triplet])\n",
    "\n",
    "snp_df[\"maternal_codon\"] = maternal_sequences\n",
    "snp_df[\"paternal_codon\"] = paternal_sequences\n",
    "\n",
    "snp_df[\"maternal_AA\"] = maternal_aa\n",
    "snp_df[\"paternal_AA\"] = paternal_aa\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f9edc353",
   "metadata": {},
   "outputs": [],
   "source": [
    "utr3_snps = snp_df[ snp_df[\"region\"] == \"UTR3\"  ]\n",
    "utr5_snps = snp_df[ snp_df[\"region\"] == \"UTR5\"  ]\n",
    "\n",
    "cds_snps        = snp_df[ snp_df[\"region\"] == \"CDS\"  ]\n",
    "synonymous_snps =  np.sum(cds_snps[\"maternal_AA\"] == cds_snps[\"paternal_AA\"] )\n",
    "\n",
    "total_cds_snps = len(cds_snps)\n",
    "\n",
    "non_syn_snps   = total_cds_snps - synonymous_snps\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df7cc82d",
   "metadata": {},
   "source": [
    "## Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c342715b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total number of snps is 528\n",
      "\n",
      "The distribution of the snps to the regions \n",
      "UTR5, CDS and UTR3 are:\n",
      "17, 187, 324, respectively.\n",
      "\n",
      "For the CDS snps,\n",
      "134 of them are synonymous and\n",
      "53 of them are non-synonymous.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"The total number of snps is {}\\n\".format(len(snp_df)))\n",
    "\n",
    "print(\"The distribution of the snps to the regions \\nUTR5, CDS and UTR3 are:\\n{}, {}, {}, respectively.\\n\".\\\n",
    "      format(len(utr5_snps), len(cds_snps), len(utr3_snps)  ))\n",
    "\n",
    "print(\"For the CDS snps,\\n{} of them are synonymous and\\n{} of them are non-synonymous.\\n\".format(synonymous_snps, non_syn_snps))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "57d3ab88",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Let's write the results\n",
    "\n",
    "snp_df.to_csv(\"./snp_dataframes/selected_snps.csv.gz\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffa99489",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9d30bfb3",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98f2af6e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "be4eb76c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "cluster_list = (cluster_1_genes, cluster_2_genes, cluster_3_genes, cluster_4_genes)\n",
    "\n",
    "for i in range(1,5):\n",
    "    this_cluster = cluster_list[i-1]\n",
    "\n",
    "    with gzip.open(\"./snp_dataframes/cluster_{}_sequences.fa.gz\".format(i), \"wt\") as output_stream,\\\n",
    "         gzip.open(\"./snp_dataframes/cluster_{}_cds_sequences.fa.gz\".format(i), \"wt\") as cds_stream,\\\n",
    "         gzip.open(\"./snp_dataframes/cluster_{}_utr3_sequences.fa.gz\".format(i), \"wt\") as utr3_stream,\\\n",
    "         gzip.open(\"./snp_dataframes/cluster_{}_utr5_sequences.fa.gz\".format(i), \"wt\") as utr5_stream:\n",
    "        for ind, contents in snp_df.iterrows():\n",
    "            if contents[\"gene\"] in this_cluster:\n",
    "                this_str = \">{}_{}\\n{}\".format(contents[\"gene\"], contents[\"position\"], contents[\"sequence\"]) \n",
    "                print(this_str, file = output_stream)\n",
    "\n",
    "                if contents[\"region\"] == \"CDS\":\n",
    "                    print(this_str, file = cds_stream)\n",
    "                if contents[\"region\"] == \"UTR3\":\n",
    "                    print(this_str, file = utr3_stream)\n",
    "                if contents[\"region\"] == \"UTR5\":\n",
    "                    print(this_str, file = utr5_stream)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "78a3cbbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene</th>\n",
       "      <th>position</th>\n",
       "      <th>maternal</th>\n",
       "      <th>paternal</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>chr_pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>region</th>\n",
       "      <th>sequence</th>\n",
       "      <th>maternal_codon</th>\n",
       "      <th>paternal_codon</th>\n",
       "      <th>maternal_AA</th>\n",
       "      <th>paternal_AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000045866</td>\n",
       "      <td>Ddx21</td>\n",
       "      <td>4410</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr10</td>\n",
       "      <td>62580562</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>TTAANTGAA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000045866</td>\n",
       "      <td>Ddx21</td>\n",
       "      <td>4399</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr10</td>\n",
       "      <td>62580573</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>ATTGNCACA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000045866</td>\n",
       "      <td>Ddx21</td>\n",
       "      <td>4063</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr10</td>\n",
       "      <td>62580909</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>CTCCNGGTA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000045866</td>\n",
       "      <td>Ddx21</td>\n",
       "      <td>3913</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr10</td>\n",
       "      <td>62581059</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>TTGGNAGGA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000045866</td>\n",
       "      <td>Ddx21</td>\n",
       "      <td>3895</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr10</td>\n",
       "      <td>62581077</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>AGGGNTCTT</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>ENSMUST00000111820</td>\n",
       "      <td>Tmppe</td>\n",
       "      <td>2175</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr9</td>\n",
       "      <td>114406531</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>GATTNCTGT</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>ENSMUST00000111820</td>\n",
       "      <td>Tmppe</td>\n",
       "      <td>2591</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr9</td>\n",
       "      <td>114406947</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>TCACNCATA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>ENSMUST00000111820</td>\n",
       "      <td>Tmppe</td>\n",
       "      <td>2708</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr9</td>\n",
       "      <td>114407064</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>TGGGNCTGA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>ENSMUST00000111820</td>\n",
       "      <td>Tmppe</td>\n",
       "      <td>2721</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr9</td>\n",
       "      <td>114407077</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>TTAGNGAGA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>ENSMUST00000111820</td>\n",
       "      <td>Tmppe</td>\n",
       "      <td>2750</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr9</td>\n",
       "      <td>114407106</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR3</td>\n",
       "      <td>CCATNGGAA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>528 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             transcript   gene  position maternal paternal chromosome  \\\n",
       "0    ENSMUST00000045866  Ddx21      4410        A        G      chr10   \n",
       "1    ENSMUST00000045866  Ddx21      4399        A        G      chr10   \n",
       "2    ENSMUST00000045866  Ddx21      4063        C        T      chr10   \n",
       "3    ENSMUST00000045866  Ddx21      3913        G        A      chr10   \n",
       "4    ENSMUST00000045866  Ddx21      3895        C        A      chr10   \n",
       "..                  ...    ...       ...      ...      ...        ...   \n",
       "523  ENSMUST00000111820  Tmppe      2175        A        G       chr9   \n",
       "524  ENSMUST00000111820  Tmppe      2591        T        G       chr9   \n",
       "525  ENSMUST00000111820  Tmppe      2708        C        T       chr9   \n",
       "526  ENSMUST00000111820  Tmppe      2721        A        T       chr9   \n",
       "527  ENSMUST00000111820  Tmppe      2750        A        G       chr9   \n",
       "\n",
       "       chr_pos strand region   sequence maternal_codon paternal_codon  \\\n",
       "0     62580562      -   UTR3  TTAANTGAA             NA             NA   \n",
       "1     62580573      -   UTR3  ATTGNCACA             NA             NA   \n",
       "2     62580909      -   UTR3  CTCCNGGTA             NA             NA   \n",
       "3     62581059      -   UTR3  TTGGNAGGA             NA             NA   \n",
       "4     62581077      -   UTR3  AGGGNTCTT             NA             NA   \n",
       "..         ...    ...    ...        ...            ...            ...   \n",
       "523  114406531      +   UTR3  GATTNCTGT             NA             NA   \n",
       "524  114406947      +   UTR3  TCACNCATA             NA             NA   \n",
       "525  114407064      +   UTR3  TGGGNCTGA             NA             NA   \n",
       "526  114407077      +   UTR3  TTAGNGAGA             NA             NA   \n",
       "527  114407106      +   UTR3  CCATNGGAA             NA             NA   \n",
       "\n",
       "    maternal_AA paternal_AA  \n",
       "0            NA          NA  \n",
       "1            NA          NA  \n",
       "2            NA          NA  \n",
       "3            NA          NA  \n",
       "4            NA          NA  \n",
       "..          ...         ...  \n",
       "523          NA          NA  \n",
       "524          NA          NA  \n",
       "525          NA          NA  \n",
       "526          NA          NA  \n",
       "527          NA          NA  \n",
       "\n",
       "[528 rows x 14 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
