{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d79bb32e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import sys\n",
    "from copy import deepcopy\n",
    "\n",
    "sys.path.insert(0, '../')\n",
    "from ref_lib.GTF import GTFfile, GTFEntry, get_gtf_contents\n",
    "from ref_lib.Fasta import FastaEntry, FastaFile\n",
    "\n",
    "from collections import defaultdict, OrderedDict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5b32b1f",
   "metadata": {},
   "source": [
    "It would be good to document this a bit more!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bc5f6556",
   "metadata": {},
   "outputs": [],
   "source": [
    "reverse_complement_dict = { \"A\" : \"T\",\n",
    "                            \"C\" : \"G\",\n",
    "                            \"G\" : \"C\",\n",
    "                            \"T\" : \"A\",\n",
    "                            \"N\" : \"N\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "19d872bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCF_FIELDS = [\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\",  \"INFO\",    \"FORMAT\",  \"CAST_EiJ\", \n",
    "              \"CHR_GENOMIC\", \"POS_GENOMIC\", \"STRAND\", \"TRANS_HEADER\", \"KOZAK_START\", \"KOZAK_END\", \"KOZAK_MOTIF\", \n",
    "              \"DUMMY_SCORE\", \"DUMMY_STRAND\"]\n",
    "\n",
    "class VcfEntry:\n",
    "\n",
    "    def __init__(self , vcf_line_contents ):\n",
    "        assert len(vcf_line_contents) >= len(VCF_FIELDS)\n",
    "        \n",
    "        self.fields = { VCF_FIELDS[i] : vcf_line_contents[i] for i in range( len(VCF_FIELDS) ) }\n",
    "        \n",
    "        \n",
    "    def __str__(self ):\n",
    "        \"\"\"\n",
    "        This needs to be rewritten \n",
    "        \"\"\"\n",
    "        return \"\\t\".join( [self.fields[f] for f in VCF_FIELDS] )\n",
    "\n",
    "\n",
    "############################################################################\n",
    "    \n",
    "class VcfFile:\n",
    "    '''\n",
    "    This is a reader for \n",
    "    '''\n",
    "    \n",
    "    def __init__(self , file):\n",
    "        myopen = open\n",
    "        if file.endswith(\".gz\"):\n",
    "            myopen = gzip.open\n",
    "\n",
    "        if(file):\n",
    "            self.f = myopen(file , \"rt\")\n",
    "        else:\n",
    "            self.f = stdin\n",
    "\n",
    "    #####################################################\n",
    "\n",
    "    def __enter__(self):\n",
    "        return self\n",
    "\n",
    "    #####################################################\n",
    "\n",
    "    def __exit__(self, exc_type, exc_val, exc_tb):\n",
    "        pass\n",
    "\n",
    "    ######################################################\n",
    "\n",
    "    def __getitem__(self, index):\n",
    "        line = self.f.readline().strip()\n",
    "        \n",
    "        while line.startswith(\"#\"):\n",
    "            line = self.f.readline().strip()\n",
    "        \n",
    "        if line == \"\":\n",
    "            raise IndexError\n",
    "        #line_contents = line.split(\"\\t\")\n",
    "        line_contents = line.split()\n",
    "        if len(line_contents) < 9:\n",
    "            raise IndexError\n",
    "        return VcfEntry(line_contents)\n",
    "                \n",
    "    #########################################################\n",
    "\n",
    "    def __del__(self):\n",
    "        self.f.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "51f9c225",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_vcf_file = \"transcriptomic_variants.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8a69713f",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_file = \"./appris_mouse_v2_riboITP_transcriptomic_variants_Kozak.vcf\"\n",
    "\n",
    "my_vcf = VcfFile(vcf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0d805b15",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_file = \"./rbp_selected_snps_intersect.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ca862844",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file   = \"../../../mouse_itp_reference/transcriptome/varnt_masked_and_filtered_mouse_transcriptome.fa.gz\" \n",
    "fasta_reader = FastaFile(fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9ffa78e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_entries = [a for a in fasta_reader]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8fdba31d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_dict = OrderedDict()\n",
    "\n",
    "for f in fasta_entries:\n",
    "    fasta_dict[f.header] = f.sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "442ab924",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_vcf = \"appris_mouse_v2_riboITP_transcriptomic_variants_Kozak_maternal_paternal_seq_added.vcf\"\n",
    "\n",
    "with open(output_vcf, \"wt\") as output_stream:\n",
    "    for entry in my_vcf:\n",
    "\n",
    "        sequence_start = int(entry.fields[\"KOZAK_START\"] )\n",
    "        sequence_end   = int(entry.fields[\"KOZAK_END\"] )\n",
    "        this_gene      = entry.fields[\"TRANS_HEADER\"]\n",
    "\n",
    "        # Note that VCF is 1-based.\n",
    "        # So we need to subtract 1 to make it 0 based\n",
    "        snp_position = int(entry.fields[\"POS\"]) - 1\n",
    "\n",
    "        snp_offset = snp_position - sequence_start \n",
    "\n",
    "        this_kozak_sequence_raw = fasta_dict[this_gene][sequence_start: sequence_end]\n",
    "\n",
    "        kozak_sequence_ref = this_kozak_sequence_raw[0: 0 +  snp_offset] +\\\n",
    "                            entry.fields[\"REF\"] +\\\n",
    "                            this_kozak_sequence_raw[ snp_offset + 1 :]\n",
    "\n",
    "        kozak_sequence_alt = this_kozak_sequence_raw[0: 0 +  snp_offset] +\\\n",
    "                            entry.fields[\"ALT\"] +\\\n",
    "                            this_kozak_sequence_raw[ snp_offset + 1 :]\n",
    "                            \n",
    "        print( \"{}\\t{}\\t{}\".format(entry, kozak_sequence_ref, kozak_sequence_alt), file=output_stream )\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45cd44d8",
   "metadata": {},
   "source": [
    "## Bed file Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "091795eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCF_FIELDS = [\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\",  \"INFO\",    \"FORMAT\",  \"CAST_EiJ\", \n",
    "              \"CHR_GENOMIC\", \"POS_GENOMIC\", \"STRAND\"]\n",
    "\n",
    "class VcfEntry:\n",
    "\n",
    "    def __init__(self , vcf_line_contents ):\n",
    "        assert len(vcf_line_contents) >= len(VCF_FIELDS)\n",
    "        \n",
    "        self.fields = { VCF_FIELDS[i] : vcf_line_contents[i] for i in range( len(VCF_FIELDS) ) }\n",
    "        \n",
    "        \n",
    "    def __str__(self ):\n",
    "        \"\"\"\n",
    "        This needs to be rewritten \n",
    "        \"\"\"\n",
    "        return \"\\t\".join( [self.fields[f] for f in VCF_FIELDS] )\n",
    "\n",
    "\n",
    "############################################################################\n",
    "    \n",
    "class VcfFile:\n",
    "    '''\n",
    "    This is a reader for \n",
    "    '''\n",
    "    \n",
    "    def __init__(self , file):\n",
    "        myopen = open\n",
    "        if file.endswith(\".gz\"):\n",
    "            myopen = gzip.open\n",
    "\n",
    "        if(file):\n",
    "            self.f = myopen(file , \"rt\")\n",
    "        else:\n",
    "            self.f = stdin\n",
    "\n",
    "    #####################################################\n",
    "\n",
    "    def __enter__(self):\n",
    "        return self\n",
    "\n",
    "    #####################################################\n",
    "\n",
    "    def __exit__(self, exc_type, exc_val, exc_tb):\n",
    "        pass\n",
    "\n",
    "    ######################################################\n",
    "\n",
    "    def __getitem__(self, index):\n",
    "        line = self.f.readline().strip()\n",
    "        \n",
    "        while line.startswith(\"#\"):\n",
    "            line = self.f.readline().strip()\n",
    "        \n",
    "        if line == \"\":\n",
    "            raise IndexError\n",
    "        #line_contents = line.split(\"\\t\")\n",
    "        line_contents = line.split()\n",
    "        if len(line_contents) < 9:\n",
    "            raise IndexError\n",
    "        return VcfEntry(line_contents)\n",
    "                \n",
    "    #########################################################\n",
    "\n",
    "    def __del__(self):\n",
    "        self.f.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "14f2c11d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "vcf_chr_dict = defaultdict(dict)\n",
    "\n",
    "main_vcf = VcfFile(main_vcf_file)\n",
    "\n",
    "for entry in main_vcf:\n",
    "\n",
    "    vcf_chr_dict[entry.fields[\"CHR_GENOMIC\"]][ int(entry.fields[\"POS_GENOMIC\"]) ] = entry\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a7d0da51",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t42851330\t42851337\tTbg97262300\tchr1\t42851336\t42851337\n",
      "chr1\t42851331\t42851338\tELAVL1\tchr1\t42851336\t42851337\n",
      "chr1\t42851331\t42851338\tRbp9\tchr1\t42851336\t42851337\n",
      "chr1\t42851331\t42851338\tZC3H14CONSTRUCT\tchr1\t42851336\t42851337\n",
      "chr1\t42851331\t42851338\telav\tchr1\t42851336\t42851337\n",
      "chr1\t42851331\t42851338\tfne\tchr1\t42851336\t42851337\n",
      "chr1\t42862479\t42862486\tNOVA1\tchr1\t42862485\t42862486\n",
      "chr1\t42862480\t42862487\tDAZAP1\tchr1\t42862485\t42862486\n",
      "chr1\t42862480\t42862487\tMAL8P140\tchr1\t42862485\t42862486\n",
      "chr1\t42862480\t42862487\tRBM41\tchr1\t42862485\t42862486\n"
     ]
    }
   ],
   "source": [
    "! head './rbp_selected_snps_intersect.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "4bd3161f",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_bed_file = \"./rbp_selected_snps_intersect_maternal_paternal_seq_added.bed\"\n",
    "\n",
    "with open(bed_file, \"rt\") as input_stream,\\\n",
    "     open(output_bed_file, \"wt\") as output_stream:\n",
    "    i = 0\n",
    "\n",
    "    for line in input_stream:\n",
    "        \n",
    "        contents = line.strip().split()\n",
    "\n",
    "        if len(contents) < 6:\n",
    "            continue\n",
    "\n",
    "        seq_start_genomic = int(contents[1])\n",
    "        seq_end_genomic   = int(contents[2])\n",
    "        chr_genomic       = contents[0] \n",
    "        snp_pos_genomic   = int(contents[6]) \n",
    "\n",
    "        # We need this (-1) adjustment because VCF is 1-based\n",
    "        left_offset  = (snp_pos_genomic -1 ) - seq_start_genomic \n",
    "        right_offset = seq_end_genomic - (snp_pos_genomic -1)\n",
    "        \n",
    "        vcf_entry = vcf_chr_dict[chr_genomic][snp_pos_genomic]\n",
    "\n",
    "        snp_pos_trans   = int(vcf_entry.fields[\"POS\"])\n",
    "\n",
    "        # snp_pos_trans is also 1-based so we need -1\n",
    "        seq_start_trans = (snp_pos_trans -1) - left_offset\n",
    "        seq_end_trans   = (snp_pos_trans -1) + right_offset \n",
    "\n",
    "        raw_sequence = fasta_dict[ vcf_entry.fields[\"CHROM\"] ][seq_start_trans: seq_end_trans]\n",
    "\n",
    "        snp_local_offset = left_offset\n",
    "\n",
    "        ref_sequence = raw_sequence[:snp_local_offset] + vcf_entry.fields[\"REF\"] + raw_sequence[snp_local_offset + 1:]\n",
    "        alt_sequence = raw_sequence[:snp_local_offset] + vcf_entry.fields[\"ALT\"] + raw_sequence[snp_local_offset + 1:]\n",
    "\n",
    "\n",
    "        #print(str(vcf_entry))\n",
    "        #print(\"----   {}   {} -------\".format(seq_start_trans, seq_end_trans))\n",
    "        #print(vcf_entry.fields[\"CHROM\"] )\n",
    "        print( \"{}\\t{}\\t{}\".format(line.strip(), ref_sequence, alt_sequence) , file = output_stream)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
