---
title: "Imprinted Genes"
output: html_document
---

We determine imprinted genes first. Then we compare the translation efficiency of imprinted genes to those of background genes.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r include,echo=FALSE,warning=FALSE}
library(cowplot)
library(ggplot2)
library(dplyr)
library(Cairo)
library(edgeR)
library(MatchIt)
library(Seurat)
```

```{r constants,echo=FALSE,warning=FALSE}
CONSTANT_ADDITION = 1
```


```{r setwd,echo=FALSE,warning=FALSE}
setwd("~/projects/mice_cell_development/mouse-itp/imprinted_genes/")
#setwd("/data/projects/mice_cell_development/mouse-itp/imprinted_genes/")
```

```{r filepaths,echo=FALSE,warning=FALSE}
published_imprinted_gene_list_file = "./imprinted_gene_list_41467_2021_23510_MOESM4_ESM.csv"


# Another imprinted gene list from
# https://www.cell.com/cell/fulltext/S0092-8674(19)30106-0?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867419301060%3Fshowall%3Dtrue#%20
tucci_et_al_imprinted_gene_list_file = "./tucci_et_al_2019_mouse_imprinted_genes.csv"

four_cell_all_snp_counts_file  = "./fourcell_all_snp_counts.csv.gz"
eight_cell_all_snp_counts_file = "./eightcell_all_snp_counts.csv.gz" 

ribo_rna_counts_file = "../cds_counts.csv.gz"

#region_lengths_file = "../../mouse_itp_reference/transcriptome/appris_mouse_v2_filtered_regions.bed"

region_lengths_file = "../appris_mouse_v2_filtered_regions.bed"
 

```

```{r region_lengths,echo=FALSE,warning=FALSE}

# Here we obtain the cds lengths

region_lengths_pre = read.csv(region_lengths_file, sep = "\t", header = FALSE)

colnames(region_lengths_pre) = c("gene", "start", "end", "region", "score", "strand")

get_gene_name = function(long_name){
   return(strsplit(long_name,  split= "|", fixed = TRUE) )
}

region_lengths_pre$gene = unlist( lapply( get_gene_name(region_lengths_pre$gene)  , "[[", 6)  )

region_lengths_pre = region_lengths_pre[ region_lengths_pre$region == "CDS" ,]

cds_lengths = region_lengths_pre$end - region_lengths_pre$start

region_lengths = data.frame(gene = region_lengths_pre$gene, cds_length = cds_lengths)

```

## Publisehd Imprinted Genes

[Santini et al ](https://www.nature.com/articles/s41467-021-23510-4#MOESM4) studies mouse embryos and provide a list of imprinted genes.
However their list doesn't indicate any direction (maternal or paternal).

We define **paternal ratio** as

$$
\frac{ \mbox{PATERNAL READS + CONSTANT} }{ \mbox{MATERNAL READS + PATERNAL READS + 2*CONSTANT}  }
$$

where constant is a positive integer. Namely, we are taking the ratio of paternal reads to the total reads after adding a constant to both maternal and paternal reads (hence the coefficinet 2 in the denominatior).

```{r echo=FALSE,warning=FALSE}
four_cell_all_snp_counts  = read.csv(four_cell_all_snp_counts_file)
eight_cell_all_snp_counts = read.csv(eight_cell_all_snp_counts_file)

four_cell_all_snp_counts$paternal_ratio  = 
  (four_cell_all_snp_counts$paternal_y + CONSTANT_ADDITION  ) / (four_cell_all_snp_counts$maternal_y + four_cell_all_snp_counts$paternal_y + 2* CONSTANT_ADDITION)

eight_cell_all_snp_counts$paternal_ratio = 
  (eight_cell_all_snp_counts$paternal_y + CONSTANT_ADDITION) / (eight_cell_all_snp_counts$maternal_y + eight_cell_all_snp_counts$paternal_y + 2*CONSTANT_ADDITION)

published_nature_gen_df = read.csv(published_imprinted_gene_list_file)

published_nature_gen_df
```

[Tucci et. al.](https://www.cell.com/cell/fulltext/S0092-8674(19)30106-0?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867419301060%3Fshowall%3Dtrue#%20)
however, indicates the direction, though their list is less comprehensive.

```{r echo=FALSE,warning=FALSE}

tucci_et_al_imprinted_gene_df_pre = read.csv(tucci_et_al_imprinted_gene_list_file)

# get rid of the isoform-psecific entries
tucci_et_al_imprinted_gene_df = tucci_et_al_imprinted_gene_df_pre[tucci_et_al_imprinted_gene_df_pre$ExpressedAllele %in% c("P", "M") ,]

tucci_et_al_imprinted_gene_df
```

## Definition of Imprinted genes

?? To be completed ??

## Distribution of SNP RNA Counts

We sum the reads overlapping with SNPs for each gene. Then we aggregate the read counts across replicates.
So, we obtain a single count for each gene in each stage of the embryo.
We want to be confident with our imprinted gene picks. So we filter out genes having less than 10 counts.

Please note that these are different (less) than regular RNA-Seq counts as we use the reads overlapping with SNPs only.

On the other hand, for TE calculations, we use all the reads. In either case, we user the reads mapping to the CDS region only.

```{r echo=FALSE,warning=FALSE}

fourcel_rna_total = four_cell_all_snp_counts$paternal_y + four_cell_all_snp_counts$maternal_y

eight_cell_rna_total = eight_cell_all_snp_counts$paternal_y + eight_cell_all_snp_counts$maternal_y

hist( log2( fourcel_rna_total[fourcel_rna_total>=0 ] ) )

hist( log2( eight_cell_rna_total[ eight_cell_rna_total >= 0 ] ) )

hist( log2( fourcel_rna_total[fourcel_rna_total>=10 ] ) )

hist( log2( eight_cell_rna_total[ eight_cell_rna_total >= 10 ] ) )

```
We pick genes with total count of at least 10.

```{r echo=FALSE,warning=FALSE}

four_cell_snp_filtered  = four_cell_all_snp_counts[ (four_cell_all_snp_counts$paternal_y + four_cell_all_snp_counts$maternal_y) >= 10, ]

eight_cell_snp_filtered = eight_cell_all_snp_counts[ (eight_cell_all_snp_counts$paternal_y + eight_cell_all_snp_counts$maternal_y) >= 10,]

print("Four cell SNP (>=10)  genes:")
print(dim(four_cell_snp_filtered)[1]  )
print("Eight cell SNP (>=10)  genes:")
print(dim(eight_cell_snp_filtered)[1])

hist( log2( four_cell_snp_filtered$paternal_ratio )  )
hist( log2( eight_cell_snp_filtered$paternal_ratio )  )
```

```{r echo=FALSE,warning=FALSE}

TAIL_SIZE   = 0.05
quant_probs = probs = c(TAIL_SIZE, 1 - TAIL_SIZE)

fourcell_quantiles = quantile( four_cell_snp_filtered$paternal_ratio , quant_probs  )

eightcell_quantiles = quantile( eight_cell_snp_filtered$paternal_ratio, quant_probs  )

print("Quantiles are (4 and 8 cells)")
print(fourcell_quantiles)
print(eightcell_quantiles)

fourcell_paternal_genes = four_cell_snp_filtered[ four_cell_snp_filtered$paternal_ratio >= fourcell_quantiles[2] , ]$gene
fourcell_maternal_genes = four_cell_snp_filtered[ four_cell_snp_filtered$paternal_ratio <= fourcell_quantiles[1] , ]$gene

print("Four cell paternal and maternal genes are:")
print( c(length(fourcell_paternal_genes), length(fourcell_maternal_genes) )  )


eightcell_paternal_genes = eight_cell_snp_filtered[ eight_cell_snp_filtered$paternal_ratio >= eightcell_quantiles[2] , ]$gene
eightcell_maternal_genes = eight_cell_snp_filtered[ eight_cell_snp_filtered$paternal_ratio <= eightcell_quantiles[1] , ]$gene

print("Eight cell paternal and maternal genes are:")
print( c(length(eightcell_paternal_genes), length(eightcell_maternal_genes) )  )

```

## Comparison of Our Imprinted Genes to the Published Ones



```{r echo=FALSE,warning=FALSE}

nature_intersection_df = merge(published_nature_gen_df, region_lengths, by, by.x = "Gene.symbol", by.y = "gene" )
published_nature_gen_df
nature_intersection_df

fourcell_common_genes_with_nature = intersect( nature_intersection_df$Gene.symbol, union( fourcell_maternal_genes, fourcell_paternal_genes  )   )

print("4 Cell: Common imprinted genes with the Santini et. al. paper:")
print(fourcell_common_genes_with_nature)

print("8 Cell: Common imprinted genes with the Santini et. al. paper:")
eightcell_common_genes_with_nature = intersect( nature_intersection_df$Gene.symbol, union( eightcell_maternal_genes, eightcell_paternal_genes  )   )

eightcell_common_genes_with_nature

#published_nature_gen_df
#fourcell_tucci_merged = merge(four_cell_snp_filtered, tucci_et_al_imprinted_gene_df, by.x = "gene", by.y = "Gene")
#fourcell_tucci_merged$gene

#eightcell_tucci_merged = merge(eight_cell_snp_filtered, tucci_et_al_imprinted_gene_df, by.x = "gene", by.y = "Gene")
#eightcell_tucci_merged$gene
```


```{r echo=FALSE,warning=FALSE}

a = merge(published_nature_gen_df, eight_cell_snp_filtered, by.x = "Gene.symbol", by.y = "gene")
a

```


## Translation Efficiency of Imprinted Genes 

```{r echo=FALSE,warning=FALSE}
ribo_rna_counts = read.csv(ribo_rna_counts_file )

ribo_rna_gene_names        = unlist(lapply( (strsplit(ribo_rna_counts$transcript, split = "-" ) ), "[[", 1  ) )

ribo_rna_counts$transcript = ribo_rna_gene_names

ribo_rna_counts
```


We apply CLR normalization to our rna and ribo data
```{r echo=TRUE,warning=FALSE}
set.seed(3)

clr_normalized_counts_pre = NormalizeData(ribo_rna_counts[,-1], 
                                          scale.factor = 10000, margin = 2, 
                                          normalization.method= "CLR",verbose = TRUE)

clr_normalized_counts = data.frame(transcript = ribo_rna_counts$transcript , clr_normalized_counts_pre)

colSums(clr_normalized_counts_pre)
print("\n\n##################\n\n")
colSums(ribo_rna_counts[,-1])
```

```{r echo=FALSE,warning=FALSE}

rnaseq_fourcell_averages_clr  = (clr_normalized_counts$X4cell.1.RNAseq + clr_normalized_counts$X4cell.2.RNAseq) / 2
rnaseq_eightcell_averages_clr = (clr_normalized_counts$X8cell.1.RNAseq + clr_normalized_counts$X8cell.2.RNAseq + clr_normalized_counts$X8cell.3.RNAseq + clr_normalized_counts$X8cell.4.RNAseq) / 4
ribo_fourcell_averages_clr    = (clr_normalized_counts$X4cell.1 + clr_normalized_counts$X4cell.2 + clr_normalized_counts$X4cell.3) / 3
ribo_eightcell_averages_clr   = (clr_normalized_counts$X8cell.1 + clr_normalized_counts$X8cell.2 + clr_normalized_counts$X8cell.3 + clr_normalized_counts$X8cell.4) / 4

fourcell_TE_clr  = ribo_fourcell_averages_clr / rnaseq_fourcell_averages_clr
eightcell_TE_clr = ribo_eightcell_averages_clr / rnaseq_eightcell_averages_clr

hist( log2(fourcell_TE_clr) )

hist( log2(eightcell_TE_clr) )

```

```{r echo=FALSE,warning=FALSE}

TE_df_clr = data.frame( transcript = ribo_rna_counts$transcript, fourcell_TE_clr = fourcell_TE_clr, eightcell_TE_clr = eightcell_TE_clr  )

fourcell_background_genes = setdiff( TE_df_clr$transcript , union( fourcell_maternal_genes, fourcell_paternal_genes  ) )

eightcell_background_genes = setdiff( TE_df_clr$transcript , union( eightcell_maternal_genes, eightcell_paternal_genes  ) )


fourcell_paternal_TE_clr   = TE_df_clr[ TE_df_clr$transcript %in% fourcell_paternal_genes,]$fourcell_TE_clr
fourcell_maternal_TE_clr   = TE_df_clr[ TE_df_clr$transcript %in% fourcell_maternal_genes,]$fourcell_TE_clr
fourcell_background_TE_clr = TE_df_clr[ TE_df_clr$transcript %in% fourcell_background_genes,]$fourcell_TE_clr


fourcell_paternal_rna_clr   = rnaseq_fourcell_averages_clr[ TE_df_clr$transcript %in% fourcell_paternal_genes    ]
fourcell_maternal_rna_clr   = rnaseq_fourcell_averages_clr[ TE_df_clr$transcript %in% fourcell_maternal_genes    ]
fourcell_background_rna_clr = rnaseq_fourcell_averages_clr[ TE_df_clr$transcript %in% fourcell_background_genes  ]

fourcell_arranged_TE_clr = c( fourcell_background_TE_clr, fourcell_paternal_TE_clr, fourcell_maternal_TE_clr )

fourcell_arranged_rna_clr = c(fourcell_background_rna_clr, fourcell_paternal_rna_clr, fourcell_maternal_rna_clr )

fourcell_TE_labels_clr = c(
  rep("background", length(fourcell_background_TE_clr)) ,
  rep("paternal", length(fourcell_paternal_TE_clr) ),
  rep("maternal",  length(fourcell_maternal_TE_clr) )
)

fourcell_TE_imprinted_clr = c(
  rep(0, length(fourcell_background_TE_clr)) ,
  rep(1, length(fourcell_paternal_TE_clr) ),
  rep(1,  length(fourcell_maternal_TE_clr) )
)




fourcell_paternal_genes_clr   = TE_df_clr[ TE_df_clr$transcript %in% fourcell_paternal_genes,]$transcript
fourcell_maternal_genes_clr   = TE_df_clr[ TE_df_clr$transcript %in% fourcell_maternal_genes,]$transcript
fourcell_background_genes_clr = TE_df_clr[ TE_df_clr$transcript %in% fourcell_background_genes,]$transcript

fourcell_arranged_genes_clr = c(
  fourcell_background_genes_clr,
  fourcell_paternal_genes_clr,
  fourcell_maternal_genes_clr
)


fourcell_TE_arranged_df_clr = data.frame( type       = fourcell_TE_labels_clr,
                                      imprinted  = fourcell_TE_imprinted_clr,
                                      transcript = fourcell_arranged_genes_clr,
                                      TE         = fourcell_arranged_TE_clr,
                                      rna        = fourcell_arranged_rna_clr )





sample_fourcell_clr = matchit(imprinted~rna, data = fourcell_TE_arranged_df_clr )

sample_fourcell_clr


matched_fourcell_clr = match.data(sample_fourcell_clr)

matched_fourcell_clr

boxplot( log2(matched_fourcell_clr$rna) ~ matched_fourcell_clr$imprinted )
boxplot( (matched_fourcell_clr$rna) ~ matched_fourcell_clr$imprinted )

boxplot( log2(matched_fourcell_clr$rna) ~ matched_fourcell_clr$type )

boxplot( log2(matched_fourcell_clr$TE) ~ matched_fourcell_clr$type )

boxplot( (matched_fourcell_clr$TE) ~ matched_fourcell_clr$type )



```

```{r echo=FALSE,warning=FALSE}



wilcox.test( matched_fourcell_clr[matched_fourcell_clr$type == "background",]$TE, matched_fourcell_clr[matched_fourcell_clr$type == "maternal",]$TE  )

wilcox.test( matched_fourcell_clr[matched_fourcell_clr$type == "background",]$TE, matched_fourcell_clr[matched_fourcell_clr$type == "paternal",]$TE  )

wilcox.test( matched_fourcell_clr[matched_fourcell_clr$type == "paternal",]$TE, matched_fourcell_clr[matched_fourcell_clr$type == "maternal",]$TE  )

```


```{r echo=FALSE,warning=FALSE}

fourcell_TE_and_lengths = merge(matched_fourcell_clr, region_lengths, by.x = "transcript", by.y = "gene")

hist( log2(fourcell_TE_and_lengths$cds_length) )

hist( log2(cds_lengths) )

median( fourcell_TE_and_lengths$cds_length )

median( cds_lengths )

boxplot( (fourcell_TE_and_lengths$cds_length) ~ fourcell_TE_and_lengths$type )


hist( log2( fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "maternal",]$cds_length  ) )

hist( log2( fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "background",]$cds_length  ) )


plot(density( log2( fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "maternal",]$cds_length  )   ) )

lines(density( log2( fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "background",]$cds_length  )   ), col = "red" )
```


```{r echo=FALSE,warning=FALSE}

wilcox.test(  fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "maternal",]$cds_length,  fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "background",]$cds_length  )

wilcox.test( fourcell_TE_and_lengths$cds_length, cds_lengths  )
```





```{r echo=FALSE,warning=FALSE}

eightcell_background_genes = setdiff( TE_df_clr$transcript , union( eightcell_maternal_genes, eightcell_paternal_genes  ) )

eightcell_background_genes = setdiff( TE_df_clr$transcript , union( eightcell_maternal_genes, eightcell_paternal_genes  ) )


eightcell_paternal_TE_clr   = TE_df_clr[ TE_df_clr$transcript %in% eightcell_paternal_genes,]$eightcell_TE_clr
eightcell_maternal_TE_clr   = TE_df_clr[ TE_df_clr$transcript %in% eightcell_maternal_genes,]$eightcell_TE_clr
eightcell_background_TE_clr = TE_df_clr[ TE_df_clr$transcript %in% eightcell_background_genes,]$eightcell_TE_clr


eightcell_paternal_rna_clr   = rnaseq_eightcell_averages_clr[ TE_df_clr$transcript %in% eightcell_paternal_genes    ]
eightcell_maternal_rna_clr   = rnaseq_eightcell_averages_clr[ TE_df_clr$transcript %in% eightcell_maternal_genes    ]
eightcell_background_rna_clr = rnaseq_eightcell_averages_clr[ TE_df_clr$transcript %in% eightcell_background_genes  ]

eightcell_arranged_TE_clr = c( eightcell_background_TE_clr, eightcell_paternal_TE_clr, eightcell_maternal_TE_clr )

eightcell_arranged_rna_clr = c(eightcell_background_rna_clr, eightcell_paternal_rna_clr, eightcell_maternal_rna_clr )

eightcell_TE_labels_clr = c(
  rep("background", length(eightcell_background_TE_clr)) ,
  rep("paternal", length(eightcell_paternal_TE_clr) ),
  rep("maternal",  length(eightcell_maternal_TE_clr) )
)

eightcell_TE_imprinted_clr = c(
  rep(0, length(eightcell_background_TE_clr)) ,
  rep(1, length(eightcell_paternal_TE_clr) ),
  rep(1,  length(eightcell_maternal_TE_clr) )
)




eightcell_paternal_genes_clr   = TE_df_clr[ TE_df_clr$transcript %in% eightcell_paternal_genes,]$transcript
eightcell_maternal_genes_clr   = TE_df_clr[ TE_df_clr$transcript %in% eightcell_maternal_genes,]$transcript
eightcell_background_genes_clr = TE_df_clr[ TE_df_clr$transcript %in% eightcell_background_genes,]$transcript

eightcell_arranged_genes_clr = c(
  eightcell_background_genes_clr,
  eightcell_paternal_genes_clr,
  eightcell_maternal_genes_clr
)

eightcell_TE_arranged_df_clr = data.frame( type       = eightcell_TE_labels_clr,
                                      imprinted  = eightcell_TE_imprinted_clr,
                                      transcript = eightcell_arranged_genes_clr,
                                      TE         = eightcell_arranged_TE_clr,
                                      rna        = eightcell_arranged_rna_clr )





sample_eightcell_clr = matchit(imprinted~rna, data = eightcell_TE_arranged_df_clr )

sample_eightcell_clr


matched_eightcell_clr = match.data(sample_eightcell_clr)

matched_eightcell_clr

boxplot( log2(matched_eightcell_clr$rna) ~ matched_eightcell_clr$imprinted )

boxplot( log2(matched_eightcell_clr$rna) ~ matched_eightcell_clr$type )

boxplot( log2(matched_eightcell_clr$TE) ~ matched_eightcell_clr$type )

boxplot( (matched_eightcell_clr$TE) ~ matched_eightcell_clr$type )

```


```{r echo=FALSE,warning=FALSE}



wilcox.test( matched_eightcell_clr[matched_eightcell_clr$type == "background",]$TE, matched_eightcell_clr[matched_eightcell_clr$type == "maternal",]$TE  )

wilcox.test( matched_eightcell_clr[matched_eightcell_clr$type == "background",]$TE, matched_eightcell_clr[matched_eightcell_clr$type == "paternal",]$TE  )

wilcox.test( matched_eightcell_clr[matched_eightcell_clr$type == "paternal",]$TE, matched_eightcell_clr[matched_eightcell_clr$type == "maternal",]$TE  )

```





























