---
title: "Imprinted Genes"
output: html_document
---

We determine imprinted genes first. Then we compare the translation efficiency of imprinted genes to those of background genes.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r include,echo=FALSE,warning=FALSE}
library(cowplot)
library(ggplot2)
library(dplyr)
library(Cairo)
library(edgeR)
library(MatchIt)
library(Seurat)
library(VennDiagram)
library(grDevices)
library(ggpubr)
library(RColorBrewer)
```

```{r constants,echo=FALSE,warning=FALSE}
CONSTANT_ADDITION = 1

FONT_LABEL_SIZE = 8
FONT_TITLE_SIZE = 9

PDF_resolution = 600
FIGURE_FONT = "helvetica"
```


```{r setwd,echo=FALSE,warning=FALSE}

#setwd("~/projects/mice_cell_development/mouse-itp/imprinted_genes/")
#setwd("/data/projects/mice_cell_development/mouse-itp/imprinted_genes/")

MBOOK = F

if(MBOOK){

    setwd("~/projects/mice_cell_development/mouse-itp/imprinted_genes/")

    PLOT_TITLE = element_text(hjust = 0.5,  face = "plain", size = FONT_TITLE_SIZE)
    AXIS_TITLE  = element_text(face = "plain", size = FONT_LABEL_SIZE)
}else{
    setwd("/data/projects/mice_cell_development/mouse-itp/imprinted_genes/")

    PLOT_TITLE = element_text(hjust = 0.5, family = FIGURE_FONT, face = "plain", size = FONT_TITLE_SIZE)
    AXIS_TITLE  = element_text(family = FIGURE_FONT, face = "plain", size = FONT_LABEL_SIZE)

}

```

```{r filepaths,echo=FALSE,warning=FALSE}
published_imprinted_gene_list_file = "./imprinted_gene_list_41467_2021_23510_MOESM4_ESM.csv"


# Another imprinted gene list from
# https://www.cell.com/cell/fulltext/S0092-8674(19)30106-0?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867419301060%3Fshowall%3Dtrue#%20
tucci_et_al_imprinted_gene_list_file = "./tucci_et_al_2019_mouse_imprinted_genes.csv"

four_cell_all_snp_counts_file  = "./fourcell_all_snp_counts.csv.gz"
eight_cell_all_snp_counts_file = "./eightcell_all_snp_counts.csv.gz" 

ribo_rna_counts_file = "../cds_counts.csv.gz"

chrx_genes_file = "./chrx_genes.tsv"

#region_lengths_file = "../../mouse_itp_reference/transcriptome/appris_mouse_v2_filtered_regions.bed"

region_lengths_file = "../appris_mouse_v2_filtered_regions.bed"
 
detailed_snp_file = "../snp/notebooks/snp_dataframes/rnaseq_detailed_snps.csv.gz"
GSE80810_allelic_ratio_file = "./GSE80810_allelic_ratio.csv"
```


```{r plot_constants,echo=FALSE,warning=FALSE}
BURNT_ORANGE = "#bf5700"
UT_BLUE      = "#005f86"

PERCENTAGE_BARPLOT_PATERNAL_COLOR =  "#f0756e"
PERCENTAGE_BARPLOT_MATERNAL_COLOR = "#7ccba2"
PERCENTAGE_BARPLOT_OTHER_COLOR    = "#045175"

CDS_GREEN  = rgb(124, 203, 162, maxColorValue = 255)
UTR5_BLUE  = rgb(4,   82,  117, maxColorValue = 255)
UTR3_GREEN = rgb(240, 116, 110,maxColorValue = 255)

#PERCENTAGE_BARPLOT_PATERNAL_COLOR =  "#999999"
#PERCENTAGE_BARPLOT_MATERNAL_COLOR = "#E69F00"
#PERCENTAGE_BARPLOT_OTHER_COLOR    = "#56B4E9"

PERCENTAGE_BARPLOT_COLORS = c(PERCENTAGE_BARPLOT_OTHER_COLOR,
                              PERCENTAGE_BARPLOT_MATERNAL_COLOR,
                              PERCENTAGE_BARPLOT_PATERNAL_COLOR
                              )




#MAIN_PERCENTAGE_RNASEQ_FILL_COLOR     = "skyblue"
#MAIN_PERCENTAGE_RIBOSEQ_FILL_COLOR    = "#abe39d"
#MAIN_PERCENTAGE_ERRORBAR_COLOR = "#413bb8"

FOURCELL_BACKGROUND_COLOR = "#d9d9d9"

COUNT_NORMALIZATION_FACTOR = 10000

#RATIO_RIBO_COLOUR = "orange"
#RATIO_RNA_COLOUR  = "blue"

RATIO_RIBO_COLOUR = BURNT_ORANGE
RATIO_RNA_COLOUR  = "navy"

PERCENTAGE_DASHED_COLOR = "#088f99"

################################################################################
#########                 F O N T   S I Z E S                          #########

FONT_LABEL_SIZE = 8
FONT_TITLE_SIZE = 9

PDF_resolution = 600
FIGURE_FONT = "helvetica"

DETAILED_BARPLOT_FONT_SIZE = 7
```

```{r region_lengths,echo=FALSE,warning=FALSE}

# Here we obtain the cds lengths

region_lengths_pre = read.csv(region_lengths_file, sep = "\t", header = FALSE)

colnames(region_lengths_pre) = c("gene", "start", "end", "region", "score", "strand")

get_gene_name = function(long_name){
   return(strsplit(long_name,  split= "|", fixed = TRUE) )
}

region_lengths_pre$gene = unlist( lapply( get_gene_name(region_lengths_pre$gene)  , "[[", 6)  )

region_lengths_pre      = region_lengths_pre[ region_lengths_pre$region == "CDS" ,]

cds_lengths             = region_lengths_pre$end - region_lengths_pre$start

region_lengths          = data.frame(gene = region_lengths_pre$gene, cds_length = cds_lengths)

```

```{r plot_save_functions,echo=FALSE,warning=FALSE}
get_output_file_path = function(file_name, output_folder = "pdf"){
  this_path = paste( output_folder, file_name, sep = "/"  )
  return(this_path)
}

save_plot_pdf = function(filename, this_plot, width = NA, height = NA){
  this_file = get_output_file_path(filename)
  print(this_file)
  ggsave(this_file, 
         plot   = this_plot, 
         device = cairo_pdf, 
         width  = width,
         height = height,
         dpi    = PDF_resolution )
  
}
```


## Publisehd Imprinted Genes

We define **paternal ratio** as

$$
\frac{ \mbox{PATERNAL READS + CONSTANT} }{ \mbox{MATERNAL READS + PATERNAL READS + 2*CONSTANT}  }
$$

where constant is a positive integer. Namely, we are taking the ratio of paternal reads to the total reads after adding a constant to both maternal and paternal reads (hence the coefficient 2 in the denominatior).

```{r echo=FALSE,warning=FALSE}
four_cell_all_snp_counts  = read.csv(four_cell_all_snp_counts_file)
eight_cell_all_snp_counts = read.csv(eight_cell_all_snp_counts_file)

four_cell_all_snp_counts$paternal_ratio  = 
  (four_cell_all_snp_counts$paternal_y + CONSTANT_ADDITION  ) / (four_cell_all_snp_counts$maternal_y + four_cell_all_snp_counts$paternal_y + 2* CONSTANT_ADDITION)

eight_cell_all_snp_counts$paternal_ratio = 
  (eight_cell_all_snp_counts$paternal_y + CONSTANT_ADDITION) / (eight_cell_all_snp_counts$maternal_y + eight_cell_all_snp_counts$paternal_y + 2*CONSTANT_ADDITION)

```


## Distribution of SNP RNA Counts

We sum the reads overlapping with SNPs for each gene. Then we aggregate the read counts across replicates.
So, we obtain a single count for each gene in each stage of the embryo.
We want to be confident with our imprinted gene picks. So we filter out genes having less than 10 counts.

Please note that these are different (less) than regular RNA-Seq counts as we use the reads overlapping with SNPs only.

On the other hand, for TE calculations, we use all the reads. In either case, we user the reads mapping to the CDS region only.

```{r echo=FALSE,warning=FALSE}

fourcel_rna_total = four_cell_all_snp_counts$paternal_y + four_cell_all_snp_counts$maternal_y

eight_cell_rna_total = eight_cell_all_snp_counts$paternal_y + eight_cell_all_snp_counts$maternal_y

hist( log2( fourcel_rna_total[fourcel_rna_total>=0 ] ) )

hist( log2( eight_cell_rna_total[ eight_cell_rna_total >= 0 ] ) )

hist( log2( fourcel_rna_total[fourcel_rna_total>=10 ] ) )

hist( log2( eight_cell_rna_total[ eight_cell_rna_total >= 10 ] ) )

```
We pick genes with total count of at least 10.

```{r echo=FALSE,warning=FALSE}

four_cell_snp_filtered_pre  = four_cell_all_snp_counts[ (four_cell_all_snp_counts$paternal_y + four_cell_all_snp_counts$maternal_y) >= 10, ]

eight_cell_snp_filtered_pre = eight_cell_all_snp_counts[ (eight_cell_all_snp_counts$paternal_y + eight_cell_all_snp_counts$maternal_y) >= 10,]

print("Four cell SNP (>=10)  genes:")
print(dim(four_cell_snp_filtered_pre)[1]  )
print("Eight cell SNP (>=10)  genes:")
print(dim(eight_cell_snp_filtered_pre)[1])

hist( ( four_cell_snp_filtered_pre$paternal_ratio )  )
hist( ( eight_cell_snp_filtered_pre$paternal_ratio )  )
```
### Genes from the X Chromosome

Note that we are excluding genes coming from the X chromosome

```{r echo=FALSE,warning=FALSE}

chrx_df = read.csv(chrx_genes_file, sep = "\t", header = F)

chrx_genes = chrx_df[, 1]

chrx_eight_cell_intersection_count = length( intersect(eight_cell_snp_filtered_pre$gene, chrx_genes) )
chrx_four_cell_intersection_count  = length( intersect(four_cell_snp_filtered_pre$gene, chrx_genes) )

print("Genes of the chrX:")
print( c("Intersection with the 4cell expressed genes is", chrx_four_cell_intersection_count)  )
print( c("Intersection with the 8cell expressed genes is", chrx_eight_cell_intersection_count)  )


eight_cell_snp_filtered = eight_cell_snp_filtered_pre[!(eight_cell_snp_filtered_pre$gene %in% chrx_genes ) , ]
four_cell_snp_filtered  = four_cell_snp_filtered_pre[ !(four_cell_snp_filtered_pre$gene %in% chrx_genes), ]


```

## A simple way of finding Imprinted Genes

After combining the reads of the replicates, we initially determine the maternal and paternal reads from the 0.5% tails of the paternal ratio distribution.
Later we control for replicate-to-replicate variability and refine this list.

```{r echo=FALSE,warning=FALSE}

TAIL_SIZE   = 0.05
quant_probs = probs = c(TAIL_SIZE, 1 - TAIL_SIZE)

fourcell_quantiles = quantile( four_cell_snp_filtered$paternal_ratio , quant_probs  )

eightcell_quantiles = quantile( eight_cell_snp_filtered$paternal_ratio, quant_probs  )

print("Quantiles are (4 and 8 cells)")
print(fourcell_quantiles)
print(eightcell_quantiles)

fourcell_paternal_genes = four_cell_snp_filtered[ four_cell_snp_filtered$paternal_ratio >= fourcell_quantiles[2] , ]$gene
fourcell_maternal_genes = four_cell_snp_filtered[ four_cell_snp_filtered$paternal_ratio <= fourcell_quantiles[1] , ]$gene

print("Four cell paternal and maternal genes are:")
print( c(length(fourcell_paternal_genes), length(fourcell_maternal_genes) )  )


eightcell_paternal_genes = eight_cell_snp_filtered[ eight_cell_snp_filtered$paternal_ratio >= eightcell_quantiles[2] , ]$gene
eightcell_maternal_genes = eight_cell_snp_filtered[ eight_cell_snp_filtered$paternal_ratio <= eightcell_quantiles[1] , ]$gene

print("Eight cell paternal and maternal genes are:")
print( c(length(eightcell_paternal_genes), length(eightcell_maternal_genes) )  )

```







## Translation Efficiency of Imprinted Genes 

```{r echo=FALSE,warning=FALSE}
ribo_rna_counts = read.csv(ribo_rna_counts_file )

ribo_rna_gene_names        = unlist(lapply( (strsplit(ribo_rna_counts$transcript, split = "-" ) ), "[[", 1  ) )

ribo_rna_counts$transcript = ribo_rna_gene_names

ribo_rna_counts
```


We apply CLR normalization to our rna and ribo data
```{r echo=TRUE,warning=FALSE}
set.seed(3)

clr_normalized_counts_pre = NormalizeData(ribo_rna_counts[,-1], 
                                          scale.factor = 10000, margin = 2, 
                                          normalization.method= "CLR",verbose = TRUE)

clr_normalized_counts = data.frame(transcript = ribo_rna_counts$transcript , clr_normalized_counts_pre)

cpm_normalized_counts_pre = cpm(ribo_rna_counts[,-1])

cpm_normalized_counts = data.frame( transcript = ribo_rna_counts$transcript, cpm_normalized_counts_pre)

colSums(clr_normalized_counts_pre)
print("\n\n##################\n\n")
colSums(cpm_normalized_counts[,-1])
```

```{r echo=FALSE,warning=FALSE}

rnaseq_fourcell_averages_clr  = (clr_normalized_counts$X4cell.1.RNAseq + clr_normalized_counts$X4cell.2.RNAseq) / 2
rnaseq_eightcell_averages_clr = (clr_normalized_counts$X8cell.1.RNAseq + clr_normalized_counts$X8cell.2.RNAseq + clr_normalized_counts$X8cell.3.RNAseq + clr_normalized_counts$X8cell.4.RNAseq) / 4
ribo_fourcell_averages_clr    = (clr_normalized_counts$X4cell.1 + clr_normalized_counts$X4cell.2 + clr_normalized_counts$X4cell.3) / 3
ribo_eightcell_averages_clr   = (clr_normalized_counts$X8cell.1 + clr_normalized_counts$X8cell.2 + clr_normalized_counts$X8cell.3 + clr_normalized_counts$X8cell.4) / 4

fourcell_TE_clr  = ribo_fourcell_averages_clr / rnaseq_fourcell_averages_clr
eightcell_TE_clr = ribo_eightcell_averages_clr / rnaseq_eightcell_averages_clr

rnaseq_onecell_averages_clr = (clr_normalized_counts$X1cell.1.RNAseq + clr_normalized_counts$X1cell.2.RNAseq + clr_normalized_counts$X1cell.3.RNAseq + clr_normalized_counts$X1cell.4.RNAseq) / 4
ribo_onecell_averages_clr   = (clr_normalized_counts$X1cell.1 + clr_normalized_counts$X1cell.2 + clr_normalized_counts$X1cell.3 + clr_normalized_counts$X1cell.4 + clr_normalized_counts$X1cell.5) / 5

onecell_TE_clr = ribo_onecell_averages_clr / rnaseq_onecell_averages_clr

rnaseq_twocell_averages_clr = (clr_normalized_counts$X2cell.1.RNAseq + clr_normalized_counts$X2cell.2.RNAseq + clr_normalized_counts$X2cell.3.RNAseq + clr_normalized_counts$X2cell.4.RNAseq) / 4
ribo_twocell_averages_clr   = (clr_normalized_counts$X2cell.1 + clr_normalized_counts$X2cell.2 + clr_normalized_counts$X2cell.3) / 3

twocell_TE_clr = rnaseq_twocell_averages_clr / ribo_twocell_averages_clr

rnaseq_mii_averages_clr = (clr_normalized_counts$MII.1.RNAseq + clr_normalized_counts$MII.2.RNAseq + clr_normalized_counts$MII.3.RNAseq + clr_normalized_counts$MII.4.RNAseq ) / 4
ribo_mii_averages_clr = (clr_normalized_counts$MII.1 + clr_normalized_counts$MII.2 + clr_normalized_counts$MII.3 + clr_normalized_counts$MII.4 + clr_normalized_counts$MII.5) / 5

mii_TE_clr = ribo_mii_averages_clr / rnaseq_mii_averages_clr


rnaseq_gv_averages_clr = (clr_normalized_counts$GV.1.RNAseq + clr_normalized_counts$GV.2.RNAseq + clr_normalized_counts$GV.3.RNAseq + clr_normalized_counts$GV.4.RNAseq) / 4
ribo_gv_averages_clr   = (clr_normalized_counts$GV.1 + clr_normalized_counts$GV.2 + clr_normalized_counts$GV.3 + clr_normalized_counts$GV.4 + clr_normalized_counts$GV.5) / 5

gv_TE_clr = ribo_gv_averages_clr / rnaseq_gv_averages_clr

TE_df_clr = data.frame( transcript   = ribo_rna_counts$transcript, 
                        fourcell_TE  = fourcell_TE_clr, 
                        eightcell_TE = eightcell_TE_clr, 
                        mii_TE       = mii_TE_clr,
                        gv_TE    = gv_TE_clr,
                        onecell_TE   = onecell_TE_clr, 
                        twocell_TE   = twocell_TE_clr
                        )

hist( log2(fourcell_TE_clr) )

hist( log2(eightcell_TE_clr) )

```
```{r echo=FALSE,warning=FALSE}

rnaseq_fourcell_averages_cpm  = (cpm_normalized_counts$X4cell.1.RNAseq + cpm_normalized_counts$X4cell.2.RNAseq) / 2
rnaseq_eightcell_averages_cpm = (cpm_normalized_counts$X8cell.1.RNAseq + cpm_normalized_counts$X8cell.2.RNAseq + cpm_normalized_counts$X8cell.3.RNAseq + cpm_normalized_counts$X8cell.4.RNAseq) / 4
ribo_fourcell_averages_cpm    = (cpm_normalized_counts$X4cell.1 + cpm_normalized_counts$X4cell.2 + cpm_normalized_counts$X4cell.3) / 3
ribo_eightcell_averages_cpm   = (cpm_normalized_counts$X8cell.1 + cpm_normalized_counts$X8cell.2 + cpm_normalized_counts$X8cell.3 + cpm_normalized_counts$X8cell.4) / 4

fourcell_TE_cpm  = ribo_fourcell_averages_cpm / rnaseq_fourcell_averages_cpm
eightcell_TE_cpm = ribo_eightcell_averages_cpm / rnaseq_eightcell_averages_cpm

rnaseq_onecell_averages_cpm = (cpm_normalized_counts$X1cell.1.RNAseq + cpm_normalized_counts$X1cell.2.RNAseq + cpm_normalized_counts$X1cell.3.RNAseq + cpm_normalized_counts$X1cell.4.RNAseq) / 4
ribo_onecell_averages_cpm   = (cpm_normalized_counts$X1cell.1 + cpm_normalized_counts$X1cell.2 + cpm_normalized_counts$X1cell.3 + cpm_normalized_counts$X1cell.4 + cpm_normalized_counts$X1cell.5) / 5

onecell_TE_cpm = ribo_onecell_averages_cpm / rnaseq_onecell_averages_cpm

rnaseq_twocell_averages_cpm = (cpm_normalized_counts$X2cell.1.RNAseq + cpm_normalized_counts$X2cell.2.RNAseq + cpm_normalized_counts$X2cell.3.RNAseq + cpm_normalized_counts$X2cell.4.RNAseq) / 4
ribo_twocell_averages_cpm   = (cpm_normalized_counts$X2cell.1 + cpm_normalized_counts$X2cell.2 + cpm_normalized_counts$X2cell.3) / 3

twocell_TE_cpm = rnaseq_twocell_averages_cpm / ribo_twocell_averages_cpm

rnaseq_mii_averages_cpm = (cpm_normalized_counts$MII.1.RNAseq + cpm_normalized_counts$MII.2.RNAseq + cpm_normalized_counts$MII.3.RNAseq + cpm_normalized_counts$MII.4.RNAseq ) / 4
ribo_mii_averages_cpm = (cpm_normalized_counts$MII.1 + cpm_normalized_counts$MII.2 + cpm_normalized_counts$MII.3 + cpm_normalized_counts$MII.4 + cpm_normalized_counts$MII.5) / 5

mii_TE_cpm = ribo_mii_averages_cpm / rnaseq_mii_averages_cpm


rnaseq_gv_averages_cpm = (cpm_normalized_counts$GV.1.RNAseq + cpm_normalized_counts$GV.2.RNAseq + cpm_normalized_counts$GV.3.RNAseq + cpm_normalized_counts$GV.4.RNAseq) / 4
ribo_gv_averages_cpm   = (cpm_normalized_counts$GV.1 + cpm_normalized_counts$GV.2 + cpm_normalized_counts$GV.3 + cpm_normalized_counts$GV.4 + cpm_normalized_counts$GV.5) / 5

gv_TE_cpm = ribo_gv_averages_cpm / rnaseq_gv_averages_cpm

TE_df_cpm = data.frame( transcript   = ribo_rna_counts$transcript, 
                        fourcell_TE  = fourcell_TE_cpm, 
                        eightcell_TE = eightcell_TE_cpm, 
                        mii_TE       = mii_TE_cpm,
                        gv_TE    = gv_TE_cpm,
                        onecell_TE   = onecell_TE_cpm, 
                        twocell_TE   = twocell_TE_cpm
                        )

hist( log2(fourcell_TE_cpm) )

hist( log2(eightcell_TE_cpm) )

```




## Imprinted genes after Bootsrapping 

We obtain the snp counts for individual experiments first. Next, we aggregate the counts by position and obtain counts for each experiment per gene.

```{r echo=FALSE,warning=FALSE}

experiments_fourcell  = c("20210607-RNAseq-4cell-cross-A", "20210607-RNAseq-4cell-cross-C")
experiments_eightcell = c("20210607-RNAseq-8cell-cross-A", "20210607-RNAseq-8cell-cross-B", "20210607-RNAseq-8cell-cross-C", "20210607-RNAseq-8cell-cross-D") 

detailed_snps = read.csv(detailed_snp_file, header = T)

detailed_snps = detailed_snps[, 1:5]

detailed_snps = detailed_snps[detailed_snps$experiment %in% c(experiments_fourcell, experiments_eightcell), ]

detailed_snps = 
  detailed_snps %>% group_by( experiment, transcript  ) %>%
    summarise(paternal_total = sum(paternal), maternal_total = sum(maternal))

detailed_snps = detailed_snps[ order(detailed_snps$transcript)  ,]

colnames(detailed_snps) = c("experiment", "gene", "paternal", "maternal")

detailed_snps$gene = unlist(lapply( strsplit(detailed_snps$gene, split = "-"), "[[", 1 ) )

detailed_snps[1000:1020,]

```


```{r echo=FALSE,warning=FALSE}

detailed_fourcell_snps = detailed_snps[ detailed_snps$experiment %in% experiments_fourcell, ]
detailed_fourcell_snps = detailed_fourcell_snps[ detailed_fourcell_snps$gene %in% four_cell_snp_filtered$gene,  ]

detailed_eightcell_snps = detailed_snps[ detailed_snps$experiment %in% experiments_eightcell,  ]
detailed_eightcell_snps = detailed_eightcell_snps[ detailed_eightcell_snps$gene %in% eight_cell_snp_filtered$gene,  ]



compute_paternal_ratio = function(unit_df, paternal_threshold){
  paternal_ratios        = (unit_df$paternal + CONSTANT_ADDITION) / (unit_df$paternal + unit_df$maternal + (2 * CONSTANT_ADDITION) )
  unit_df$paternal_ratio = paternal_ratios
  unit_df$is_paternal     = as.integer( paternal_ratios > paternal_threshold )
  unit_df$is_maternal     = as.integer( paternal_ratios < (1 - paternal_threshold) )
  
  return(unit_df)
}

unit_sample = function( snp_df , paternal_threshold ){
  experiments         = unique(snp_df$experiment)
  sampled_experiments = sample( experiments, replace = T, size = length( experiments  ) )
  
  initial_df = snp_df[ snp_df$experiment == sampled_experiments[1], -1  ]
  
  #print(sampled_experiments)
  
  for (e in sampled_experiments[-1]){
    this_df             = snp_df[ snp_df$experiment == e,   ]
    initial_df$paternal = initial_df$paternal + this_df$paternal
    initial_df$maternal = initial_df$maternal + this_df$maternal
  }
  
  result = compute_paternal_ratio(initial_df, paternal_threshold)
  
  return(result)
}

bulk_sample = function( snp_df, paternal_threshold, rounds = 100  ){
  experiments         = unique(snp_df$experiment)
  
  genes = snp_df[ snp_df$experiment == experiments[1],   ]$gene
  
  paternal_hits = rep(0, length(genes))
  maternal_hits = rep(0, length(genes))
  
  for(i in 1:rounds){
    this_sampling_df = unit_sample(snp_df, paternal_threshold = 0.7)
    paternal_hits = paternal_hits + this_sampling_df$is_paternal
    maternal_hits = maternal_hits + this_sampling_df$is_maternal
  }
  
  result = data.frame( gene = genes, paternal = paternal_hits, maternal = maternal_hits  )
  
  
  return(result)
  
}

#a = unit_sample(detailed_eightcell_snps)
#compute_paternal_ratio(a)


#unit_sample(detailed_eightcell_snps, paternal_threshold = 0.7)
```


```{r echo=FALSE,warning=FALSE}

set.seed(3)

eightcell_bootstrap_df = bulk_sample(detailed_eightcell_snps, paternal_threshold = 0.7, rounds = 1000)
fourcell_bootstrap_df  = bulk_sample(detailed_fourcell_snps, paternal_threshold  = 0.7, rounds = 1000) 

```



```{r echo=FALSE,warning=FALSE}

hist( eightcell_bootstrap_df$maternal, breaks = seq(0,1000, 50)   )

hist( fourcell_bootstrap_df$maternal   )
```


```{r echo=FALSE,warning=FALSE}

BOOTSTRAP_THRESHOLD = 800

eightcell_paternal_genes_bootstrap   = eightcell_bootstrap_df[eightcell_bootstrap_df$paternal > BOOTSTRAP_THRESHOLD, ]$gene 
eightcell_maternal_genes_bootstrap   = eightcell_bootstrap_df[eightcell_bootstrap_df$maternal > BOOTSTRAP_THRESHOLD, ]$gene 
eightcell_background_genes_bootstrap = setdiff( eightcell_bootstrap_df$gene , c(eightcell_paternal_genes_bootstrap, eightcell_maternal_genes_bootstrap) )

fourcell_paternal_genes_bootstrap   = fourcell_bootstrap_df[fourcell_bootstrap_df$paternal > BOOTSTRAP_THRESHOLD, ]$gene
fourcell_maternal_genes_bootstrap   = fourcell_bootstrap_df[fourcell_bootstrap_df$maternal > BOOTSTRAP_THRESHOLD, ]$gene
fourcell_background_genes_bootstrap = setdiff( fourcell_bootstrap_df$gene , c(fourcell_paternal_genes_bootstrap, fourcell_maternal_genes_bootstrap) )


print( c(  "Paternal Genes from 4cells and 8 cells are:", length(fourcell_paternal_genes_bootstrap), length(eightcell_paternal_genes_bootstrap) ) )
print( c(  "Maternal Genes from 4cells and 8 cells are:", length(fourcell_maternal_genes_bootstrap), length(eightcell_maternal_genes_bootstrap) ) )

```


```{r echo=FALSE,warning=FALSE}


mymain = function(TE_df, rna_averages, gene_lengths, picker_name , paternal_genes, maternal_genes, background_genes){
  paternal_TE = TE_df[ TE_df$transcript %in% paternal_genes,   ][picker_name][,1]
  maternal_TE = TE_df[ TE_df$transcript %in% maternal_genes,   ][picker_name][,1]
  bg_TE       = TE_df[ TE_df$transcript %in% background_genes, ][picker_name][,1]
  
  paternal_rna   = rna_averages[ TE_df$transcript %in% paternal_genes ]
  maternal_rna   = rna_averages[ TE_df$transcript %in% maternal_genes ]
  background_rna = rna_averages[ TE_df$transcript %in% background_genes ]
  
  arranged_TE =  c( bg_TE, paternal_TE, maternal_TE )
  arrenged_rna = c( background_rna, paternal_rna, maternal_rna)
  
  TE_labels = c(
      rep("background", length(bg_TE)) ,
      rep("paternal", length(paternal_TE) ),
      rep("maternal",  length(maternal_TE) )
  )
  
  TE_imprinted = c(
    rep(0, length(bg_TE)) ,
    rep(1, length(paternal_TE) ),
    rep(1,  length(maternal_TE) )
  )
  
  paternal_genes_TE   = TE_df[TE_df$transcript %in% paternal_genes,]$transcript
  maternal_genes_TE   = TE_df[TE_df$transcript %in% maternal_genes,]$transcript
  background_genes_TE = TE_df[TE_df$transcript %in% background_genes,]$transcript
  
  TE_arranged_genes = c(background_genes_TE, paternal_genes_TE, maternal_genes_TE)
  
  TE_arranged_df_pre = data.frame(  type       = TE_labels,
                                imprinted  = TE_imprinted,
                                gene       = TE_arranged_genes,
                                TE         = arranged_TE,
                                rna        = arrenged_rna )
  
  TE_arranged_df = merge(TE_arranged_df_pre, gene_lengths, by.x = "gene", by.y= "gene")
  
  return(TE_arranged_df)
  
}


```



```{r echo=FALSE,warning=FALSE}

set.seed(3)

fourcell_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_fourcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "fourcell_TE"  , 
       paternal_genes   = fourcell_paternal_genes_bootstrap, 
       maternal_genes   = fourcell_maternal_genes_bootstrap, 
       background_genes = fourcell_background_genes_bootstrap)

fourcell_TE



sample_fourcell = matchit(imprinted~rna + cds_length, data = fourcell_TE )

sample_fourcell


matched_fourcell = match.data(sample_fourcell)

matched_fourcell

boxplot( log2(fourcell_TE$rna) ~ fourcell_TE$imprinted )

boxplot( log2(matched_fourcell$rna) ~ matched_fourcell$imprinted )

boxplot( log2(matched_fourcell$rna) ~ matched_fourcell$type )

boxplot( fourcell_TE$TE ~ fourcell_TE$type )

boxplot( matched_fourcell$TE ~ matched_fourcell$type )


plot(density( log2( matched_fourcell[matched_fourcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_fourcell[matched_fourcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_fourcell[matched_fourcell$type == "background",]$cds_length  )   ), col = "red" )

```

```{r echo=FALSE,warning=FALSE}
wilcox.test( matched_fourcell[matched_fourcell$type == "background", ]$TE, matched_fourcell[matched_fourcell$type == "maternal", ]$TE  )

wilcox.test( matched_fourcell[matched_fourcell$type == "background", ]$TE, matched_fourcell[matched_fourcell$type == "paternal", ]$TE  )

wilcox.test( matched_fourcell[matched_fourcell$type == "background", ]$TE, c(matched_fourcell[matched_fourcell$type == "paternal", ]$TE, matched_fourcell[matched_fourcell$type == "maternal", ]$TE)  )

wilcox.test( matched_fourcell[matched_fourcell$type == "maternal", ]$TE, matched_fourcell[matched_fourcell$type == "paternal", ]$TE  )
```

```{r echo=FALSE,warning=FALSE}
wilcox.test( fourcell_TE[fourcell_TE$type == "background", ]$TE, fourcell_TE[fourcell_TE$type == "maternal", ]$TE  )

wilcox.test( fourcell_TE[fourcell_TE$type == "background", ]$TE, fourcell_TE[fourcell_TE$type == "paternal", ]$TE  )

wilcox.test( fourcell_TE[fourcell_TE$type == "background", ]$TE, c(fourcell_TE[fourcell_TE$type == "paternal", ]$TE, fourcell_TE[fourcell_TE$type == "maternal", ]$TE)  )

wilcox.test( fourcell_TE[fourcell_TE$type == "maternal", ]$TE, fourcell_TE[fourcell_TE$type == "paternal", ]$TE  )
```



```{r echo=FALSE,warning=FALSE}

four_cell_snp_filtered$total_rna_snp = four_cell_snp_filtered$maternal_y + four_cell_snp_filtered$paternal_y


fourcell_imprented_totals = four_cell_snp_filtered[four_cell_snp_filtered$gene %in% fourcell_TE[fourcell_TE$imprinted == 1, ]$gene, ]$total_rna_snp
fourcell_background_totals = four_cell_snp_filtered[four_cell_snp_filtered$gene %in% fourcell_TE[fourcell_TE$imprinted == 0, ]$gene, ]$total_rna_snp

plot(  density(  log2(fourcell_imprented_totals) ) )
lines(  density( log2(fourcell_background_totals) ), col="red" )

hist( log2(fourcell_imprented_totals)) 
lines( hist( log2(fourcell_imprented_totals), plot = F) ) 

four_cell_snp_filtered$paternal_ratio = four_cell_snp_filtered$paternal_y / four_cell_snp_filtered$total_rna_snp
four_cell_snp_filtered[four_cell_snp_filtered$paternal_ratio == 0,]
four_cell_snp_filtered[four_cell_snp_filtered$paternal_ratio == 1,]

```

```{r echo=FALSE,warning=FALSE}

set.seed(3)

fourcell_TE_cpm = 
  mymain(
       TE_df            = TE_df_cpm, 
       rna_averages     = rnaseq_fourcell_averages_cpm, 
       gene_lengths     = region_lengths, 
       picker_name      = "fourcell_TE"  , 
       paternal_genes   = fourcell_paternal_genes_bootstrap, 
       maternal_genes   = fourcell_maternal_genes_bootstrap, 
       background_genes = fourcell_background_genes_bootstrap)

fourcell_TE_cpm



sample_fourcell_cpm = matchit(imprinted~rna + cds_length, data = fourcell_TE_cpm )

sample_fourcell_cpm


matched_fourcell_cpm = match.data(sample_fourcell_cpm)

matched_fourcell_cpm

boxplot( log2(fourcell_TE_cpm$rna) ~ fourcell_TE_cpm$imprinted )

boxplot( log2(matched_fourcell_cpm$rna) ~ matched_fourcell_cpm$imprinted )

boxplot( log2(matched_fourcell_cpm$rna) ~ matched_fourcell_cpm$type )

boxplot( fourcell_TE_cpm$TE ~ fourcell_TE_cpm$type )

boxplot( matched_fourcell_cpm$TE ~ matched_fourcell_cpm$type )


plot(density( log2( matched_fourcell_cpm[matched_fourcell_cpm$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_fourcell_cpm[matched_fourcell_cpm$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_fourcell_cpm[matched_fourcell_cpm$type == "background",]$cds_length  )   ), col = "red" )

```

```{r echo=FALSE,warning=FALSE}
wilcox.test( fourcell_TE_cpm[fourcell_TE_cpm$type == "background", ]$TE, fourcell_TE_cpm[fourcell_TE_cpm$type == "maternal", ]$TE  )

wilcox.test( fourcell_TE_cpm[fourcell_TE_cpm$type == "background", ]$TE, fourcell_TE_cpm[fourcell_TE_cpm$type == "paternal", ]$TE  )

wilcox.test( fourcell_TE_cpm[fourcell_TE_cpm$type == "background", ]$TE, c(fourcell_TE_cpm[fourcell_TE_cpm$type == "paternal", ]$TE, fourcell_TE_cpm[fourcell_TE_cpm$type == "maternal", ]$TE)  )

wilcox.test( fourcell_TE_cpm[fourcell_TE_cpm$type == "maternal", ]$TE, fourcell_TE_cpm[fourcell_TE_cpm$type == "paternal", ]$TE  )
```

```{r echo=FALSE,warning=FALSE}
wilcox.test( matched_fourcell_cpm[matched_fourcell_cpm$type == "background", ]$TE, matched_fourcell_cpm[matched_fourcell_cpm$type == "maternal", ]$TE  )

wilcox.test( matched_fourcell_cpm[matched_fourcell_cpm$type == "background", ]$TE, matched_fourcell_cpm[matched_fourcell_cpm$type == "paternal", ]$TE  )

wilcox.test( matched_fourcell_cpm[matched_fourcell_cpm$type == "background", ]$TE, c(matched_fourcell_cpm[matched_fourcell_cpm$type == "paternal", ]$TE, matched_fourcell_cpm[matched_fourcell_cpm$type == "maternal", ]$TE)  )

wilcox.test( matched_fourcell_cpm[matched_fourcell_cpm$type == "maternal", ]$TE, matched_fourcell_cpm[matched_fourcell_cpm$type == "paternal", ]$TE  )

wilcox.test( matched_fourcell_cpm[matched_fourcell_cpm$imprinted == 1, ]$TE, matched_fourcell_cpm[matched_fourcell_cpm$imprinted == 0, ]$TE  )
```

```{r echo=FALSE,warning=FALSE}

eightcell_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_eightcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "eightcell_TE"  , 
       paternal_genes   = eightcell_paternal_genes_bootstrap, 
       maternal_genes   = eightcell_maternal_genes_bootstrap, 
       background_genes = eightcell_background_genes_bootstrap)

eightcell_TE



sample_eightcell = matchit(imprinted~rna + cds_length, data = eightcell_TE )

sample_eightcell


matched_eightcell = match.data(sample_eightcell)

matched_eightcell

boxplot( log2(matched_eightcell$rna) ~ matched_eightcell$imprinted )

boxplot( log2(matched_eightcell$rna) ~ matched_eightcell$type )

boxplot( matched_eightcell$TE ~ matched_eightcell$type )


plot(density( log2( matched_eightcell[matched_eightcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_eightcell[matched_eightcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_eightcell[matched_eightcell$type == "background",]$cds_length  )   ), col = "red" )

```



```{r echo=FALSE,warning=FALSE}
wilcox.test( matched_eightcell[matched_eightcell$type == "background", ]$TE, matched_eightcell[matched_eightcell$type == "maternal", ]$TE  )

wilcox.test( matched_eightcell[matched_eightcell$type == "background", ]$TE, matched_eightcell[matched_eightcell$type == "paternal", ]$TE  )

wilcox.test( matched_eightcell[matched_eightcell$type == "background", ]$TE, c(matched_eightcell[matched_eightcell$type == "paternal", ]$TE, matched_eightcell[matched_eightcell$type == "maternal", ]$TE )  )

wilcox.test( matched_eightcell[matched_eightcell$type == "maternal", ]$TE, matched_eightcell[matched_eightcell$type == "paternal", ]$TE  )
```

```{r echo=FALSE,warning=FALSE}
wilcox.test( eightcell_TE[eightcell_TE$type == "background", ]$TE, eightcell_TE[eightcell_TE$type == "maternal", ]$TE  )

wilcox.test( eightcell_TE[eightcell_TE$type == "background", ]$TE, eightcell_TE[eightcell_TE$type == "paternal", ]$TE  )

wilcox.test( eightcell_TE[eightcell_TE$type == "background", ]$TE, c(eightcell_TE[eightcell_TE$type == "paternal", ]$TE, eightcell_TE[eightcell_TE$type == "maternal", ]$TE )  )

wilcox.test( matched_eightcell[matched_eightcell$type == "maternal", ]$TE, matched_eightcell[matched_eightcell$type == "paternal", ]$TE  )
```

```{r echo=FALSE,warning=FALSE}

set.seed(3)

eightcell_TE_cpm = 
  mymain(
       TE_df            = TE_df_cpm, 
       rna_averages     = rnaseq_eightcell_averages_cpm, 
       gene_lengths     = region_lengths, 
       picker_name      = "eightcell_TE"  , 
       paternal_genes   = eightcell_paternal_genes_bootstrap, 
       maternal_genes   = eightcell_maternal_genes_bootstrap, 
       background_genes = eightcell_background_genes_bootstrap)

eightcell_TE_cpm



sample_eightcell_cpm = matchit(imprinted~rna + cds_length, data = eightcell_TE_cpm )

sample_eightcell_cpm


matched_eightcell_cpm = match.data(sample_eightcell_cpm)

matched_eightcell_cpm

boxplot( log2(eightcell_TE_cpm$rna) ~ eightcell_TE_cpm$imprinted )

boxplot( log2(matched_eightcell_cpm$rna) ~ matched_eightcell_cpm$imprinted )

boxplot( log2(matched_eightcell_cpm$rna) ~ matched_eightcell_cpm$type )

boxplot( eightcell_TE_cpm$TE ~ eightcell_TE_cpm$type )

boxplot( matched_eightcell_cpm$TE ~ matched_eightcell_cpm$type )


plot(density( log2( matched_eightcell_cpm[matched_eightcell_cpm$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_eightcell_cpm[matched_eightcell_cpm$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_eightcell_cpm[matched_eightcell_cpm$type == "background",]$cds_length  )   ), col = "red" )

```

```{r echo=FALSE,warning=FALSE}
wilcox.test( eightcell_TE_cpm[eightcell_TE_cpm$type == "background", ]$TE, eightcell_TE_cpm[eightcell_TE_cpm$type == "maternal", ]$TE  )

wilcox.test( eightcell_TE_cpm[eightcell_TE_cpm$type == "background", ]$TE, eightcell_TE_cpm[eightcell_TE_cpm$type == "paternal", ]$TE  )

wilcox.test( eightcell_TE_cpm[eightcell_TE_cpm$type == "background", ]$TE, c(eightcell_TE_cpm[eightcell_TE_cpm$type == "paternal", ]$TE, eightcell_TE_cpm[eightcell_TE_cpm$type == "maternal", ]$TE)  )

wilcox.test( eightcell_TE_cpm[eightcell_TE_cpm$type == "maternal", ]$TE, eightcell_TE_cpm[eightcell_TE_cpm$type == "paternal", ]$TE  )
```

```{r echo=FALSE,warning=FALSE}
wilcox.test( matched_fourcell_cpm[matched_fourcell_cpm$type == "background", ]$TE, matched_fourcell_cpm[matched_fourcell_cpm$type == "maternal", ]$TE  )

wilcox.test( matched_fourcell_cpm[matched_fourcell_cpm$type == "background", ]$TE, matched_fourcell_cpm[matched_fourcell_cpm$type == "paternal", ]$TE  )

wilcox.test( matched_fourcell_cpm[matched_fourcell_cpm$type == "background", ]$TE, c(matched_fourcell_cpm[matched_fourcell_cpm$type == "paternal", ]$TE, matched_fourcell_cpm[matched_fourcell_cpm$type == "maternal", ]$TE)  )

wilcox.test( matched_fourcell_cpm[matched_fourcell_cpm$type == "maternal", ]$TE, matched_fourcell_cpm[matched_fourcell_cpm$type == "paternal", ]$TE  )
```


## Mono Allelic Genes

Among our imprinted genes, 
If a gene has (zero paternal counts and more than one paternal counts) or
(has more than one paternal counts and zero maternal counts), it is called a mono alelic gene.
Here we investigate the translation efficiency of mono allelic genes.




```{r echo=FALSE,warning=FALSE}

mono_allelic_eightcell_paternal_df = eight_cell_snp_filtered[ eight_cell_snp_filtered$maternal_y == 0 & eight_cell_snp_filtered$paternal_y > 0 & eight_cell_snp_filtered$gene %in% eightcell_paternal_genes_bootstrap,  ]

mono_allelic_eightcell_paternal_df

mono_allelic_eightcell_paternal_genes = eight_cell_snp_filtered[ eight_cell_snp_filtered$maternal_y == 0 & eight_cell_snp_filtered$paternal_y > 0 & eight_cell_snp_filtered$gene %in% eightcell_paternal_genes_bootstrap,  ]$gene


mono_allelic_eightcell_maternal_genes = eight_cell_snp_filtered[ eight_cell_snp_filtered$paternal_y == 0 & eight_cell_snp_filtered$maternal_y > 0 & eight_cell_snp_filtered$gene %in% eightcell_maternal_genes_bootstrap,  ]$gene

print( c("Eight cell Monoallelic paternal and maternal genes", length(mono_allelic_eightcell_paternal_genes), length(mono_allelic_eightcell_maternal_genes))  )


mono_allelic_fourcell_paternal_genes = four_cell_snp_filtered[ four_cell_snp_filtered$maternal_y == 0 & four_cell_snp_filtered$paternal_y > 0 & four_cell_snp_filtered$gene %in% fourcell_paternal_genes_bootstrap,  ]$gene


mono_allelic_fourcell_maternal_genes = four_cell_snp_filtered[ four_cell_snp_filtered$paternal_y == 0 & four_cell_snp_filtered$maternal_y > 0 & four_cell_snp_filtered$gene %in% fourcell_maternal_genes_bootstrap,  ]$gene

print( c("four cell Monoallelic paternal and maternal genes", length(mono_allelic_fourcell_paternal_genes), length(mono_allelic_fourcell_maternal_genes))  )

mono_allelic_eightcell_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_eightcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "eightcell_TE"  , 
       paternal_genes   = mono_allelic_eightcell_paternal_genes, 
       maternal_genes   = mono_allelic_eightcell_maternal_genes, 
       background_genes = eightcell_background_genes_bootstrap)

mono_allelic_eightcell_TE


mono_allelic_fourcell_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_fourcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "fourcell_TE"  , 
       paternal_genes   = mono_allelic_fourcell_paternal_genes, 
       maternal_genes   = mono_allelic_fourcell_maternal_genes, 
       background_genes = fourcell_background_genes_bootstrap)

mono_allelic_fourcell_TE

```


```{r echo=FALSE,warning=FALSE}

length(mono_allelic_eightcell_paternal_genes)
length(mono_allelic_fourcell_maternal_genes)
```

```{r echo=FALSE,warning=FALSE}

length(eightcell_paternal_genes_bootstrap)
length(fourcell_paternal_genes_bootstrap)
length(intersect( eightcell_paternal_genes_bootstrap, fourcell_paternal_genes_bootstrap  ))

length(eightcell_maternal_genes_bootstrap)
length(fourcell_maternal_genes_bootstrap)
length(intersect( eightcell_maternal_genes_bootstrap, fourcell_maternal_genes_bootstrap  ))

```

### Eight Cell Mono Allelic Genes

```{r echo=FALSE,warning=FALSE}

sample_mono_allelic_eightcell = matchit(imprinted~rna + cds_length, data = mono_allelic_eightcell_TE )

sample_mono_allelic_eightcell


matched_mono_allelic_eightcell = match.data(sample_mono_allelic_eightcell)

matched_mono_allelic_eightcell

boxplot( log2(matched_mono_allelic_eightcell$rna) ~ matched_mono_allelic_eightcell$imprinted )

boxplot( log2(matched_mono_allelic_eightcell$rna) ~ matched_mono_allelic_eightcell$type )

boxplot( mono_allelic_eightcell_TE$TE ~ mono_allelic_eightcell_TE$type )

boxplot( matched_mono_allelic_eightcell$TE ~ matched_mono_allelic_eightcell$type )


plot(density( log2( matched_mono_allelic_eightcell[matched_mono_allelic_eightcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_mono_allelic_eightcell[matched_mono_allelic_eightcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_mono_allelic_eightcell[matched_mono_allelic_eightcell$type == "background",]$cds_length  )   ), col = "red" )



```




```{r echo=FALSE,warning=FALSE}

wilcox.test( matched_mono_allelic_eightcell[matched_mono_allelic_eightcell$imprinted == 0, ]$TE , matched_mono_allelic_eightcell[matched_mono_allelic_eightcell$imprinted == 1, ]$TE  )

```

### Four Cell Mono Allelic Genes

```{r echo=FALSE,warning=FALSE}

sample_mono_allelic_fourcell = matchit(imprinted~rna + cds_length, data = mono_allelic_fourcell_TE )

sample_mono_allelic_fourcell


matched_mono_allelic_fourcell = match.data(sample_mono_allelic_fourcell)

matched_mono_allelic_fourcell

boxplot( log2(matched_mono_allelic_fourcell$rna) ~ matched_mono_allelic_fourcell$imprinted )

boxplot( log2(matched_mono_allelic_fourcell$rna) ~ matched_mono_allelic_fourcell$type )

boxplot( mono_allelic_fourcell_TE$TE ~ mono_allelic_fourcell_TE$type, varwidth = T )

boxplot( matched_mono_allelic_fourcell$TE ~ matched_mono_allelic_fourcell$type )


plot(density( log2( matched_mono_allelic_fourcell[matched_mono_allelic_fourcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_mono_allelic_fourcell[matched_mono_allelic_fourcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_mono_allelic_fourcell[matched_mono_allelic_fourcell$type == "background",]$cds_length  )   ), col = "red" )



```


```{r echo=FALSE,warning=FALSE}

wilcox.test( matched_mono_allelic_fourcell[matched_mono_allelic_fourcell$imprinted == 0, ]$TE , matched_mono_allelic_fourcell[matched_mono_allelic_fourcell$imprinted == 1, ]$TE  )

```


### Multiple SNP Concordence

Now we check that multiple SNPS 

```{r echo=FALSE,warning=FALSE}

raw_snps = read.csv(detailed_snp_file)

raw_snps_four_cell = raw_snps[raw_snps$experiment %in% experiments_fourcell, ]
raw_snps_eight_cell = raw_snps[raw_snps$experiment %in% experiments_eightcell, ]

compute_support_ratio = function(this_df){
  result = 
    this_df %>% group_by(experiment, transcript) %>%
    mutate(support_ratio_of_replicate = sum(support) / snp_positions)
  
  result = 
    result %>% group_by(transcript, position) %>%
    mutate(support_ratio_combined = sum(support_ratio_of_replicate))
  
  return(result)
}

find_well_supported_genes = function(this_df, threshold){
  result = this_df[this_df$support_ratio_combined > threshold,]$transcript 
  return( unique(result) )
} 

```

```{r echo=FALSE,warning=FALSE}

snps_pos_four_cell_pre = raw_snps_four_cell[, 1:5]

snps_pos_four_cell_pre$transcript = unlist( lapply(strsplit(snps_pos_four_cell_pre$transcript, split = "-" ) , "[[", 1))

snps_pos_four_cell_pre = snps_pos_four_cell_pre[ snps_pos_four_cell_pre$paternal + snps_pos_four_cell_pre$maternal > 0 , ]

snps_pos_four_cell_pre = 
  snps_pos_four_cell_pre %>% 
  group_by(experiment, transcript) %>%
  mutate(snp_positions = length(position))


snps_pos_four_cell_pre = snps_pos_four_cell_pre[snps_pos_four_cell_pre$snp_positions > 1, ]

snps_pos_four_cell_pre$paternal_ratio = (snps_pos_four_cell_pre$paternal + 1) / (snps_pos_four_cell_pre$paternal + snps_pos_four_cell_pre$maternal + 2)

snps_pos_four_cell_paternal = snps_pos_four_cell_pre[ snps_pos_four_cell_pre$transcript %in% fourcell_paternal_genes_bootstrap,  ]
snps_pos_four_cell_maternal = snps_pos_four_cell_pre[ snps_pos_four_cell_pre$transcript %in% fourcell_maternal_genes_bootstrap,  ]

snps_pos_four_cell_paternal$support = as.integer( snps_pos_four_cell_paternal$paternal_ratio > 0.5 )
snps_pos_four_cell_maternal$support = as.integer( snps_pos_four_cell_maternal$paternal_ratio < 0.5 )



snps_pos_four_cell_paternal = compute_support_ratio(snps_pos_four_cell_paternal)
snps_pos_four_cell_maternal = compute_support_ratio(snps_pos_four_cell_maternal)

snps_pos_four_cell_paternal
snps_pos_four_cell_maternal


# There needs to be 0.6 support from each replicate hence the thereshold 1.2
multiple_snp_supported_fourcell_paternal_genes = find_well_supported_genes(snps_pos_four_cell_paternal, 1.2)

multiple_snp_supported_fourcell_maternal_genes = find_well_supported_genes(snps_pos_four_cell_maternal, 1.2)
#snps_pos_four_cell_paternal$paternal_ratio = (snps_pos_four_cell_paternal$paternal + 1) / (snps_pos_four_cell_maternal$paternal + snps_pos_four_cell_paternal$maternal + 2)

multiple_snp_supported_fourcell_paternal_genes
multiple_snp_supported_fourcell_maternal_genes
```


```{r echo=FALSE,warning=FALSE}
length(unique(snps_pos_four_cell_maternal$transcript) )
length(  multiple_snp_supported_fourcell_maternal_genes )
```

```{r echo=FALSE,warning=FALSE}
snps_pos_eight_cell_pre = raw_snps_eight_cell[, 1:5]

snps_pos_eight_cell_pre$transcript = unlist( lapply(strsplit(snps_pos_eight_cell_pre$transcript, split = "-" ) , "[[", 1))

snps_pos_eight_cell_pre = snps_pos_eight_cell_pre[ snps_pos_eight_cell_pre$paternal + snps_pos_eight_cell_pre$maternal > 0 , ]

snps_pos_eight_cell_pre = 
  snps_pos_eight_cell_pre %>% 
  group_by(experiment, transcript) %>%
  mutate(snp_positions = length(position))

snps_pos_eight_cell_pre = snps_pos_eight_cell_pre[snps_pos_eight_cell_pre$snp_positions > 1, ]

snps_pos_eight_cell_pre$paternal_ratio = (snps_pos_eight_cell_pre$paternal + 1) / (snps_pos_eight_cell_pre$paternal + snps_pos_eight_cell_pre$maternal + 2)

snps_pos_eight_cell_paternal = snps_pos_eight_cell_pre[ snps_pos_eight_cell_pre$transcript %in% eightcell_paternal_genes_bootstrap,  ]
snps_pos_eight_cell_maternal = snps_pos_eight_cell_pre[ snps_pos_eight_cell_pre$transcript %in% eightcell_maternal_genes_bootstrap,  ]

snps_pos_eight_cell_paternal$support = as.integer( snps_pos_eight_cell_paternal$paternal_ratio > 0.5 )
snps_pos_eight_cell_maternal$support = as.integer( snps_pos_eight_cell_maternal$paternal_ratio < 0.5 )



snps_pos_eight_cell_paternal = compute_support_ratio(snps_pos_eight_cell_paternal)
snps_pos_eight_cell_maternal = compute_support_ratio(snps_pos_eight_cell_maternal)

snps_pos_eight_cell_paternal
snps_pos_eight_cell_maternal


# There needs to be 0.6 support from each replicate hence the thereshold 2.4
multiple_snp_supported_eightcell_paternal_genes = find_well_supported_genes(snps_pos_eight_cell_paternal, 2.4)

multiple_snp_supported_eightcell_maternal_genes = find_well_supported_genes(snps_pos_eight_cell_maternal, 2.4)

multiple_snp_supported_eightcell_paternal_genes
multiple_snp_supported_eightcell_maternal_genes
```


```{r echo=FALSE,warning=FALSE}

set.seed(3)

fourcell_multiple_snp_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_fourcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "fourcell_TE"  , 
       paternal_genes   = multiple_snp_supported_fourcell_paternal_genes, 
       maternal_genes   = multiple_snp_supported_fourcell_maternal_genes, 
       background_genes = fourcell_background_genes_bootstrap)

fourcell_multiple_snp_TE



sample_multiple_snp_fourcell = matchit(imprinted~rna + cds_length, data = fourcell_multiple_snp_TE )

sample_multiple_snp_fourcell


matched_multiple_snp_fourcell = match.data(sample_multiple_snp_fourcell)

matched_multiple_snp_fourcell

boxplot( log2(fourcell_multiple_snp_TE$rna) ~ fourcell_multiple_snp_TE$imprinted )

boxplot( log2(matched_multiple_snp_fourcell$rna) ~ matched_multiple_snp_fourcell$imprinted )

boxplot( log2(matched_multiple_snp_fourcell$rna) ~ matched_multiple_snp_fourcell$type )

boxplot( fourcell_multiple_snp_TE$TE ~ fourcell_multiple_snp_TE$type )

boxplot( matched_multiple_snp_fourcell$TE ~ matched_multiple_snp_fourcell$type )


plot(density( log2( matched_multiple_snp_fourcell[matched_multiple_snp_fourcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_multiple_snp_fourcell[matched_multiple_snp_fourcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_multiple_snp_fourcell[matched_multiple_snp_fourcell$type == "background",]$cds_length  )   ), col = "red" )
```


```{r echo=FALSE,warning=FALSE}
wilcox.test(  fourcell_multiple_snp_TE[fourcell_multiple_snp_TE$imprinted == 0, ]$TE, fourcell_multiple_snp_TE[fourcell_multiple_snp_TE$imprinted == 1, ]$TE )

wilcox.test(  matched_multiple_snp_fourcell[matched_multiple_snp_fourcell$imprinted == 0, ]$TE, matched_multiple_snp_fourcell[matched_multiple_snp_fourcell$imprinted == 1, ]$TE )

wilcox.test(  fourcell_multiple_snp_TE[fourcell_TE$type == "background", ]$TE, fourcell_TE[fourcell_TE$type == "paternal", ]$TE )

wilcox.test(  fourcell_multiple_snp_TE[fourcell_multiple_snp_TE$type == "background", ]$TE, fourcell_multiple_snp_TE[fourcell_multiple_snp_TE$type == "maternal", ]$TE )

wilcox.test(  fourcell_multiple_snp_TE[fourcell_multiple_snp_TE$type == "maternal", ]$TE, fourcell_multiple_snp_TE[fourcell_multiple_snp_TE$type == "paternal", ]$TE )
```


```{r echo=FALSE,warning=FALSE}

set.seed(3)

fourcell_mono_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_fourcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "fourcell_TE"  , 
       paternal_genes   = mono_allelic_fourcell_paternal_genes, 
       maternal_genes   = mono_allelic_fourcell_maternal_genes, 
       background_genes = fourcell_background_genes_bootstrap)


sample_mono_fourcell = matchit(imprinted~rna + cds_length, data = fourcell_mono_TE )

matched_mono_fourcell = match.data(sample_mono_fourcell)


eightcell_mono_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_eightcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "eightcell_TE"  , 
       paternal_genes   = mono_allelic_eightcell_paternal_genes, 
       maternal_genes   = mono_allelic_eightcell_maternal_genes, 
       background_genes = eightcell_background_genes_bootstrap)


sample_mono_eightcell = matchit(imprinted~rna + cds_length, data = eightcell_mono_TE )

matched_mono_eightcell = match.data(sample_mono_eightcell)




```





```{r echo=FALSE,warning=FALSE}

set.seed(3)

eightcell_multiple_snp_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_eightcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "eightcell_TE"  , 
       paternal_genes   = multiple_snp_supported_eightcell_paternal_genes, 
       maternal_genes   = multiple_snp_supported_eightcell_maternal_genes, 
       background_genes = eightcell_background_genes_bootstrap)

eightcell_multiple_snp_TE



sample_multiple_snp_eightcell = matchit(imprinted~rna + cds_length, data = eightcell_multiple_snp_TE )

sample_multiple_snp_eightcell


matched_multiple_snp_eightcell = match.data(sample_multiple_snp_eightcell)

matched_multiple_snp_eightcell

boxplot( log2(eightcell_multiple_snp_TE$rna) ~ eightcell_multiple_snp_TE$imprinted )

boxplot( log2(matched_multiple_snp_eightcell$rna) ~ matched_multiple_snp_eightcell$imprinted )

boxplot( log2(matched_multiple_snp_eightcell$rna) ~ matched_multiple_snp_eightcell$type )

boxplot( eightcell_multiple_snp_TE$TE ~ eightcell_multiple_snp_TE$type )

boxplot( matched_multiple_snp_eightcell$TE ~ matched_multiple_snp_eightcell$type )


plot(density( log2( matched_multiple_snp_eightcell[matched_multiple_snp_eightcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.8) )

lines(density( log2( matched_multiple_snp_eightcell[matched_multiple_snp_eightcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_multiple_snp_eightcell[matched_multiple_snp_eightcell$type == "background",]$cds_length  )   ), col = "red" )
```

```{r echo=FALSE,warning=FALSE}
wilcox.test(  eightcell_multiple_snp_TE[eightcell_multiple_snp_TE$imprinted == 0, ]$TE, eightcell_multiple_snp_TE[eightcell_multiple_snp_TE$imprinted == 1, ]$TE )

wilcox.test(  matched_multiple_snp_eightcell[matched_multiple_snp_eightcell$imprinted == 0, ]$TE, matched_multiple_snp_eightcell[matched_multiple_snp_eightcell$imprinted == 1, ]$TE )
```


## Comparison with GSE80810

In [Borensztein M, ..., Heard. E., Xist-dependent imprinted X inactivation and the early developmental consequences of its failure] (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5337400/), they crossed 
females from the strain C57BL/6J  with males from the strain Cast/EiJ. They called this crossing **BC**.  We compared our 4cell and 8cell data to their wild type BC crossings.
They reported the paternal ratios for individual replicates. So we averaged the ratios to obtain a single comparison for each cell number (4 cell and 8cell). 

We made our comparison directional-wise. Namely, to compare our paternal genes, we counted the  genes with ratio > 0.5 and for maternal genes, we counted the genes with ratio < 0.5 in Borensztein et. al.


```{r echo=FALSE,warning=FALSE}
GSE80810_allelic_ratios = read.csv(GSE80810_allelic_ratio_file)

a = colnames(GSE80810_allelic_ratios)
GSE80810_allelic_ratios = GSE80810_allelic_ratios[, c(1,16:20,26:44)]

GSE80810_8C_paternal_intersection = GSE80810_allelic_ratios[GSE80810_allelic_ratios$Genes %in% eightcell_paternal_genes_bootstrap,-c(1,2,3,4,5,6)]
GSE80810_8C_paternal_averages = apply(GSE80810_8C_paternal_intersection , MARGIN = 1, FUN = mean, na.rm = 1) 
GSE80810_8C_paternal_agreements = sum(as.integer(GSE80810_8C_paternal_averages > 0.5 ) , na.rm = T)

print(c("8 cell paternal", GSE80810_8C_paternal_agreements, "agreements out of ", dim(GSE80810_8C_paternal_intersection)[1]))



GSE80810_8C_maternal_intersection = GSE80810_allelic_ratios[GSE80810_allelic_ratios$Genes %in% eightcell_maternal_genes_bootstrap,-c(1,2,3,4,5,6)]
GSE80810_8C_maternal_averages = apply(GSE80810_8C_maternal_intersection , MARGIN = 1, FUN = mean, na.rm = 1)
GSE80810_8C_maternal_agreements = sum(as.integer(GSE80810_8C_maternal_averages < 0.5 ) , na.rm = T)

print(c("8 cell maternal", GSE80810_8C_maternal_agreements, "agreements out of ", dim(GSE80810_8C_maternal_intersection)[1]))



GSE80810_4C_paternal_intersection = GSE80810_allelic_ratios[GSE80810_allelic_ratios$Genes %in% fourcell_paternal_genes_bootstrap,-c(1,2,3,4,5,6)]
GSE80810_4C_paternal_averages = apply(GSE80810_4C_paternal_intersection , MARGIN = 1, FUN = mean, na.rm = 1) 
GSE80810_4C_paternal_agreements = sum(as.integer(GSE80810_4C_paternal_averages > 0.5 ) , na.rm = T)

print(c("4 cell paternal", GSE80810_4C_paternal_agreements, "agreements out of ", dim(GSE80810_4C_paternal_intersection)[1]))



GSE80810_4C_maternal_intersection = GSE80810_allelic_ratios[GSE80810_allelic_ratios$Genes %in% fourcell_maternal_genes_bootstrap,-c(1,2,3,4,5,6)]
GSE80810_4C_maternal_averages = apply(GSE80810_4C_maternal_intersection , MARGIN = 1, FUN = mean, na.rm = 1) 
GSE80810_4C_maternal_agreements = sum(as.integer(GSE80810_4C_maternal_averages < 0.5 ) , na.rm = T)

print(c("4 cell maternal", GSE80810_4C_maternal_agreements, "agreements out of ", dim(GSE80810_4C_maternal_intersection)[1]))

```

## Scatter Plot of Paternal Ratios

```{r echo=FALSE, warning=FALSE}

snp_ratios_merged = merge(  four_cell_snp_filtered, eight_cell_snp_filtered,
                           by.x = "gene", by.y = "gene", all = T  )

snp_ratios_merged = snp_ratios_merged[, c(1,6,12)]

colnames(snp_ratios_merged) = c("gene", "fourcell_ratio", "eightcell_ratio")

decide_snp_class_pre = function(gene, 
                                fourcell_paternal_genes, fourcell_maternal_genes,
                                eightcell_maternal_genes, eightcell_paternal_genes){

  significant_fourcell_genes  = c(fourcell_paternal_genes, fourcell_maternal_genes)
  significant_eightcell_genes = c(eightcell_maternal_genes, eightcell_paternal_genes)
  
  int_4_8 = intersect(significant_fourcell_genes, significant_eightcell_genes)
  
  if(gene %in% int_4_8  ){
    return("4_and_8")
  } 
  else if(gene %in% significant_fourcell_genes){
    return("4")
  }
  else if(gene %in% significant_eightcell_genes){
    return("8")
  }
  else {
    return("None")
  }
}

decide_snp_class = Vectorize(decide_snp_class_pre, vectorize.args = "gene")


plot_snp_scatter = function(this_title , snp_ratios, 
                            fourcell_paternal_genes, fourcell_maternal_genes,
                            eightcell_maternal_genes, eightcell_paternal_genes  ){
  
  tick_offset = 0.06

  tick_positions = c(0.7, 0.3, 1, 0)
  tick_labels    = c("0.7", "0.3", "1.0", "0.0")
  
  
  
  snp_ratios$gene_type = decide_snp_class(snp_ratios$gene,
                                          fourcell_paternal_genes = fourcell_paternal_genes,
                                          fourcell_maternal_genes = fourcell_maternal_genes,
                                          eightcell_maternal_genes = eightcell_maternal_genes,
                                          eightcell_paternal_genes = eightcell_paternal_genes)
  
  
  
  snp_ratios$gene_type = 
    factor(snp_ratios$gene_type, 
         labels = c("4 & 8 cell","4 cell", "8 cell", "non-specific"),
         levels = c("4_and_8", "4", "8", "None"))
  
  #snp_ratios = snp_ratios[(! is.na(snp_ratios$fourcell_ratio) & !is.na(snp_ratios$eightcell_ratio) ),]
  

  
  ratio_scatter_plot = 
    ggplot( snp_ratios, aes(x = fourcell_ratio, y = eightcell_ratio, color = gene_type) ) +

      geom_point( alpha = 0.4, size = 0.5) + theme_bw() + 
            theme(
            panel.border      = element_blank(),
            panel.grid        = element_blank(),
            plot.title        = PLOT_TITLE,
            panel.background  = element_blank(),
            axis.text       = element_blank(),
            axis.title.y      = AXIS_TITLE,
            axis.title.x      = AXIS_TITLE,
            axis.ticks        = element_blank(),
            legend.background = element_blank(),
            legend.title      = element_blank()
 #           legend.position   = "bottom"
          ) +
      geom_hline(yintercept=0.5, size = 0.35) +
      geom_vline(xintercept=0.5, size = 0.35) + 
      scale_color_manual(values = c(UTR5_BLUE, CDS_GREEN, UTR3_GREEN ,"gray") ) +
      labs(title = this_title, x = "4 cell", y = "8 cell", color = "Type")
  
  
 
  
  ratio_scatter_plot = 
    ratio_scatter_plot + 
    annotate("text", 
              label = c("0.5",tick_labels), 
              x = rep( 0.5 - tick_offset , 5  ), 
              y = c(0.5 - tick_offset, tick_positions), size = 3) +
    annotate("text", 
              label =tick_labels, 
              y = rep( 0.5 - tick_offset , 4  ), 
              x = tick_positions, size = 3)
  
 
    
    ratio_scatter_plot = 
      ratio_scatter_plot + 
        geom_segment(data = data.frame(w = c(tick_positions, tick_positions)), 
                     aes(x     = c(rep(0.5 - tick_offset*0.2, length(tick_positions)), tick_positions) , 
                         xend = c( rep( 0.5 + tick_offset*0.2, length(tick_positions) ), tick_positions),
                         y = c(tick_positions, rep(0.5 - tick_offset*0.2, length(tick_positions))), 
                         yend = c(tick_positions, rep(0.5 + tick_offset*0.2, length(tick_positions))) ), 
                     color = "black", size = 0.25 )
    
    return(ratio_scatter_plot)
}






```


```{r echo=FALSE,warning=FALSE}
scatter_plot_of_bootstrapped = 
  plot_snp_scatter("Paternal Ratios" , 
                   snp_ratios = snp_ratios_merged, 
                   fourcell_paternal_genes = fourcell_paternal_genes_bootstrap, 
                   fourcell_maternal_genes = fourcell_maternal_genes_bootstrap,
                   eightcell_maternal_genes = eightcell_maternal_genes_bootstrap, 
                   eightcell_paternal_genes = eightcell_paternal_genes_bootstrap  )
  
scatter_plot_of_bootstrapped
```

## Common genes

We determine the number of common (paternal and maternal) genes between 4 cell and 8 cell stages.

```{r echo=TRUE, message=FALSE, warning=FALSE, paged.print=FALSE}


generate_venn_diagram = function( fourcell_set, eightcell_set, title  ){

  result = 
  venn.diagram(  x = list(fourcell_set, eightcell_set ),
                 category.names = c("4 cell", "8 Cell"),
                 main = title, 
                 filename = NULL, 
                 disable.logging = T,
                 output   = F,
                 inverted = T,
                 
                 # Numbers
                 fontfamily = "sans",
                 cex = 0.6,
                 
                 # Circles
                 lwd = 1,
                 lty = 'blank',
                 fill = c(CDS_GREEN, UTR3_GREEN),
                         
                 imagetype="svg" ,
                 height = 300 , 
                 width = 480 , 
                 resolution = PDF_resolution,
                 
                 cat.cex = 0.6,
                 cat.fontfamily = "sans",
                 cat.pos = c(30,330),
                 cat.dist = c(0.04, 0.04 ) )

}


```


```{r echo=FALSE,warning=FALSE}
venn_paternal_genes_bootstrap = 
  generate_venn_diagram( fourcell_set   =  fourcell_paternal_genes_bootstrap, 
                         eightcell_set  = eightcell_paternal_genes,
                         title          = "Paternal Genes"  )

venn_maternal_genes_bootstrap = 
  generate_venn_diagram( fourcell_set   =  fourcell_maternal_genes_bootstrap, 
                         eightcell_set  = eightcell_maternal_genes,
                         title          = "Maternal Genes"  )

pdf("pdf/venn_bootstrap.pdf", width = 1.5, height = 1.5)
grid.draw(venn_maternal_genes_bootstrap)
grid.newpage()
grid.draw(venn_paternal_genes_bootstrap)
dev.off()

grid.draw(venn_maternal_genes_bootstrap)
grid.newpage()
grid.draw(venn_paternal_genes_bootstrap)
grid.newpage()
```


```{r echo=FALSE,warning=FALSE}

scatter_plot_of_multisnp = 
  plot_snp_scatter("Multi SNP Supported Genes" , 
                   snp_ratios = snp_ratios_merged, 
                   fourcell_paternal_genes = multiple_snp_supported_fourcell_paternal_genes, 
                   fourcell_maternal_genes = multiple_snp_supported_fourcell_maternal_genes,
                   eightcell_maternal_genes = multiple_snp_supported_eightcell_maternal_genes, 
                   eightcell_paternal_genes = multiple_snp_supported_eightcell_paternal_genes  )
  

scatter_plot_of_multisnp
```

```{r echo=FALSE,warning=FALSE}
venn_paternal_genes_multisnp = 
  generate_venn_diagram( fourcell_set   =  multiple_snp_supported_fourcell_paternal_genes, 
                         eightcell_set  = multiple_snp_supported_eightcell_paternal_genes,
                         title          = "Paternal Genes"  )

venn_maternal_genes_multisnp = 
  generate_venn_diagram( fourcell_set   =  multiple_snp_supported_fourcell_maternal_genes, 
                         eightcell_set  = multiple_snp_supported_eightcell_maternal_genes,
                         title          = "Maternal Genes"  )

pdf("pdf/venn_multisnp.pdf", width = 1.5, height = 1.5)
grid.draw(venn_maternal_genes_multisnp)
grid.newpage()
grid.draw(venn_paternal_genes_multisnp)
dev.off()

grid.draw(venn_maternal_genes_multisnp)
grid.newpage()
grid.draw(venn_paternal_genes_multisnp)
grid.newpage()
```

```{r echo=FALSE,warning=FALSE}
scatter_plot_of_monoallelic = 
  plot_snp_scatter("Monoallelic Genes" , 
                   snp_ratios = snp_ratios_merged, 
                   fourcell_paternal_genes = mono_allelic_fourcell_paternal_genes, 
                   fourcell_maternal_genes = mono_allelic_fourcell_maternal_genes,
                   eightcell_maternal_genes = mono_allelic_eightcell_maternal_genes, 
                   eightcell_paternal_genes = mono_allelic_eightcell_paternal_genes  )
  
scatter_plot_of_monoallelic
```

```{r echo=FALSE,warning=FALSE}
venn_paternal_genes_mono = 
  generate_venn_diagram( fourcell_set   =  mono_allelic_fourcell_paternal_genes, 
                         eightcell_set  = mono_allelic_eightcell_paternal_genes,
                         title          = "Paternal Genes"  )

venn_maternal_genes_mono = 
  generate_venn_diagram( fourcell_set   =  mono_allelic_fourcell_maternal_genes, 
                         eightcell_set  = mono_allelic_eightcell_maternal_genes,
                         title          = "Maternal Genes"  )

pdf("pdf/venn_mono.pdf", width = 1.5, height = 1.5)
grid.draw(venn_maternal_genes_mono)
grid.newpage()
grid.draw(venn_paternal_genes_mono)
dev.off()

grid.draw(venn_maternal_genes_mono)
grid.newpage()
grid.draw(venn_paternal_genes_mono)
grid.newpage()
```

```{r echo=FALSE,warning=FALSE}
 
```


```{r echo=FALSE,warning=FALSE}

```


```{r echo=FALSE,warning=FALSE}


save_plot_pdf("scatter_plot_of_bootstrapped_with_legend.pdf", scatter_plot_of_bootstrapped, width = 3.6, height = 2 )
save_plot_pdf("scatter_plot_of_bootstrapped_no_legend.pdf", scatter_plot_of_bootstrapped + theme(legend.position = "none"), width = 2, height = 2 )

save_plot_pdf("scatter_plot_of_multisnp_with_legend.pdf", scatter_plot_of_multisnp, width = 3.6, height = 2 )
save_plot_pdf("scatter_plot_of_multisnp_no_legend.pdf", scatter_plot_of_multisnp + theme(legend.position = "none"), width = 2, height = 2 )

save_plot_pdf("scatter_plot_of_monoallelic_with_legend.pdf", scatter_plot_of_monoallelic, width = 3.6, height = 2 )
save_plot_pdf("scatter_plot_of_monoallelic_no_legend.pdf", scatter_plot_of_monoallelic + theme(legend.position = "none"), width = 2, height = 2 )
```

## Barplots of the Manuscript

```{r echo=FALSE,warning=FALSE}

twoplot = function(TE_df, title, this_color, ymax = 1.5, type = "box", jitter = F, keep_y_axis = T, write_pval = F){
  
  TE_df$imprinted_chr = as.character(TE_df$imprinted)
  
  
  
  this_wilcox_test = wilcox.test(TE_df[TE_df$imprinted == 0, ]$TE, TE_df[TE_df$imprinted == 1, ]$TE)
  
  this_pvalue = ""
  
  if(write_pval){
    this_pvalue = sprintf("p = %1.1e", this_wilcox_test$p.value)
  }
  
   p_label =  annotate("text", 
              label = this_pvalue, 
              x = 1.7, 
              y = ymax - 0.05, size = 2)
  
  p =  ggplot( TE_df, aes(x = imprinted_chr, y = TE, fill = imprinted_chr) ) + p_label
  
  if(type == "violin"){
    p = p + geom_violin(draw_quantiles = c(0.5, 0.5), scale = "count", trim = FALSE, na.rm = TRUE)
  }
  else{
    p = p + geom_boxplot(outlier.shape = NA, varwidth = T)
    if(jitter){
      p = p + geom_jitter(size = 1, alpha = 0.3)
    }
  }
  
  this_y_axis = element_blank()
  this_y_axis_text = element_blank()
  this_y_axis_title = element_blank()
  
  if(keep_y_axis){
    this_y_axis = element_line(colour = "black", size = 0.35)
    this_y_axis_text = AXIS_TITLE
    this_y_axis_title = AXIS_TITLE
  }
  
  
  p = p +
    scale_y_continuous(limits = c(0, ymax)) + 
    theme_bw() +
    theme(plot.title   = PLOT_TITLE,
            panel.border = element_blank(),
            panel.grid   = element_blank(),
            axis.text.y       = this_y_axis_text,
            axis.text.x       =  element_text(face = "plain", size = FONT_LABEL_SIZE, angle = 45),
            axis.title.y      = this_y_axis_title,
            axis.title.x      = element_blank(),
            axis.line.y      = this_y_axis,
            legend.title      = element_blank(),
          legend.position     = "None"  ) + 
    scale_x_discrete(labels = c("Non-specific", "Allele Specific")) + 
    labs(title = title, 
           x     = "Genes", 
           y     = "Translation Efficiency") + 
    scale_fill_manual(values=c("gray", this_color))
  
  if(! keep_y_axis){
    p = p + theme(axis.ticks.y = element_blank())
  }
  

  
  return(p)
}


TWO_STAT = stat_compare_means( aes(label = paste0("p ", ..p.format..)), label.x = 1.5, label.y.npc = 0.95, size = 2.5, na.rm = F, method = "wilcox", method.args = list(tol.root = 1e-4, alternative = "two.sided" ) )

p = twoplot(fourcell_TE, title = "Four Cell", this_color = CDS_GREEN, type = "bar", jitter = F, keep_y_axis = F, write_pval = T)
p 


```

```{r echo=FALSE,warning=FALSE}
threeplot = function(TE_df, title, this_color, ymax = 1.5, type = "box", jitter = F, selector = "TE", keep_y_axis = T, write_pval = T){
  
  TE_df["value"] = TE_df[selector]
  
  ylabel = selector
  
  if(selector == "TE"){
    ylabel = "Translation Efficiency"
  }
  
  this_wilcox_test_maternal = wilcox.test(TE_df[TE_df$type == "background", ]$TE,TE_df[TE_df$type == "maternal", ]$TE)
  this_wilcox_test_paternal = wilcox.test(TE_df[TE_df$type == "background", ]$TE,TE_df[TE_df$type == "paternal", ]$TE)
  
  this_maternal_pvalue = ""
  this_paternal_pvalue = ""
  
  if(write_pval){
    this_maternal_pvalue = sprintf("p = %1.1e", this_wilcox_test_maternal$p.value)
    this_paternal_pvalue = sprintf("p = %1.1e", this_wilcox_test_paternal$p.value)
  }
  
  p_maternal_label =  annotate("text", 
              label = this_maternal_pvalue, 
              x = 2, 
              y = ymax - 0.05, size = 2)
  
  p_paternal_label =  annotate("text", 
              label = this_paternal_pvalue, 
              x = 3, 
              y = ymax - 0.05, size = 2)
  
  this_y_axis = element_blank()
  this_y_axis_text = element_blank()
  this_y_axis_title = element_blank()
  
  if(keep_y_axis){
    this_y_axis = element_line(colour = "black", size = 0.35)
    this_y_axis_text = AXIS_TITLE
    this_y_axis_title = AXIS_TITLE
  }
  
  p = ggplot( TE_df, aes(x = type, y = value, fill = type) ) + p_maternal_label + p_paternal_label
  
  if(type == "violin"){
    p = p + geom_violin(draw_quantiles = c(0.5, 0.5), scale = "count")
  }
  else{
    p = p + geom_boxplot(outlier.shape = NA, varwidth = T, )
    if(jitter){
      p = p + geom_jitter(size = 1, alpha = 0.3)
    }
  }
  p = p +
    scale_y_continuous(limits = c(0, ymax)) + 
    theme_bw() +
    theme(plot.title   = PLOT_TITLE,
            axis.text.y       = this_y_axis_text,
            axis.text.x       =  element_text(face = "plain", size = FONT_LABEL_SIZE, angle = 45),
            axis.title.y      = this_y_axis_title,
            axis.title.x      = element_blank(),
            axis.line.y       = this_y_axis,
            panel.border      = element_blank(),
            panel.grid        = element_blank(),

            legend.title      = element_blank(),
          legend.position     = "None"  ) + 
    scale_x_discrete(labels = c("Non-specific", "Maternal", "Paternal")) + 
    labs(title = title, 
           x     = "", 
           y     = ylabel) + 
    scale_fill_manual(values=c("gray", this_color, this_color))
  
  if(! keep_y_axis){
    p = p + theme(axis.ticks.y = element_blank())
  }
  
  return(p)
}

#threeplot(fourcell_TE, title = "4cell", this_color = c(CDS_GREEN), ymax = 1.5, type = "box", jitter = F, selector = "TE")
```


```{r echo=FALSE,warning=FALSE}

compare_cds_lengths = function(TE_df, title = "CDS Lengths", this_color = CDS_GREEN){
  result = ggplot( TE_df, aes(x = cds_length ,  color = type) ) +
    geom_density( aes(linetype = type)) + 
     scale_linetype_manual(values=c("solid","twodash", "dotted") ) + 
    scale_color_manual(values = c("gray", this_color, this_color)) + 
    theme_bw() +
    theme(plot.title   = PLOT_TITLE,
        panel.border = element_blank(),
        panel.grid   = element_blank(),
        axis.text.y       = AXIS_TITLE,
        axis.text.x       = AXIS_TITLE,
        axis.title.y      = AXIS_TITLE,
        axis.title.x      = AXIS_TITLE,
        axis.line.y      = element_line(colour = "black", size = 0.35),
        axis.line.x      = element_line(colour = "black", size = 0.35),
        legend.title      = element_blank() ) + 
    labs(title = title, 
       x     = "CDS Length", 
       y     = "Density") +
    scale_x_continuous(trans = "log2")

  result = result + theme(legend.position = c(0.8,0.7))
  
  return(result)
}

```



## Fourcell Bootstrapped

```{r echo=FALSE,warning=FALSE}

fourcell_main_boxplot = twoplot(fourcell_TE, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F)
fourcell_main_boxplot + TWO_STAT


p = twoplot(fourcell_TE, title = "Four Cell", this_color = CDS_GREEN, type = "violin", jitter = F)
p + TWO_STAT

wilcox.test( fourcell_TE[fourcell_TE$imprinted == 1,]$TE, fourcell_TE[fourcell_TE$imprinted == 0,]$TE )
#p = twoplot(fourcell_TE, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = T)
#p + TWO_STAT


p = threeplot(fourcell_TE, title = "Four Cell TE", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(fourcell_TE, title = "Four Cell RNA", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(fourcell_TE, this_color = CDS_GREEN)
```

## Fourcell Bootstrapped CPM

```{r echo=FALSE,warning=FALSE}

fourcell_main_boxplot = twoplot(fourcell_TE_cpm, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F)
fourcell_main_boxplot + TWO_STAT


p = twoplot(fourcell_TE_cpm, title = "Four Cell", this_color = CDS_GREEN, type = "violin", jitter = F)
p + TWO_STAT

wilcox.test( fourcell_TE_cpm[fourcell_TE_cpm$imprinted == 1,]$TE, fourcell_TE_cpm[fourcell_TE_cpm$imprinted == 0,]$TE )
#p = twoplot(fourcell_TE, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = T)
#p + TWO_STAT


p = threeplot(fourcell_TE_cpm, title = "Four Cell TE", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(fourcell_TE_cpm, title = "Four Cell RNA", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(fourcell_TE_cpm, this_color = CDS_GREEN)
```

## Fourcell Matched

```{r echo=FALSE,warning=FALSE}

fourcell_bootstrapped_matched_boxplot = twoplot(matched_fourcell, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F)
fourcell_bootstrapped_matched_boxplot + TWO_STAT


twoplot(matched_fourcell_cpm, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F) + TWO_STAT

wilcox.test(matched_fourcell_cpm[matched_fourcell_cpm$imprinted == 1,]$TE, matched_fourcell_cpm[matched_fourcell_cpm$imprinted == 0,]$TE)

#fourcell_bootstrapped_matched_boxplot + TWO_STAT

p = twoplot(matched_fourcell, title = "Four Cell", this_color = CDS_GREEN, type = "violin", jitter = F)
p + TWO_STAT

wilcox.test( matched_fourcell[matched_fourcell$imprinted == 1,]$TE, matched_fourcell[matched_fourcell$imprinted == 0,]$TE )
#p = twoplot(fourcell_TE, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = T)
#p + TWO_STAT


p = threeplot(matched_fourcell, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(matched_fourcell, title = "Four Cell RNA", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(matched_fourcell, this_color = CDS_GREEN)
```
## Fourcell Multi-SNP Supported

```{r echo=FALSE,warning=FALSE}

fourcell_multisnp_boxplot = twoplot(fourcell_multiple_snp_TE, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F)
fourcell_multisnp_boxplot + TWO_STAT


p = twoplot(fourcell_multiple_snp_TE, title = "Four Cell", this_color = CDS_GREEN, type = "violin", jitter = F)
p + TWO_STAT

p = threeplot(fourcell_multiple_snp_TE, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(fourcell_multiple_snp_TE, title = "Four Cell RNA", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(fourcell_multiple_snp_TE, this_color = CDS_GREEN)
```

## Fourcell Mono-Allelic

```{r echo=FALSE,warning=FALSE}

fourcell_mono_boxplot = twoplot(fourcell_mono_TE, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F)
fourcell_mono_boxplot + TWO_STAT


p = twoplot(fourcell_mono_TE, title = "Four Cell", this_color = CDS_GREEN, type = "violin", jitter = F)
p + TWO_STAT

p = threeplot(fourcell_mono_TE, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(fourcell_mono_TE, title = "Four Cell RNA", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(fourcell_mono_TE, this_color = CDS_GREEN)
```
## Fourcell Matched Mono-Allelic

```{r echo=FALSE,warning=FALSE}

fourcell_matched_mono_boxplot = twoplot(matched_mono_fourcell, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F)
fourcell_matched_mono_boxplot + TWO_STAT


p = twoplot(matched_mono_fourcell, title = "Four Cell", this_color = CDS_GREEN, type = "violin", jitter = F)
p + TWO_STAT

p = threeplot(matched_mono_fourcell, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(matched_mono_fourcell, title = "Four Cell RNA", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(matched_mono_fourcell, this_color = CDS_GREEN)
```

## One Cell TE From Four Cell

```{r echo=FALSE,warning=FALSE}

onecell_TE_from_4 = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_onecell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "onecell_TE"  , 
       paternal_genes   = fourcell_paternal_genes_bootstrap, 
       maternal_genes   = fourcell_maternal_genes_bootstrap, 
       background_genes = fourcell_background_genes_bootstrap)

sample_onecell_TE_from_4 = matchit(imprinted~rna + cds_length, data = onecell_TE_from_4 )


matched_onecell_TE_from_4 = match.data(sample_onecell_TE_from_4)



onecell_from_fourcell_boxplot = twoplot(matched_onecell_TE_from_4, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F)
onecell_from_fourcell_boxplot + TWO_STAT


p = twoplot(matched_onecell_TE_from_4, title = "Four Cell", this_color = CDS_GREEN, type = "violin", jitter = F)
p + TWO_STAT

p = threeplot(matched_onecell_TE_from_4, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(matched_onecell_TE_from_4, title = "Four Cell RNA", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(matched_onecell_TE_from_4, this_color = CDS_GREEN)

p = threeplot(onecell_TE_from_4, title = "Four Cell (Unmatched)", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = twoplot(onecell_TE_from_4, title = "Four Cell (Unmatched)", this_color = CDS_GREEN, type = "violin", jitter = F)
p + TWO_STAT
```



## MII TE From Four Cell

```{r echo=FALSE,warning=FALSE}

mii_TE_from_4 = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_onecell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "mii_TE"  , 
       paternal_genes   = fourcell_paternal_genes_bootstrap, 
       maternal_genes   = fourcell_maternal_genes_bootstrap, 
       background_genes = fourcell_background_genes_bootstrap)

sample_mii_TE_from_4 = matchit(imprinted~rna + cds_length, data = mii_TE_from_4 )


matched_mii_TE_from_4 = match.data(sample_mii_TE_from_4)



mii_from_fourcell_boxplot = twoplot(matched_mii_TE_from_4, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F)
mii_from_fourcell_boxplot + TWO_STAT


mii_TE_from_8 = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_onecell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "mii_TE"  , 
       paternal_genes   = eightcell_paternal_genes_bootstrap, 
       maternal_genes   = eightcell_maternal_genes_bootstrap, 
       background_genes = eightcell_background_genes_bootstrap)

sample_mii_TE_from_8 = matchit(imprinted~rna + cds_length, data = mii_TE_from_8 )


matched_mii_TE_from_8 = match.data(sample_mii_TE_from_8)

mii_from_eightcell_boxplot = twoplot(matched_mii_TE_from_8, title = "eight Cell", this_color = CDS_GREEN, type = "box", jitter = F)
mii_from_eightcell_boxplot + TWO_STAT

p = twoplot(matched_mii_TE_from_4, title = "Four Cell", this_color = CDS_GREEN, type = "violin", jitter = F)
p + TWO_STAT

p = threeplot(matched_mii_TE_from_4, title = "Four Cell", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(matched_mii_TE_from_4, title = "Four Cell RNA", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(matched_mii_TE_from_4, this_color = CDS_GREEN)

p = threeplot(mii_TE_from_4, title = "Four Cell (Unmatched)", this_color = CDS_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = twoplot(mii_TE_from_4, title = "Four Cell (Unmatched)", this_color = CDS_GREEN, type = "violin", jitter = F)
p + TWO_STAT
```




































## eightcell Bootstrapped

```{r echo=FALSE,warning=FALSE}

eightcell_main_boxplot = twoplot(eightcell_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F)
eightcell_main_boxplot + TWO_STAT


p = twoplot(eightcell_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "violin", jitter = F)
p + TWO_STAT

wilcox.test( eightcell_TE[eightcell_TE$imprinted == 1,]$TE, eightcell_TE[eightcell_TE$imprinted == 0,]$TE )
#p = twoplot(eightcell_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = T)
#p + TWO_STAT


p = threeplot(eightcell_TE, title = "eight Cell TE", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(eightcell_TE, title = "eight Cell RNA", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(eightcell_TE, this_color = UTR3_GREEN)
```
## eightcell Bootstrapped CPM

```{r echo=FALSE,warning=FALSE}

eightcell_main_boxplot = twoplot(eightcell_TE_cpm, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F)
eightcell_main_boxplot + TWO_STAT


p = twoplot(eightcell_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "violin", jitter = F)
p + TWO_STAT

#wilcox.test( eightcell_TE[eightcell_TE_cpm$imprinted == 1,]$TE, eightcell_TE[eightcell_TE$imprinted == 0,]$TE )
#p = twoplot(eightcell_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = T)
#p + TWO_STAT


p = threeplot(eightcell_TE_cpm, title = "eight Cell TE", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(eightcell_TE_cpm, title = "eight Cell RNA", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(eightcell_TE_cpm, this_color = UTR3_GREEN)
```


## eightcell Matched

```{r echo=FALSE,warning=FALSE}

eightcell_bootstrapped_matched_boxplot = twoplot(matched_eightcell, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F)
eightcell_bootstrapped_matched_boxplot + TWO_STAT


p = twoplot(matched_eightcell, title = "eight Cell", this_color = UTR3_GREEN, type = "violin", jitter = F)
p + TWO_STAT

wilcox.test( matched_eightcell[matched_eightcell$imprinted == 1,]$TE, matched_eightcell[matched_eightcell$imprinted == 0,]$TE )
#p = twoplot(eightcell_TE, title = "eight Cell", this_color = CDS_GREEN, type = "box", jitter = T)
#p + TWO_STAT


p = threeplot(matched_eightcell, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(matched_eightcell, title = "eight Cell RNA", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(matched_eightcell, this_color = UTR3_GREEN)
```
## eightcell Multi-SNP Supported

```{r echo=FALSE,warning=FALSE}

eightcell_multisnp_boxplot = twoplot(eightcell_multiple_snp_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F)
eightcell_multisnp_boxplot + TWO_STAT


p = twoplot(eightcell_multiple_snp_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "violin", jitter = F)
p + TWO_STAT

p = threeplot(eightcell_multiple_snp_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(eightcell_multiple_snp_TE, title = "eight Cell RNA", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(eightcell_multiple_snp_TE, this_color = UTR3_GREEN)
```

## eightcell Mono-Allelic

```{r echo=FALSE,warning=FALSE}

eightcell_mono_boxplot = twoplot(eightcell_mono_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F)
eightcell_mono_boxplot + TWO_STAT


p = twoplot(eightcell_mono_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "violin", jitter = F)
p + TWO_STAT

p = threeplot(eightcell_mono_TE, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(eightcell_mono_TE, title = "eight Cell RNA", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(eightcell_mono_TE, this_color = UTR3_GREEN)
```
## eightcell Matched Mono-Allelic

```{r echo=FALSE,warning=FALSE}

eightcell_matched_mono_boxplot = twoplot(matched_mono_eightcell, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F)
eightcell_matched_mono_boxplot + TWO_STAT


p = twoplot(matched_mono_eightcell, title = "eight Cell", this_color = UTR3_GREEN, type = "violin", jitter = F)
p + TWO_STAT

p = threeplot(matched_mono_eightcell, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(matched_mono_eightcell, title = "eight Cell RNA", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(matched_mono_eightcell, this_color = UTR3_GREEN)
```

## One Cell TE From eight Cell

```{r echo=FALSE,warning=FALSE}

onecell_TE_from_8 = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_onecell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "onecell_TE"  , 
       paternal_genes   = eightcell_paternal_genes_bootstrap, 
       maternal_genes   = eightcell_maternal_genes_bootstrap, 
       background_genes = eightcell_background_genes_bootstrap)

sample_onecell_TE_from_8 = matchit(imprinted~rna + cds_length, data = onecell_TE_from_8 )


matched_onecell_TE_from_8 = match.data(sample_onecell_TE_from_8)



onecell_from_eightcell_boxplot = twoplot(matched_onecell_TE_from_8, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F)
onecell_from_eightcell_boxplot + TWO_STAT


p = twoplot(matched_onecell_TE_from_8, title = "eight Cell", this_color = UTR3_GREEN, type = "violin", jitter = F)
p + TWO_STAT

p = threeplot(matched_onecell_TE_from_8, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(matched_onecell_TE_from_8, title = "eight Cell RNA", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(matched_onecell_TE_from_8, this_color = UTR3_GREEN)

p = threeplot(onecell_TE_from_8, title = "eight Cell (Unmatched)", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = twoplot(onecell_TE_from_8, title = "eight Cell (Unmatched)", this_color = UTR3_GREEN, type = "violin", jitter = F)
p + TWO_STAT
```



## MII TE From eight Cell

```{r echo=FALSE,warning=FALSE}

mii_TE_from_8 = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_onecell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "mii_TE"  , 
       paternal_genes   = eightcell_paternal_genes_bootstrap, 
       maternal_genes   = eightcell_maternal_genes_bootstrap, 
       background_genes = eightcell_background_genes_bootstrap)

sample_mii_TE_from_8 = matchit(imprinted~rna + cds_length, data = mii_TE_from_8 )


matched_mii_TE_from_8 = match.data(sample_mii_TE_from_8)



mii_from_eightcell_boxplot = twoplot(matched_mii_TE_from_8, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F)
mii_from_eightcell_boxplot + TWO_STAT


p = twoplot(matched_mii_TE_from_8, title = "eight Cell", this_color = UTR3_GREEN, type = "violin", jitter = F)
p + TWO_STAT

p = threeplot(matched_mii_TE_from_8, title = "eight Cell", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = threeplot(matched_mii_TE_from_8, title = "eight Cell RNA", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 6, selector = "rna" )
p = p + stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(5.25, 5.65) )
p

compare_cds_lengths(matched_mii_TE_from_8, this_color = UTR3_GREEN)

p = threeplot(mii_TE_from_8, title = "eight Cell (Unmatched)", this_color = UTR3_GREEN, type = "box", jitter = F, ymax = 1.8)
p = p +  stat_compare_means(comparisons = list(  c(1,2), c(1,3)  ), label.y = c(1.55, 1.65) )
p

p = twoplot(mii_TE_from_8, title = "eight Cell (Unmatched)", this_color = UTR3_GREEN, type = "violin", jitter = F)
p + TWO_STAT
```






















```{r echo=FALSE,warning=FALSE}

all_snps_eight_cell = raw_snps[raw_snps$experiment %in% experiments_eightcell, ]

all_snps_eight_cell$transcript = unlist( lapply( strsplit(all_snps_eight_cell$transcript, split = "-" ) , "[[", 1)  )

all_snps_four_cell = raw_snps[raw_snps$experiment %in% experiments_fourcell, ]

all_snps_four_cell$transcript = unlist( lapply( strsplit(all_snps_four_cell$transcript, split = "-" ) , "[[", 1)  )

mono_allelic_eightcell_maternal_genes
```


```{r echo=FALSE,warning=FALSE}

this_gene = "Slc24a5"

all_snps_eight_cell[all_snps_eight_cell$transcript == this_gene,]

eightcell_TE[eightcell_TE$gene == this_gene , ]

this_gene = "Tubg2"

all_snps_eight_cell[all_snps_eight_cell$transcript == this_gene,]

eightcell_TE[eightcell_TE$gene == this_gene , ]

```





```{r echo=FALSE,warning=FALSE}

mono_allelic_eightcell_paternal_genes

eightcell_TE[eightcell_TE$gene %in% mono_allelic_eightcell_paternal_genes, ]
```


```{r echo=FALSE,warning=FALSE}

this_gene = "Prkce"

all_snps_eight_cell[all_snps_eight_cell$transcript == this_gene,]

eightcell_TE[eightcell_TE$gene == this_gene , ]

this_gene = "Ptprh"

all_snps_eight_cell[all_snps_eight_cell$transcript == this_gene,]

eightcell_TE[eightcell_TE$gene == this_gene , ]


this_gene = "Shc3"

all_snps_eight_cell[all_snps_eight_cell$transcript == this_gene,]

eightcell_TE[eightcell_TE$gene == this_gene , ]
```


```{r echo=FALSE,warning=FALSE}

mono_allelic_fourcell_paternal_genes

fourcell_TE[fourcell_TE$gene %in% mono_allelic_fourcell_paternal_genes, ]
```


```{r echo=FALSE,warning=FALSE}

this_gene = "Prkcz"

all_snps_four_cell[all_snps_four_cell$transcript == this_gene,]

fourcell_TE[fourcell_TE$gene == this_gene , ]

this_gene = "Syt5"

all_snps_four_cell[all_snps_four_cell$transcript == this_gene,]

fourcell_TE[fourcell_TE$gene == this_gene , ]


this_gene = "Atg10"

all_snps_four_cell[all_snps_four_cell$transcript == this_gene,]

fourcell_TE[fourcell_TE$gene == this_gene , ]

```


```{r echo=FALSE,warning=FALSE}

mono_allelic_fourcell_maternal_genes

```


```{r echo=FALSE,warning=FALSE}
this_gene = "Tesc"

all_snps_four_cell[all_snps_four_cell$transcript == this_gene,]

fourcell_TE[fourcell_TE$gene == this_gene , ]

this_gene = "Tacc2"

all_snps_four_cell[all_snps_four_cell$transcript == this_gene,]

fourcell_TE[fourcell_TE$gene == this_gene , ]

```


```{r echo=FALSE,warning=FALSE}

a = all_snps_four_cell[all_snps_four_cell$transcript %in% fourcell_background_genes_bootstrap, ]
```


```{r echo=FALSE,warning=FALSE}

this_gene = "Actr5"

all_snps_four_cell[all_snps_four_cell$transcript == this_gene,]

fourcell_TE[fourcell_TE$gene == this_gene , ]

this_gene = "Adam17"

all_snps_four_cell[all_snps_four_cell$transcript == this_gene,]

fourcell_TE[fourcell_TE$gene == this_gene , ]

this_gene = "Anapc7"

all_snps_four_cell[all_snps_four_cell$transcript == this_gene,]

fourcell_TE[fourcell_TE$gene == this_gene , ]
```


```{r echo=FALSE,warning=FALSE}

b = all_snps_eight_cell[all_snps_eight_cell$transcript %in% eightcell_background_genes_bootstrap, ]

```


```{r echo=FALSE,warning=FALSE}

this_gene = "Plpp5"

all_snps_eight_cell[all_snps_eight_cell$transcript == this_gene,]

eightcell_TE[eightcell_TE$gene == this_gene , ]


this_gene = "Pld3"

all_snps_eight_cell[all_snps_eight_cell$transcript == this_gene,]

eightcell_TE[eightcell_TE$gene == this_gene , ]


this_gene = "Pik3c2a"

all_snps_eight_cell[all_snps_eight_cell$transcript == this_gene,]

eightcell_TE[eightcell_TE$gene == this_gene , ]



this_gene = "Pemt"

all_snps_eight_cell[all_snps_eight_cell$transcript == this_gene,]

eightcell_TE[eightcell_TE$gene == this_gene , ]

```


```{r echo=FALSE,warning=FALSE}

# Prepare the data structures for the plots
## TO BE ADDED: Confidence interval for TE
COUNT_NORMALIZATION_FACTOR = 100000

example_4cell_gene_list = c("Prkcz", "Tesc", "Anapc7" )
example_4cell_TE_df = fourcell_TE[fourcell_TE$gene %in% example_4cell_gene_list, ]
example_4cell_TE_df$stage = rep("4 cell", 3)

example_8cell_gene_list = c("Shc3","Plpp5", "Tubg2" )
example_8cell_TE_df     = eightcell_TE[eightcell_TE$gene %in% example_8cell_gene_list, ]
example_8cell_TE_df$stage = rep("8 cell", 3)

example_TE_df = rbind(example_4cell_TE_df, example_8cell_TE_df)
example_TE_df

## We added normalized counts
all_snps_four_cell = 
  all_snps_four_cell %>% 
  group_by(experiment) %>%
  mutate( experiment_total = sum(paternal + maternal) ) %>%
  mutate( paternal_per_k = (paternal / experiment_total)* COUNT_NORMALIZATION_FACTOR,
          maternal_per_k = (maternal / experiment_total)*COUNT_NORMALIZATION_FACTOR )

all_snps_eight_cell = 
  all_snps_eight_cell %>% 
  group_by(experiment) %>%
  mutate( experiment_total = sum(paternal + maternal) ) %>%
  mutate( paternal_per_k = (paternal / experiment_total)* COUNT_NORMALIZATION_FACTOR,
          maternal_per_k = (maternal / experiment_total)*COUNT_NORMALIZATION_FACTOR )

example_fourcell_snps = all_snps_four_cell[all_snps_four_cell$transcript %in% example_4cell_gene_list,]
example_fourcell_snps = example_fourcell_snps[, c(1:5,13,14)]  
  
example_eightcell_snps = all_snps_eight_cell[all_snps_eight_cell$transcript %in% example_8cell_gene_list,]
example_eightcell_snps = example_eightcell_snps[, c(1:5, 13,14)]
  example_eightcell_snps

fourcell_background_median_TE  = median(fourcell_TE[fourcell_TE$type == "background", ]$TE, na.rm = T )
eightcell_background_median_TE = median(eightcell_TE[eightcell_TE$type == "background", ]$TE, na.rm = T )

print( c(fourcell_background_median_TE, eightcell_background_median_TE ) )

## We will add two dummy experiments with 0 counts to 4cell
## this way, the bar widths will be equal

fourcell_e1_data = example_fourcell_snps[example_fourcell_snps$experiment == 	"20210607-RNAseq-4cell-cross-A",]
fourcell_e1_data$maternal = 0
fourcell_e1_data$paternal = 0
fourcell_e1_data$maternal_per_k = 0
fourcell_e1_data$paternal_per_k = 0
fourcell_e1_data$experiment = "D1"


example_fourcell_snps = rbind(example_fourcell_snps, fourcell_e1_data)
fourcell_e1_data$experiment = "D2"
example_fourcell_snps = rbind(example_fourcell_snps, fourcell_e1_data)
example_fourcell_snps
```

```{r echo=FALSE,warning=FALSE}

generate_blank_plot = function(bg = "white"){
  
  this_background = element_blank()
    
  if(bg != "white"){
    this_background = element_rect(fill = bg)
  }  
  
  df <- data.frame()
  p = ggplot(df) + geom_point() + 
    xlim(0, 4) + ylim(0, 100) + 
    theme(
      panel.border     = element_blank(),
      panel.grid       = element_blank(),
      panel.background = this_background,
      axis.title.x     = element_blank(),
      axis.title.y     = element_blank(),
      axis.text.x      = element_blank(),
      axis.text.y      = element_blank(),
      axis.ticks.x     = element_blank(),
      plot.background  = element_blank(),
      axis.ticks.y     = element_blank()
    ) 
  return(p)
}

format_y_axis = function(x){
  return( sprintf("%.0f", x) )
}

make_y_axis = function(ymax){
  if(ymax == 5){
    result = c(0, 5)
  }
  else if(ymax %% 2 == 0){
    result = c(0, ymax / 2, ymax)
  } 
  else if(ymax %% 3 == 0){
    result = c(0, ymax / 3, 2*( ymax / 3 ) , ymax)
  } 
  else{
    result = c(0, ymax)
  }
  
  return(result)
}

y_scale_expand = c(0,0)



generate_legend_unit = function(snp_df, gene){
  
  gene_data = 
    snp_df %>%
    filter( transcript == gene ) %>%
    filter(paternal_per_k > 0 | maternal_per_k > 0)
  
  number_of_snps = length( unique(gene_data$position  ) )

    barplot_colors = colorRampPalette(brewer.pal(9,"Blues"))( number_of_snps + 5 )[-seq(1,5)] 
  
  df = data.frame( x = rep(1,number_of_snps), y = rep(1,number_of_snps), position = seq(1,number_of_snps))
  
  p = ggplot(data=df, 
             aes(x    = x, 
                 y    = y, 
                 fill = factor(position, levels = sort(unique( df$position  ) )  )  )  )  + 
    geom_bar(stat="identity", width= 0.8, color = "black", size = 0.3) + 
    theme(
      panel.border     = element_blank(),
      panel.grid       = element_blank(),
      panel.background = element_blank(),
      axis.title.x     = element_blank(),
      legend.position  = "none",
      axis.title.y     = element_blank(),
      axis.text.x = element_blank(),
      axis.ticks.x=element_blank(),
      axis.ticks.y=element_blank(),
      plot.background = element_blank(),
      axis.text.y = element_blank(),
      plot.title = element_text(color = "black", size=7, face="bold", hjust = 0.5),
    ) + 
    scale_fill_manual(values = barplot_colors)  + 
    #ggtitle(title_texts[[experiment_type]])
    ggtitle( paste( as.character(number_of_snps), sep = " " ) )
  
  #if(experiment_type == "rna"){
  #  p = p + scale_y_reverse()
  #}
  
  return(p + rotate())
}


unit_snp_plot = function(snp_df, gene, allele_type = "paternal", ymax = 100){
  
  gene_data = snp_df[ snp_df$transcript == gene,]
  
  number_of_snps = length( unique(gene_data$position  ) )
  
  this_color_palette = colorRampPalette(brewer.pal(9,"Blues"))( number_of_snps + 5 )[-seq( 1,  5) ]
  
  yticks = make_y_axis(ymax)
  
  if(allele_type == "paternal"){
    plot_base = ggplot(data=gene_data, 
                       aes(x    = experiment, 
                           y    = paternal_per_k, 
                           fill = factor(position, levels = sort(unique( gene_data$position  ) )  )  )  ) 
  }else{
    plot_base = ggplot(data=gene_data, 
                       aes(x    = experiment, 
                           y    = maternal_per_k, 
                           fill = factor(position, levels = sort(unique( gene_data$position  ) )  )  )  )
  }
  
  p = plot_base + 
        geom_bar(stat="identity", width= 1) + 
    theme(
      panel.border     = element_blank(),
      panel.grid       = element_blank(),
      panel.background = element_blank(),
      axis.title.x     = element_blank(),
      legend.position  = "none",
      # plot.margin      = margin(5.5, 5.5, 0, 5.5),
      axis.title.y     = element_blank(),
      axis.text.x = element_blank(),
      axis.ticks.x=element_blank(),
      #plot.background = background_canvas,
      axis.text.y = element_text( face = "plain", size = DETAILED_BARPLOT_FONT_SIZE),
      # axis.ticks.y=element_blank()
    ) +
    scale_fill_manual(values = this_color_palette)
  
    if(allele_type == "maternal"){
      p = p + scale_y_reverse(labels = format_y_axis, 
                                              limits = c(ymax, 0), 
                                              breaks   = yticks,
                                              expand = y_scale_expand )
    }
    else{
      p = p + scale_y_continuous( labels = format_y_axis, 
                                                  limits = c(0, ymax),
                                                  breaks   = yticks,
                                                  expand = y_scale_expand  )
    }
  
  return(p)
}

paternal_label = ggdraw() + 
  draw_label(
    "paternal",
    fontface   = 'plain',
    size       = FONT_LABEL_SIZE,
    angle      = 90
  ) +
  theme(
    plot.margin = margin(0, 0, 0, 0)
  )

maternal_label = ggdraw() + 
  draw_label(
    "maternal",
    fontface   = 'plain',
    size       = FONT_LABEL_SIZE,
    angle      = 90
  ) +
  theme(
    plot.margin = margin(0, 0, 0, 0)
  )

separator_1 = generate_blank_plot()

TITLE_HEIGHT  = 0.2
LEGEND_HEIGHT = 0.6

snp_paternal_maternal_label = plot_grid(separator_1, paternal_label, maternal_label, separator_1, ncol = 1,
                                        rel_heights = c(TITLE_HEIGHT, 1, 1, LEGEND_HEIGHT))

plot_snps_of_gene = function(snp_df, gene){
  
  
    this_title = ggdraw() + 
    draw_label(
      gene,
      fontface   = 'plain',
      size       = FONT_LABEL_SIZE,
      angle      = 0
    ) +
    theme(
      plot.margin = margin(0, 0, 0, 0)
    )
  
  gene_sums = snp_df %>%
    filter( transcript == gene ) %>%
    group_by(experiment) %>%
    mutate(paternal_sum = sum(paternal_per_k), maternal_sum = sum(maternal_per_k))
  
  ymax = max( max(gene_sums$paternal_sum), max( gene_sums$maternal_sum ) )
  
  paternal_plot = unit_snp_plot(snp_df, gene = gene, allele_type = "paternal", ymax = ymax)
  maternal_plot = unit_snp_plot(snp_df, gene = gene, allele_type = "maternal", ymax = ymax)
  
  this_legend = generate_legend_unit(snp_df = snp_df, gene = gene)
  this_plot = plot_grid(this_title, paternal_plot, maternal_plot, this_legend, ncol = 1, 
                        rel_heights = c(TITLE_HEIGHT, 1, 1, LEGEND_HEIGHT))
  
  return(this_plot)
}




unit_snp_plot(example_fourcell_snps, gene = "Anapc7", allele_type = "paternal")

plot_snps_of_gene(example_fourcell_snps, gene = "Anapc7") 



```
```{r echo=FALSE,warning=FALSE}



snp_plot_anapc7 = plot_snps_of_gene(example_fourcell_snps, gene = "Anapc7") 

snp_plot_prkcz = plot_snps_of_gene(example_fourcell_snps, gene = "Prkcz")

snp_plot_tesc = plot_snps_of_gene(example_fourcell_snps, gene = "Tesc")

snp_plot_4cell = plot_grid(snp_plot_anapc7,  snp_plot_tesc, snp_plot_prkcz, nrow = 1)

snp_plot_plpp5 = plot_snps_of_gene(example_eightcell_snps, gene = "Plpp5") 

snp_plot_tubg2 = plot_snps_of_gene(example_eightcell_snps, gene = "Tubg2")

snp_plot_shc3 = plot_snps_of_gene(example_eightcell_snps, gene = "Shc3")

snp_plot8_cell = plot_grid(snp_plot_plpp5, snp_plot_tubg2, snp_plot_shc3, nrow = 1)
snp_plot8_cell

snp_pos_plot = plot_grid(snp_paternal_maternal_label, snp_plot_4cell, snp_plot8_cell , nrow = 1,
                         rel_widths = c(0.15, 1, 1))

snp_pos_plot

save_plot_pdf("snp_pos_plot.pdf", snp_pos_plot, width = 4, height = 2)
#snp_plot_4cell


```



# Bootstrapping to determine confidence intervals for Tes

```{r echo=FALSE,warning=FALSE}
set.seed(3)
example_genes_count = clr_normalized_counts[clr_normalized_counts$transcript %in% c(example_4cell_gene_list, example_8cell_gene_list),]

example_genes_sample_ordered = example_genes_count$transcript

example_genes_count_4cell_rna   = example_genes_count[,c(18,19)]
example_genes_count_4cell_ribo  = example_genes_count[,c(32,33,34)]

example_genes_count_8cell_ribo = example_genes_count[, 35:38]
example_genes_count_8cell_rna  = example_genes_count[,  19:22]

TE_sample_unit = function(){
  
  experiments_4cell_rna = sample(1: length(example_genes_count_4cell_rna), 
                                 size    = length(example_genes_count_4cell_rna), 
                                 replace = T)
  
  experiments_4cell_ribo = sample(1: length(example_genes_count_4cell_ribo), 
                                 size    = length(example_genes_count_4cell_ribo), 
                                 replace = T)
  
  experiments_8cell_rna = sample(1: length(example_genes_count_8cell_rna), 
                                 size    = length(example_genes_count_8cell_rna), 
                                 replace = T)
  
  experiments_8cell_ribo = sample(1: length(example_genes_count_8cell_ribo), 
                                 size    = length(example_genes_count_8cell_ribo), 
                                 replace = T)
  
  average_4cell_rna = apply(example_genes_count_4cell_rna[,experiments_4cell_rna], MARGIN = 1, FUN = mean) 
  average_4cell_ribo = apply(example_genes_count_4cell_ribo[,experiments_4cell_ribo], MARGIN = 1, FUN = mean) 
  average_8cell_rna = apply(example_genes_count_8cell_rna[, experiments_8cell_rna], MARGIN = 1, FUN = mean) 
  average_8cell_ribo = apply(example_genes_count_8cell_ribo[, experiments_8cell_ribo], MARGIN = 1, FUN = mean) 
  
  TE_4cell_sampling_TE = average_4cell_ribo / average_4cell_rna
  TE_8cell_sampling_TE = average_8cell_ribo / average_8cell_rna
  
  return(list(fourcell = TE_4cell_sampling_TE, eightcell = TE_8cell_sampling_TE) )
}

make_sampling_TE_matrices = function(rounds = 100){
  first_round  = TE_sample_unit()
  fourcell_TE  = first_round$fourcell
  eightcell_TE = first_round$eightcell
  
  for( i in 2:rounds){
    this_round  = TE_sample_unit()
    fourcell_TE = rbind(fourcell_TE, this_round$fourcell)
    eightcell_TE = rbind(eightcell_TE, this_round$eightcell)
  }
  

  return(list(fourcell = fourcell_TE, eightcell = eightcell_TE))
} 

example_TE_from_sampling = make_sampling_TE_matrices(100)
```


```{r echo=FALSE,warning=FALSE}
example_TE_4cell_from_sampling =  example_TE_from_sampling$fourcell
example_TE_8cell_from_sampling =  example_TE_from_sampling$eightcell

colnames(example_TE_4cell_from_sampling) = example_genes_sample_ordered
colnames(example_TE_8cell_from_sampling) = example_genes_sample_ordered

example_TE_4cell_from_sampling = example_TE_4cell_from_sampling[,example_4cell_gene_list]
#example_TE_4cell_from_sampling

example_TE_8cell_from_sampling = example_TE_8cell_from_sampling[,example_8cell_gene_list]
#example_TE_8cell_from_sampling

fourcell_TE_sampling_means  = apply(example_TE_4cell_from_sampling, MARGIN = 2, FUN = mean)
eightcell_TE_sampling_means = apply(example_TE_8cell_from_sampling, MARGIN = 2, FUN = mean)

fourcell_TE_sampling_confidencelower = apply(example_TE_4cell_from_sampling, MARGIN = 2, FUN = quantile, probs = 0.05 )
eightcell_TE_sampling_confidencelower = apply(example_TE_8cell_from_sampling, MARGIN = 2, FUN = quantile, probs = 0.05 )

fourcell_TE_sampling_confidenceupper = apply(example_TE_4cell_from_sampling, MARGIN = 2, FUN = quantile, probs = 0.95 )
eightcell_TE_sampling_confidenceupper = apply(example_TE_8cell_from_sampling, MARGIN = 2, FUN = quantile, probs = 0.95 )

stages = c(rep("4cell", 3), rep("8cell", 3))

example_TE_with_conf_intervals = data.frame( gene = c(names(fourcell_TE_sampling_means), names(eightcell_TE_sampling_means)),
                                            TE    = c(fourcell_TE_sampling_means, eightcell_TE_sampling_means),
                                            stage = stages,
                                            confidence_lower = c(fourcell_TE_sampling_confidencelower, eightcell_TE_sampling_confidencelower),
                                            confidence_upper = c(fourcell_TE_sampling_confidenceupper, eightcell_TE_sampling_confidenceupper)
                                           )

example_TE_with_conf_intervals = merge(example_TE_with_conf_intervals, example_TE_df[,c(1,2)] )

position_converter = function(x){
  if(x == "4cell"){
    return(1)
  }
  else{
    return(2)
  }
}

position_converter = Vectorize(position_converter)

example_TE_with_conf_intervals$position = position_converter(example_TE_with_conf_intervals$stage)

example_TE_with_conf_intervals
```

```{r echo=FALSE,warning=FALSE}

thick_width = 0.3

plot_example_TE = function(TE_df, fourcell_median, eightcell_median){
  
  p_colors = c("gray", CDS_GREEN, CDS_GREEN, gray, UTR3_GREEN, UTR3_GREEN) 
  
  label_y_positions = TE_df$TE + c(0, 0, 0, 0, 0, 0)
  label_x_positions = TE_df$position + c(-0.15, -0.15, 0.5, 0, 0, 0.5)
  text_df = data.frame(x=label_x_positions, y = label_y_positions, gene = TE_df$gene) 
  
  p = ggplot(data = TE_df, aes(x = stage, y = TE)) + 
    geom_point( aes(color = gene, label = as.character(1:6)), position = position_dodge(width = 0.3), size = 1.8 ) + 
    geom_errorbar(aes(ymin=confidence_lower, ymax=confidence_upper, color = gene), 
             position=position_dodge(width = 0.3), width = 0.1) +
    geom_text(data = text_df, aes(x=x, y=y, label = gene), nudge_x = -0.2, size = 3) +
    #geom_text(aes( label = gene), nudge_x = -0.2, size = 3) +
    geom_segment(data = data.frame(w = c(1, 2), m = c(fourcell_median, eightcell_median)),
             aes(x     = w - thick_width ,
                 xend = w + thick_width,
                 y = m,
                 yend =m ) ,
             color = "#999999", size = 0.45 ) +

    theme_bw() + 
    theme(
      panel.border      = element_blank(),
      panel.grid        = element_blank(),
      plot.title        = PLOT_TITLE,
      panel.background  = element_blank(),
      axis.text.y       = AXIS_TITLE,
      axis.text.x       = AXIS_TITLE,
      axis.title.y      = AXIS_TITLE,
      axis.title.x      = AXIS_TITLE,
      legend.position   = "none",
      axis.line         = element_line(colour = "black", size = 0.35), 
    ) +
    scale_color_manual(values=c("#999999", "#999999", CDS_GREEN, UTR3_GREEN, CDS_GREEN, UTR3_GREEN)) + 
    labs(title = "Example Genes", y = "Translation Efficiency", x = "stage") 
  return(p)
}

main_example_TE_plot = plot_example_TE(example_TE_with_conf_intervals, fourcell_median = fourcell_background_median_TE, eightcell_median = eightcell_background_median_TE)

main_example_TE_plot
```


```{r echo=FALSE,warning=FALSE}

fourcell_main_boxplot = twoplot(fourcell_TE, title = "4 Cell", this_color = CDS_GREEN, type = "box", jitter = F, write_pval = T) 
eightcell_main_boxplot = twoplot(eightcell_TE, title = "8 Cell", this_color = UTR3_GREEN, type = "box", jitter = F, keep_y_axis = F, write_pval = T) 
fourcell_mono_boxplot = twoplot(fourcell_mono_TE, title = "4 Cell Mono", this_color = CDS_GREEN, type = "box", jitter = F, keep_y_axis = F, write_pval = T)
eightcell_mono_boxplot = twoplot(eightcell_mono_TE, title = "8 Cell Mono", this_color = UTR3_GREEN, type = "box", jitter = F, keep_y_axis = F, write_pval = T)


scatter_legend = get_legend(scatter_plot_of_bootstrapped + theme( legend.text=element_text(size=8), legend.title = element_blank() ) )

first_row_plot = plot_grid(scatter_plot_of_bootstrapped + theme(legend.position = "right", legend.text=element_text(size=6), legend.title=element_blank()), 
          fourcell_main_boxplot  ,
          eightcell_main_boxplot  ,
          fourcell_mono_boxplot ,
          eightcell_mono_boxplot , nrow =1, 
          rel_widths = c(3.4, 1.3, 1, 1, 1))



#imprinted_plot = plot_grid(first_row_plot, second_row_plot, ncol = 1, rel_heights = c(1, 1.1))
imprinted_plot = first_row_plot

save_plot_pdf("imprinted_main.pdf", imprinted_plot, width = 7, height = 2.3)
imprinted_plot
```




## Supplementary Figures

```{r echo=FALSE,warning=FALSE}

multisnp_title = ggdraw() + 
  draw_label(
    "Genes Supported by Multiple Snps",
    fontface   = 'plain',
    size       = FONT_TITLE_SIZE,
    angle      = 0
  ) +
  theme(
    plot.margin = margin(0, 0, 0, 0)
  )

matched_title = ggdraw() + 
  draw_label(
    "Genes with similar RNA abundance and CDS length",
    fontface   = 'plain',
    size       = FONT_TITLE_SIZE,
    angle      = 0
  ) +
  theme(
    plot.margin = margin(0, 0, 0, 0)
  )

example_gene_title = ggdraw() + 
  draw_label(
    "Examples of genes and their TE",
    fontface   = 'plain',
    size       = FONT_TITLE_SIZE,
    angle      = 0
  ) +
  theme(
    plot.margin = margin(0, 0, 0, 0)
  )

fourcell_multisnp_boxplot = twoplot(fourcell_multiple_snp_TE, title = "4 Cell", this_color = CDS_GREEN, type = "box", jitter = F, keep_y_axis = T, write_pval = T)
eightcell_multisnp_boxplot = twoplot(eightcell_multiple_snp_TE, title = "8 Cell", this_color = UTR3_GREEN, type = "box", jitter = F, keep_y_axis = F, write_pval = T) 

supp_first_row = plot_grid(scatter_plot_of_multisnp + labs(title = "Paternal Ratio"),
                           separator_1, fourcell_multisnp_boxplot, eightcell_multisnp_boxplot, nrow = 1, rel_widths = c(2.7,1,1,1) )

supp_first_row = plot_grid(multisnp_title, supp_first_row, ncol = 1, rel_heights = c(0.1, 1) )


fourcell_matched_boxplot = twoplot(matched_fourcell, title = "4 Cell CLR", this_color = CDS_GREEN, type = "box", jitter = F, keep_y_axis = T, write_pval = T) 

eightcell_matched_boxplot = twoplot(matched_eightcell, title = "8 Cell CLR", this_color = UTR3_GREEN, type = "box", jitter = F, keep_y_axis = F, write_pval = T) 

fourcell_cpm_matched_boxplot = twoplot(matched_fourcell_cpm, title = "4 Cell CPM", this_color = CDS_GREEN, type = "box", jitter = F, keep_y_axis = T, write_pval = T)

eightcell_cpm_matched_boxplot = twoplot(matched_eightcell_cpm, title = "8 Cell CPM", this_color = UTR3_GREEN, type = "box", jitter = F, keep_y_axis = F, write_pval = T) 

matched_box_plot = plot_grid(fourcell_matched_boxplot, eightcell_matched_boxplot, fourcell_cpm_matched_boxplot, eightcell_cpm_matched_boxplot, nrow = 1)
supp_second_row = plot_grid(matched_title, matched_box_plot, ncol = 1, rel_heights = c(0.1, 1))

example_genes_row_plot = plot_grid( snp_pos_plot,  main_example_TE_plot, nrow = 1, rel_widths = c(1.5,1))
supp_example_gene_row = plot_grid(example_gene_title , example_genes_row_plot, ncol = 1, rel_heights = c(0.1, 1))




#twoplot()

onecell_TE_title = ggdraw() + 
  draw_label(
    "TE at 1 Cell Embryo",
    fontface   = 'plain',
    size       = FONT_TITLE_SIZE,
    angle      = 0
  ) +
  theme(
    plot.margin = margin(0, 0, 0, 0)
  )

mii_TE_title = ggdraw() + 
  draw_label(
    "TE at MII",
    fontface   = 'plain',
    size       = FONT_TITLE_SIZE,
    angle      = 0
  ) +
  theme(
    plot.margin = margin(0, 0, 0, 0)
  )


matched_onecell_TE_from_4_plot = twoplot(matched_onecell_TE_from_4, title = "Specificity from 4 cell", this_color = CDS_GREEN, type = "box", jitter = F, write_pval = T, keep_y_axis = T) 
matched_onecell_TE_from_8_plot = twoplot(matched_onecell_TE_from_8, title = "Specificity from 8 cell", this_color = UTR3_GREEN, type = "box", jitter = F, write_pval = T, keep_y_axis = F)

onecell_TE_base_plot = plot_grid( matched_onecell_TE_from_4_plot, matched_onecell_TE_from_8_plot, nrow = 1)

onecell_TE_plot = plot_grid(onecell_TE_title, onecell_TE_base_plot, ncol = 1, rel_heights = c(0.1, 1))
#onecell_TE_base_plot
#


matched_onecell_TE_from_4_plot = twoplot(matched_mii_TE_from_4, title = "Specificity from 4 cell", this_color = CDS_GREEN, type = "box", jitter = F, write_pval = T, keep_y_axis = T) 
matched_onecell_TE_from_8_plot = twoplot(matched_mii_TE_from_8, title = "Specificity from 8 cell", this_color = UTR3_GREEN, type = "box", jitter = F, write_pval = T, keep_y_axis = F) 

mii_TE_base_plot = plot_grid( matched_onecell_TE_from_4_plot, matched_onecell_TE_from_8_plot, nrow = 1 )

mii_TE_plot = plot_grid(mii_TE_title, mii_TE_base_plot, ncol = 1, rel_heights = c(0.1, 1))

TE_at_other_stages_row = plot_grid(onecell_TE_plot, mii_TE_plot, nrow = 1)





maternal_paternal_title = ggdraw() + 
  draw_label(
    "Translation Efficiency of Maternal and Paternal Genes",
    fontface   = 'plain',
    size       = FONT_TITLE_SIZE,
    angle      = 0
  ) +
  theme(
    plot.margin = margin(0, 0, 0, 0)
  )

maternal_paternal_TE_4cell_plot = threeplot(fourcell_TE, title = "4 cell", this_color = c(CDS_GREEN), ymax = 1.5, type = "box", jitter = F, selector = "TE", keep_y_axis = T)
maternal_paternal_TE_8cell_plot = threeplot(eightcell_TE, title = "8 cell", this_color = c(UTR3_GREEN), ymax = 1.5, type = "box", jitter = F, selector = "TE", keep_y_axis = F)
maternal_paternal_plot_row_base = plot_grid(separator_1, maternal_paternal_TE_4cell_plot, maternal_paternal_TE_8cell_plot, separator_1, nrow = 1, rel_widths = c(0.5, 0.7, 0.7, 0.5))
maternal_paternal_plot_row      = plot_grid(maternal_paternal_title, maternal_paternal_plot_row_base, ncol = 1, rel_heights = c(0.1, 1))

imprinted_supp = plot_grid(maternal_paternal_plot_row, supp_first_row, supp_second_row,supp_example_gene_row, TE_at_other_stages_row,  ncol = 1)
save_plot_pdf("imprinted_supp.pdf", imprinted_supp, width = 7, height = 10.5)
#imprinted_supp
```


## Effect Sizes
```{r echo=FALSE,warning=FALSE}

report_effect_sizes = function(fourcell_df, eightcell_df){
  effect_size_eightcell = 
    median(eightcell_df[eightcell_df$imprinted == 1, ]$TE, na.rm = T ) /
    median(eightcell_df[eightcell_df$imprinted == 0, ]$TE , na.rm = T)
  
  effect_size_fourcell = 
    median(fourcell_df[fourcell_df$imprinted == 1, ]$TE, na.rm = T ) /
    median(fourcell_df[fourcell_df$imprinted == 0, ]$TE , na.rm = T)
  
  print( c("Effect sizes are: 4-cell", effect_size_fourcell, "8-cell", effect_size_eightcell))  
}



```

### Bootstrapped
```{r echo=FALSE,warning=FALSE}
report_effect_sizes(fourcell_TE, eightcell_TE)

```

### Multi-SNP
```{r echo=FALSE,warning=FALSE}
report_effect_sizes(fourcell_multiple_snp_TE , eightcell_multiple_snp_TE)

```

### Mono-Allelic
```{r echo=FALSE,warning=FALSE}
report_effect_sizes(fourcell_mono_TE , eightcell_mono_TE)

```

### Matched
```{r echo=FALSE,warning=FALSE}
report_effect_sizes(matched_fourcell , matched_eightcell)

```


### Once Cell
```{r echo=FALSE,warning=FALSE}
report_effect_sizes(matched_onecell_TE_from_4 , matched_onecell_TE_from_8)

```

### MII
```{r echo=FALSE,warning=FALSE}
report_effect_sizes(matched_mii_TE_from_4 , matched_mii_TE_from_8)

```


### Multi-SNP
```{r echo=FALSE,warning=FALSE}
report_effect_sizes(fourcell_multiple_snp_TE , eightcell_multiple_snp_TE)

```
