{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f3645943",
   "metadata": {},
   "source": [
    "# EXTRACT SNP COUNTS OF DETECTED GENES\n",
    "\n",
    "\n",
    "We extract the gene lists that are snp-expressed in 4 and 8 cell stages.\n",
    "\n",
    "The definition of expressed (or detected) genes come from our prop-test. \n",
    "See the definition of `perform_prop_test` for details"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0928c34b",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "990895ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import os\n",
    "\n",
    "from multiprocessing import Pool\n",
    "from matplotlib import pyplot as plt\n",
    "from functools import reduce\n",
    "\n",
    "import pickle\n",
    "\n",
    "# We use rpy2 for the prop.test and padjust in R\n",
    "import rpy2\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "from rpy2.robjects.packages import importr\n",
    "\n",
    "\n",
    "from IPython.display import display, HTML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "929a2112",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython\n",
    "%matplotlib inline\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [16, 8]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96fe2e3d",
   "metadata": {},
   "source": [
    "## Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8f48026f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The values of the confidence interval coming from the proptest\n",
    "# must be less than this value\n",
    "PROPTEST_ERROR_THRESHOLD = 0.05\n",
    "\n",
    "# After adjusting the p-values of the proptest\n",
    "# we pick the FDR values less than this threshold\n",
    "FDR_THRESHOLD = 0.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a4436eb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "75c8c644",
   "metadata": {},
   "source": [
    "## Input & Output Folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e087702b",
   "metadata": {},
   "outputs": [],
   "source": [
    "INPUT_FOLDER  = \"../snp\"\n",
    "OUTPUT_FOLDER = \".\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81600d55",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4b7c40b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RNA_SEQ EXPERIMENTS\n",
    "# Note that we excluded the experiment 20210607-RNAseq-4cell-cross-B\n",
    "\n",
    "rnaseq_one_cell_exps = [\n",
    "\"20210607-RNAseq-1cell-cross-A\",\n",
    "\"20210607-RNAseq-1cell-cross-B\",\n",
    "\"20210607-RNAseq-1cell-cross-C\",\n",
    "\"20210607-RNAseq-1cell-cross-D\"]\n",
    "\n",
    "rnaseq_two_cell_exps = [\n",
    "\"20210607-RNAseq-2cell-cross-A\",\n",
    "\"20210607-RNAseq-2cell-cross-B\",\n",
    "\"20210607-RNAseq-2cell-cross-C\",\n",
    "\"20210607-RNAseq-2cell-cross-D\"\n",
    "]\n",
    "\n",
    "rnaseq_four_cell_exps = [\n",
    "\"20210607-RNAseq-4cell-cross-A\",\n",
    "#\"20210607-RNAseq-4cell-cross-B\",\n",
    "\"20210607-RNAseq-4cell-cross-C\"\n",
    "]\n",
    "\n",
    "rnaseq_eight_cell_exps = [\n",
    "\"20210607-RNAseq-8cell-cross-A\",\n",
    "\"20210607-RNAseq-8cell-cross-B\",\n",
    "\"20210607-RNAseq-8cell-cross-C\",\n",
    "\"20210607-RNAseq-8cell-cross-D\"\n",
    "]\n",
    "\n",
    "rnaseq_gv_epxs = [\n",
    "\"20210607-RNAseq-GV-A\",\n",
    "\"20210607-RNAseq-GV-B\",\n",
    "\"20210607-RNAseq-GV-C\",\n",
    "\"20210607-RNAseq-GV-D\"\n",
    "]\n",
    "\n",
    "rnaseq_mii_exps = [\n",
    "\"20210607-RNAseq-MII-A\",\n",
    "\"20210607-RNAseq-MII-B\",\n",
    "\"20210607-RNAseq-MII-C\",\n",
    "\"20210607-RNAseq-MII-D\"\n",
    "]\n",
    "\n",
    "rnaseq_all_exps = rnaseq_one_cell_exps + rnaseq_two_cell_exps + \\\n",
    "                  rnaseq_four_cell_exps + rnaseq_eight_cell_exps +\\\n",
    "                  rnaseq_gv_epxs + rnaseq_mii_exps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "287c0d62",
   "metadata": {},
   "outputs": [],
   "source": [
    "## RIBOSOME PROFILING EXPERIMENTS\n",
    "\n",
    "riboseq_one_cell_exps = [\n",
    "\"20210513-ITP-1cell-cross-50-A\",\n",
    "\"20210513-ITP-1cell-cross-50-B\",\n",
    "\"20210513-ITP-1cell-cross-50-C\",\n",
    "\"20210513-ITP-1cell-cross-50-D\",\n",
    "\"20210513-ITP-1cell-cross-50-E\"]\n",
    "    \n",
    "riboseq_two_cell_exps = [\n",
    "\"20210513-ITP-2cell-cross-50-B\",\n",
    "\"20210513-ITP-2cell-cross-50-C\",\n",
    "\"20210513-ITP-2cell-cross-50-F\",\n",
    "]\n",
    "\n",
    "riboseq_four_cell_exps = [\n",
    "\"20210513-ITP-4cell-cross-50-B\",\n",
    "\"20210513-ITP-4cell-cross-50-C\",\n",
    "\"20210513-ITP-4cell-cross-50-D\",\n",
    "]\n",
    "    \n",
    "riboseq_eigth_cell_exps = [\n",
    "\"20210513-ITP-8cell-cross-50-A\",\n",
    "\"20210513-ITP-8cell-cross-50-B\",\n",
    "\"20210513-ITP-8cell-cross-50-C\",\n",
    "\"20210513-ITP-8cell-cross-50-D\",\n",
    "]\n",
    "    \n",
    "    \n",
    "riboseq_mii_exps = [\n",
    "\"20210301-ITP-MII-25-B\",\n",
    "\"20210301-ITP-MII-50-A\",\n",
    "\"20210301-ITP-MII-50-B\",\n",
    "\"20210318-ITP-MII-50-B\",\n",
    "#\"20210614-ITP-MII-50-A\",\n",
    "#\"20210614-ITP-MII-50-B\",\n",
    "#\"20210614-ITP-MII-50-C\",\n",
    "\"20210614-ITP-MII-50-D\",\n",
    "#\"20210614-ITP-MII-50-E\",\n",
    "#\"20210614-ITP-MII-50-F\",\n",
    "]    \n",
    "    \n",
    "    \n",
    "riboseq_gv_epxs = [\n",
    "\"20210614-ITP-GV-50-A\",\n",
    "\"20210614-ITP-GV-50-B\",\n",
    "\"20210614-ITP-GV-50-C\",\n",
    "\"20210614-ITP-GV-50-D\",\n",
    "\"20210614-ITP-GV-50-E\",\n",
    "\"20210614-ITP-GV-50-F\",\n",
    "]\n",
    "\n",
    "riboseq_all_exps = riboseq_one_cell_exps  + riboseq_two_cell_exps + \\\n",
    "                   riboseq_four_cell_exps + riboseq_eigth_cell_exps + \\\n",
    "                   riboseq_mii_exps       + riboseq_gv_epxs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "86f98d1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## FUNCTIONS TO READ SNP FILES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "88fe0cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_CDS(name_str):\n",
    "    \"\"\"\n",
    "    Returns CDS boundaries from the transcript header\n",
    "    \"\"\"\n",
    "    \n",
    "    contents = name_str.split(\"|\")\n",
    "    \n",
    "    for c in contents:\n",
    "        if c.startswith(\"CDS\"):\n",
    "            boundaries = c.split(\":\")[1].split(\"-\")\n",
    "            return (int(boundaries[0]), int(boundaries[1]))\n",
    "    \n",
    "    print(name_str)\n",
    "    raise ValueError \n",
    "    \n",
    "    \n",
    "    \n",
    "############################################################\n",
    "    \n",
    "    \n",
    "    \n",
    "def get_df(snp_count_file):\n",
    "    \"\"\"\n",
    "    Reads SNP cont file into a dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    count_df         = pd.read_csv(snp_count_file, sep = \"\\t\")\n",
    "    transcript_names = list( map( lambda x: x.split(\"|\")[4], count_df[\"transcript\"]) )\n",
    "\n",
    "    boundaries       = list( map(extract_CDS, count_df[\"transcript\"])  )\n",
    "\n",
    "    start_positions  = list( map(lambda x: x[0], boundaries ) )\n",
    "    stop_positions   = list( map(lambda x: x[1], boundaries ) )\n",
    "\n",
    "    snp_total = count_df[\"A\"] + count_df[\"C\"] + count_df[\"G\"] +  count_df[\"T\"]\n",
    "\n",
    "    count_df[\"transcript\"] = transcript_names\n",
    "    count_df[\"snp_total\"]  = snp_total\n",
    "    count_df[\"CDS_start\"]  = start_positions\n",
    "    count_df[\"CDS_stop\"]   = stop_positions\n",
    "\n",
    "\n",
    "    count_df.sort_values(by = ['snp_total'], ascending=False, inplace=True)\n",
    "    \n",
    "    return count_df    \n",
    "\n",
    "#################################################################\n",
    "\n",
    "def get_ref_alt_counts(count_df):\n",
    "    \"\"\"\n",
    "    Extracts the reference (maternal) and alternative (paternal) \n",
    "    counts from the dataframes\n",
    "    \"\"\"\n",
    "    \n",
    "    ref_count   = 0\n",
    "    alt_count   = 0\n",
    "    total_count = 0\n",
    "    \n",
    "    cds_ref_count   = 0\n",
    "    cds_alt_count   = 0\n",
    "    cds_total_count = 0\n",
    "\n",
    "    for index, row in count_df.iterrows():\n",
    "        total_count += row[\"A\"] + row[\"C\"] + row[\"G\"] + row[\"T\"]\n",
    "        ref_count   += row[ row[\"REF\"] ]\n",
    "        alt_count   += row[ row[\"ALT\"] ]\n",
    "        \n",
    "        if row[\"position\"] >= row[\"CDS_start\"] and\\\n",
    "           row[\"position\"] <= row[\"CDS_stop\"]:\n",
    "                cds_total_count += row[\"A\"] + row[\"C\"] + row[\"G\"] + row[\"T\"]\n",
    "                cds_ref_count   += row[ row[\"REF\"] ]\n",
    "                cds_alt_count   += row[ row[\"ALT\"] ]            \n",
    "\n",
    "        \n",
    "    return {\"ref_count\":     ref_count,     \"alt_count\": alt_count,         \"total_count\": total_count,\n",
    "            \"cds_ref_count\": cds_ref_count, \"cds_alt_count\": cds_alt_count, \"cds_total_count\": cds_total_count}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfaa15af",
   "metadata": {},
   "source": [
    "## RNA-Seq SNP Calls\n",
    "\n",
    "We read SNP calls from RNA-Seq into dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c92ed715",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnaseq_experiment_dfs = dict()\n",
    "\n",
    "for e in rnaseq_all_exps:\n",
    "    this_file = os.path.join( INPUT_FOLDER, \"snp_counts_rnaseq\", e + \".tsv.gz\")\n",
    "    rnaseq_experiment_dfs[e] = get_df(this_file)\n",
    "    rnaseq_experiment_dfs[e][\"experiment\"] = e\n",
    "    \n",
    "rnaseq_reference_alt_counts = dict()\n",
    "\n",
    "with Pool(4) as p:\n",
    "    rnaseq_counts_array = p.map( get_ref_alt_counts, [ rnaseq_experiment_dfs[e] for e in rnaseq_all_exps ] )\n",
    "    \n",
    "for i in range(len(rnaseq_all_exps)):\n",
    "    rnaseq_reference_alt_counts[rnaseq_all_exps[i]] = rnaseq_counts_array[i]\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "508c19b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref_count</th>\n",
       "      <th>alt_count</th>\n",
       "      <th>total_count</th>\n",
       "      <th>cds_ref_count</th>\n",
       "      <th>cds_alt_count</th>\n",
       "      <th>cds_total_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20210607-RNAseq-1cell-cross-A</th>\n",
       "      <td>179914</td>\n",
       "      <td>314</td>\n",
       "      <td>180540</td>\n",
       "      <td>179914</td>\n",
       "      <td>314</td>\n",
       "      <td>180540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20210607-RNAseq-1cell-cross-B</th>\n",
       "      <td>180976</td>\n",
       "      <td>373</td>\n",
       "      <td>181700</td>\n",
       "      <td>180976</td>\n",
       "      <td>373</td>\n",
       "      <td>181700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               ref_count  alt_count  total_count  \\\n",
       "20210607-RNAseq-1cell-cross-A     179914        314       180540   \n",
       "20210607-RNAseq-1cell-cross-B     180976        373       181700   \n",
       "\n",
       "                               cds_ref_count  cds_alt_count  cds_total_count  \n",
       "20210607-RNAseq-1cell-cross-A         179914            314           180540  \n",
       "20210607-RNAseq-1cell-cross-B         180976            373           181700  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnaseq_snp_count_df  = pd.DataFrame.from_dict(rnaseq_reference_alt_counts).transpose()\n",
    "rnseq_snp_count_file = os.path.join(OUTPUT_FOLDER, \"rnaseq_experimentwise_snp_counts.csv\")\n",
    "rnaseq_snp_count_df.to_csv(rnseq_snp_count_file)\n",
    "\n",
    "rnaseq_snp_count_df[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4a583b79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>snp_total</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>CDS_stop</th>\n",
       "      <th>experiment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68514</th>\n",
       "      <td>Rps19-203</td>\n",
       "      <td>303</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>521</td>\n",
       "      <td>1</td>\n",
       "      <td>616</td>\n",
       "      <td>1139</td>\n",
       "      <td>298</td>\n",
       "      <td>735</td>\n",
       "      <td>20210607-RNAseq-4cell-cross-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66330</th>\n",
       "      <td>Dppa3-201</td>\n",
       "      <td>290</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>691</td>\n",
       "      <td>403</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1095</td>\n",
       "      <td>114</td>\n",
       "      <td>566</td>\n",
       "      <td>20210607-RNAseq-4cell-cross-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4478</th>\n",
       "      <td>Rps26-201</td>\n",
       "      <td>299</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>505</td>\n",
       "      <td>0</td>\n",
       "      <td>517</td>\n",
       "      <td>1022</td>\n",
       "      <td>270</td>\n",
       "      <td>617</td>\n",
       "      <td>20210607-RNAseq-4cell-cross-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66329</th>\n",
       "      <td>Dppa3-201</td>\n",
       "      <td>263</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>509</td>\n",
       "      <td>0</td>\n",
       "      <td>437</td>\n",
       "      <td>946</td>\n",
       "      <td>114</td>\n",
       "      <td>566</td>\n",
       "      <td>20210607-RNAseq-4cell-cross-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23498</th>\n",
       "      <td>Rpl15-202</td>\n",
       "      <td>275</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>441</td>\n",
       "      <td>0</td>\n",
       "      <td>474</td>\n",
       "      <td>0</td>\n",
       "      <td>915</td>\n",
       "      <td>120</td>\n",
       "      <td>734</td>\n",
       "      <td>20210607-RNAseq-4cell-cross-C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript  position REF ALT    A    C    G    T  snp_total  CDS_start  \\\n",
       "68514  Rps19-203       303   C   T    1  521    1  616       1139        298   \n",
       "66330  Dppa3-201       290   A   C  691  403    1    0       1095        114   \n",
       "4478   Rps26-201       299   C   T    0  505    0  517       1022        270   \n",
       "66329  Dppa3-201       263   C   T    0  509    0  437        946        114   \n",
       "23498  Rpl15-202       275   A   G  441    0  474    0        915        120   \n",
       "\n",
       "       CDS_stop                     experiment  \n",
       "68514       735  20210607-RNAseq-4cell-cross-C  \n",
       "66330       566  20210607-RNAseq-4cell-cross-C  \n",
       "4478        617  20210607-RNAseq-4cell-cross-C  \n",
       "66329       566  20210607-RNAseq-4cell-cross-C  \n",
       "23498       734  20210607-RNAseq-4cell-cross-C  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnaseq_experiment_dfs[\"20210607-RNAseq-4cell-cross-C\"][:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd5c4f75",
   "metadata": {},
   "source": [
    "## Ribosome Profiling SNP Calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bd0bd8d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "riboseq_experiment_dfs = dict()\n",
    "\n",
    "for e in riboseq_all_exps:\n",
    "    this_file = os.path.join(INPUT_FOLDER,\"snp_counts_riboseq\", e + \".tsv.gz\")\n",
    "    riboseq_experiment_dfs[e] = get_df(this_file)\n",
    "    riboseq_experiment_dfs[e][\"experiment\"] = e\n",
    "    \n",
    "\n",
    "riboseq_reference_alt_counts = dict()\n",
    "\n",
    "with Pool(4) as p:\n",
    "    riboseq_counts_array = p.map( get_ref_alt_counts, [ riboseq_experiment_dfs[e] for e in riboseq_all_exps ] )\n",
    "    \n",
    "for i in range(len(riboseq_all_exps)):\n",
    "    riboseq_reference_alt_counts[riboseq_all_exps[i]] = riboseq_counts_array[i]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f829ac6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for e in riboseq_all_exps:\n",
    "#    print(e)\n",
    "#    print(\"Alt: {:.1f}    Ref: {:.1f}\".format( riboseq_reference_alt_percentages[e][\"cds_alt_percentage\"],\n",
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eab17dc2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3345f003",
   "metadata": {},
   "source": [
    "## SNP Data at Highest Detail\n",
    "\n",
    "We combine all of the SNP data into one dataframe. \n",
    "For each experiment, transcript and SNP position there is a separate row with SNP information.\n",
    "Note that, in our context, reference (REF) corresponds to maternal strain and alternative (ALT) corresponds to paternal strain."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1250bd7b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c631c580",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_maternal_paternal_snps(this_df):\n",
    "    maternal_counts = list()\n",
    "    paternal_counts = list()\n",
    "    \n",
    "    \n",
    "    for i, contents in this_df.iterrows():\n",
    "        this_maternal = contents[ contents[\"REF\"] ]\n",
    "        this_paternal = contents[ contents[\"ALT\"] ]\n",
    "        paternal_counts.append(this_paternal)\n",
    "        maternal_counts.append(this_maternal)\n",
    "        \n",
    "    this_df[\"maternal\"] = maternal_counts\n",
    "    this_df[\"paternal\"] = paternal_counts\n",
    "    \n",
    "    return this_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1e962201",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find ribosome profiling maternal / poaternal counts \n",
    "\n",
    "with Pool(4) as p:\n",
    "    riboseq_detailed_count_dfs = p.map( find_maternal_paternal_snps, list(riboseq_experiment_dfs.values()) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d217d2af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find rna-seq maternal / poaternal counts \n",
    "\n",
    "with Pool(4) as p:\n",
    "    rnaseq_detailed_count_dfs = p.map( find_maternal_paternal_snps, list(rnaseq_experiment_dfs.values()) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a194d915",
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_snp_dfs_into_on_df(df_list):\n",
    "    \n",
    "    res_columns = [\"experiment\", \"transcript\", \"position\", \"paternal\",\\\n",
    "                   \"maternal\", \"REF\", \"ALT\", \"A\", \"C\", \"G\", \"T\"]\n",
    "    \n",
    "    result_df   =  df_list[0].copy()\n",
    "    \n",
    "    for i in range(1, len(df_list)):\n",
    "        result_df = result_df.append( df_list[i] )\n",
    "        \n",
    "    return result_df[res_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "69b5c07e",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_riboseq_snp_df = combine_snp_dfs_into_on_df(riboseq_detailed_count_dfs)\n",
    "combined_rnaseq_snp_df  = combine_snp_dfs_into_on_df(rnaseq_detailed_count_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "00e9f5bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "high_res_riboseq_snp_file = os.path.join(OUTPUT_FOLDER, \"riboseq_detailed_snps.csv.gz\")\n",
    "combined_riboseq_snp_df.to_csv( high_res_riboseq_snp_file, index = False )\n",
    "\n",
    "high_res_rnaseq_snp_file = os.path.join(OUTPUT_FOLDER, \"rnaseq_detailed_snps.csv.gz\")\n",
    "combined_rnaseq_snp_df.to_csv( high_res_rnaseq_snp_file, index = False )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "680f700c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "813b890b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment</th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>paternal</th>\n",
       "      <th>maternal</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21047</th>\n",
       "      <td>20210513-ITP-1cell-cross-50-A</td>\n",
       "      <td>Spin1-201</td>\n",
       "      <td>848</td>\n",
       "      <td>0</td>\n",
       "      <td>139</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>139</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22356</th>\n",
       "      <td>20210513-ITP-1cell-cross-50-A</td>\n",
       "      <td>Zbed3-202</td>\n",
       "      <td>471</td>\n",
       "      <td>0</td>\n",
       "      <td>124</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22358</th>\n",
       "      <td>20210513-ITP-1cell-cross-50-A</td>\n",
       "      <td>Zbed3-202</td>\n",
       "      <td>707</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23711</th>\n",
       "      <td>20210513-ITP-1cell-cross-50-A</td>\n",
       "      <td>E330034G19Rik-203</td>\n",
       "      <td>257</td>\n",
       "      <td>0</td>\n",
       "      <td>74</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>74</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22354</th>\n",
       "      <td>20210513-ITP-1cell-cross-50-A</td>\n",
       "      <td>Zbed3-202</td>\n",
       "      <td>320</td>\n",
       "      <td>4</td>\n",
       "      <td>52</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29124</th>\n",
       "      <td>20210614-ITP-GV-50-F</td>\n",
       "      <td>Slc4a8-201</td>\n",
       "      <td>3090</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29122</th>\n",
       "      <td>20210614-ITP-GV-50-F</td>\n",
       "      <td>Slc4a8-201</td>\n",
       "      <td>2048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29121</th>\n",
       "      <td>20210614-ITP-GV-50-F</td>\n",
       "      <td>Slc4a8-201</td>\n",
       "      <td>1791</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29120</th>\n",
       "      <td>20210614-ITP-GV-50-F</td>\n",
       "      <td>Slc4a8-201</td>\n",
       "      <td>1494</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85338</th>\n",
       "      <td>20210614-ITP-GV-50-F</td>\n",
       "      <td>Mid1-205</td>\n",
       "      <td>1745</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2218814 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          experiment         transcript  position  paternal  \\\n",
       "21047  20210513-ITP-1cell-cross-50-A          Spin1-201       848         0   \n",
       "22356  20210513-ITP-1cell-cross-50-A          Zbed3-202       471         0   \n",
       "22358  20210513-ITP-1cell-cross-50-A          Zbed3-202       707         0   \n",
       "23711  20210513-ITP-1cell-cross-50-A  E330034G19Rik-203       257         0   \n",
       "22354  20210513-ITP-1cell-cross-50-A          Zbed3-202       320         4   \n",
       "...                              ...                ...       ...       ...   \n",
       "29124           20210614-ITP-GV-50-F         Slc4a8-201      3090         0   \n",
       "29122           20210614-ITP-GV-50-F         Slc4a8-201      2048         0   \n",
       "29121           20210614-ITP-GV-50-F         Slc4a8-201      1791         0   \n",
       "29120           20210614-ITP-GV-50-F         Slc4a8-201      1494         0   \n",
       "85338           20210614-ITP-GV-50-F           Mid1-205      1745         0   \n",
       "\n",
       "       maternal REF ALT    A    C  G    T  \n",
       "21047       139   C   G    0  139  0    0  \n",
       "22356       124   T   C    0    0  0  124  \n",
       "22358       100   A   G  100    0  0    0  \n",
       "23711        74   A   C   74    0  0    0  \n",
       "22354        52   T   G    4    0  4   52  \n",
       "...         ...  ..  ..  ...  ... ..  ...  \n",
       "29124         0   A   G    0    0  0    0  \n",
       "29122         0   T   A    0    0  0    0  \n",
       "29121         0   G   A    0    0  0    0  \n",
       "29120         0   G   A    0    0  0    0  \n",
       "85338         0   A   G    0    0  0    0  \n",
       "\n",
       "[2218814 rows x 11 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_riboseq_snp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50504f5a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "98418cc0",
   "metadata": {},
   "source": [
    "## Aggregating SNP Counts\n",
    "\n",
    "We sum SNPs accross replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ded20cf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_two_snp_dfs(df_1, df_2):\n",
    "    \"\"\"\n",
    "    This is the helper function to add dataframes\n",
    "    \n",
    "    Adds two dataframes and produces one that is compatible with other SNP dataframes\n",
    "    (Have the same column names)\n",
    "    \"\"\"\n",
    "    \n",
    "    index_col_name = \"trans_pos\"\n",
    "    \n",
    "    zipped_df1_transcripts_df1_positions = zip( df_1[\"transcript\"], df_1[\"position\"] )\n",
    "    zipped_df2_transcripts_df2_positions = zip( df_2[\"transcript\"], df_2[\"position\"] )\n",
    "    \n",
    "    df1_index_vec = list( map( lambda x: str(x[0]) + \"_\" + str(x[1]), zipped_df1_transcripts_df1_positions ) )\n",
    "    df2_index_vec = list( map( lambda x: str(x[0]) + \"_\" + str(x[1]), zipped_df2_transcripts_df2_positions ) )\n",
    "    \n",
    "    df_1_indexed = df_1.copy()\n",
    "    df_2_indexed = df_2.copy()\n",
    "    \n",
    "    df_1_indexed[index_col_name] = df1_index_vec\n",
    "    df_2_indexed[index_col_name] = df2_index_vec\n",
    "    \n",
    "    df_1_indexed.set_index(index_col_name, inplace = True)\n",
    "    df_2_indexed.set_index(index_col_name, inplace = True)\n",
    "    \n",
    "    merged_df = df_1_indexed.merge( df_2_indexed, on = [\"trans_pos\"] )\n",
    "    \n",
    "    df_dict = {\n",
    "        #\"trnas_pos\":  merged_df.index,\n",
    "        \"transcript\": merged_df[\"transcript_x\"],\n",
    "        \"position\":   merged_df[\"position_x\"],\n",
    "        \"REF\":        merged_df[\"REF_x\"],\n",
    "        \"ALT\":        merged_df[\"ALT_x\"],\n",
    "        \"A\":          np.add(merged_df[\"A_x\"] , merged_df[\"A_y\"]),\n",
    "        \"C\":          np.add(merged_df[\"C_x\"] , merged_df[\"C_y\"]),\n",
    "        \"G\":          np.add(merged_df[\"G_x\"] , merged_df[\"G_y\"]),\n",
    "        \"T\":          np.add(merged_df[\"T_x\"] , merged_df[\"T_y\"]),\n",
    "        \"snp_total\":  np.add(merged_df[\"snp_total_x\"] , merged_df[\"snp_total_y\"]),\n",
    "        \"CDS_start\":  merged_df[\"CDS_start_x\"],\n",
    "        \"CDS_stop\":  merged_df[\"CDS_stop_x\"]     \n",
    "    }\n",
    "    \n",
    "    result_df = pd.DataFrame.from_dict(df_dict)\n",
    "    \n",
    "    return (result_df)\n",
    "\n",
    "\n",
    "##############################################################################\n",
    "\n",
    "\n",
    "def add_snp_dfs(df_list):\n",
    "    \"\"\"\n",
    "    THIS IS THE MAIN FUNCTION TO CALL!!!\n",
    "    Adds a list of dataframes containing SNPS\n",
    "    \"\"\"\n",
    "    result_df =  reduce( lambda x, y : add_two_snp_dfs(x,y), df_list  )\n",
    "    result_df.sort_values(by = ['snp_total'], ascending=False, inplace=True)\n",
    "    return result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1cd304b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>snp_total</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>CDS_stop</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trans_pos</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Spin1-201_848</th>\n",
       "      <td>Spin1-201</td>\n",
       "      <td>848</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>320</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>323</td>\n",
       "      <td>261</td>\n",
       "      <td>1049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zbed3-202_471</th>\n",
       "      <td>Zbed3-202</td>\n",
       "      <td>471</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>217</td>\n",
       "      <td>217</td>\n",
       "      <td>174</td>\n",
       "      <td>860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zbed3-202_707</th>\n",
       "      <td>Zbed3-202</td>\n",
       "      <td>707</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>218</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>218</td>\n",
       "      <td>174</td>\n",
       "      <td>860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E330034G19Rik-203_257</th>\n",
       "      <td>E330034G19Rik-203</td>\n",
       "      <td>257</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>125</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>125</td>\n",
       "      <td>56</td>\n",
       "      <td>1111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zbed3-202_320</th>\n",
       "      <td>Zbed3-202</td>\n",
       "      <td>320</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>130</td>\n",
       "      <td>150</td>\n",
       "      <td>174</td>\n",
       "      <td>860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gpd1-201_825</th>\n",
       "      <td>Gpd1-201</td>\n",
       "      <td>825</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>1143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gpd1-201_508</th>\n",
       "      <td>Gpd1-201</td>\n",
       "      <td>508</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>1143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gpd1-201_462</th>\n",
       "      <td>Gpd1-201</td>\n",
       "      <td>462</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>1143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Smarcd1-201_1699</th>\n",
       "      <td>Smarcd1-201</td>\n",
       "      <td>1699</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>194</td>\n",
       "      <td>1741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mid1-205_1745</th>\n",
       "      <td>Mid1-205</td>\n",
       "      <td>1745</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>345</td>\n",
       "      <td>2387</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85339 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              transcript  position REF ALT    A    C   G    T  \\\n",
       "trans_pos                                                                       \n",
       "Spin1-201_848                  Spin1-201       848   C   G    0  320   2    1   \n",
       "Zbed3-202_471                  Zbed3-202       471   T   C    0    0   0  217   \n",
       "Zbed3-202_707                  Zbed3-202       707   A   G  218    0   0    0   \n",
       "E330034G19Rik-203_257  E330034G19Rik-203       257   A   C  125    0   0    0   \n",
       "Zbed3-202_320                  Zbed3-202       320   T   G    8    2  10  130   \n",
       "...                                  ...       ...  ..  ..  ...  ...  ..  ...   \n",
       "Gpd1-201_825                    Gpd1-201       825   T   C    0    0   0    0   \n",
       "Gpd1-201_508                    Gpd1-201       508   C   T    0    0   0    0   \n",
       "Gpd1-201_462                    Gpd1-201       462   C   T    0    0   0    0   \n",
       "Smarcd1-201_1699             Smarcd1-201      1699   G   A    0    0   0    0   \n",
       "Mid1-205_1745                   Mid1-205      1745   A   G    0    0   0    0   \n",
       "\n",
       "                       snp_total  CDS_start  CDS_stop  \n",
       "trans_pos                                              \n",
       "Spin1-201_848                323        261      1049  \n",
       "Zbed3-202_471                217        174       860  \n",
       "Zbed3-202_707                218        174       860  \n",
       "E330034G19Rik-203_257        125         56      1111  \n",
       "Zbed3-202_320                150        174       860  \n",
       "...                          ...        ...       ...  \n",
       "Gpd1-201_825                   0         94      1143  \n",
       "Gpd1-201_508                   0         94      1143  \n",
       "Gpd1-201_462                   0         94      1143  \n",
       "Smarcd1-201_1699               0        194      1741  \n",
       "Mid1-205_1745                  0        345      2387  \n",
       "\n",
       "[85339 rows x 11 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "add_two_snp_dfs( riboseq_experiment_dfs[riboseq_one_cell_exps[0]], riboseq_experiment_dfs[riboseq_one_cell_exps[1]] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "fa43fdd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "riboseq_snp_total_one_cell   = add_snp_dfs( [riboseq_experiment_dfs[e] for e in  riboseq_one_cell_exps] )\n",
    "riboseq_snp_total_two_cell   = add_snp_dfs( [riboseq_experiment_dfs[e] for e in  riboseq_two_cell_exps] )\n",
    "riboseq_snp_total_four_cell  = add_snp_dfs( [riboseq_experiment_dfs[e] for e in  riboseq_four_cell_exps] )\n",
    "riboseq_snp_total_eight_cell = add_snp_dfs( [riboseq_experiment_dfs[e] for e in  riboseq_eigth_cell_exps] )\n",
    "\n",
    "\n",
    "riboseq_snp_total_mii = add_snp_dfs( [riboseq_experiment_dfs[e] for e in  riboseq_mii_exps] )\n",
    "riboseq_snp_total_gv  = add_snp_dfs( [riboseq_experiment_dfs[e] for e in  riboseq_gv_epxs] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "be9cf965",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnaseq_snp_total_one_cell   = add_snp_dfs( [rnaseq_experiment_dfs[e] for e in  rnaseq_one_cell_exps] )\n",
    "rnaseq_snp_total_two_cell   = add_snp_dfs( [rnaseq_experiment_dfs[e] for e in  rnaseq_two_cell_exps] )\n",
    "rnaseq_snp_total_four_cell  = add_snp_dfs( [rnaseq_experiment_dfs[e] for e in  rnaseq_four_cell_exps] )\n",
    "rnaseq_snp_total_eight_cell = add_snp_dfs( [rnaseq_experiment_dfs[e] for e in  rnaseq_eight_cell_exps] )\n",
    "\n",
    "\n",
    "rnaseq_snp_total_mii = add_snp_dfs( [rnaseq_experiment_dfs[e] for e in  rnaseq_mii_exps] )\n",
    "rnaseq_snp_total_gv  = add_snp_dfs( [rnaseq_experiment_dfs[e] for e in  rnaseq_gv_epxs] )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cb7e893",
   "metadata": {},
   "source": [
    "We add the paternal / maternal counts ratios to our dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1298244a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_maternal_ratio(snp_df):\n",
    "    maternal_ratios      = list()\n",
    "    maternal_count       = list()\n",
    "    paternal_count       = list()\n",
    "    mat_pat_count        = list()\n",
    "    \n",
    "    for i, e in snp_df.iterrows():\n",
    "        \n",
    "        maternal_nuc  = e[\"REF\"]\n",
    "        paternal_nuc  = e[\"ALT\"]\n",
    "        mat_pat_total = e[maternal_nuc] + e[paternal_nuc]\n",
    "        \n",
    "        # Avoid division by 0\n",
    "        if mat_pat_total == 0:\n",
    "            mat_pat_total = 1\n",
    "            \n",
    "        maternal_ratio = e[maternal_nuc] / mat_pat_total\n",
    "        \n",
    "        mat_pat_count.append(mat_pat_total)\n",
    "        maternal_ratios.append(maternal_ratio)\n",
    "        maternal_count.append(e[maternal_nuc])\n",
    "        paternal_count.append(e[paternal_nuc])\n",
    "        \n",
    "    snp_df[\"maternal_count\"] = maternal_count \n",
    "    snp_df[\"paternal_count\"] = paternal_count \n",
    "    snp_df[\"mat_pat_total\"]  = mat_pat_count \n",
    "    snp_df[\"maternal_ratio\"] = maternal_ratios\n",
    "    \n",
    "    return snp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ca27fedd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>snp_total</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>CDS_stop</th>\n",
       "      <th>maternal_count</th>\n",
       "      <th>paternal_count</th>\n",
       "      <th>mat_pat_total</th>\n",
       "      <th>maternal_ratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trans_pos</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Padi6-201_165</th>\n",
       "      <td>Padi6-201</td>\n",
       "      <td>165</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>609</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>612</td>\n",
       "      <td>64</td>\n",
       "      <td>2112</td>\n",
       "      <td>609</td>\n",
       "      <td>0</td>\n",
       "      <td>609</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zbed3-202_471</th>\n",
       "      <td>Zbed3-202</td>\n",
       "      <td>471</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>522</td>\n",
       "      <td>531</td>\n",
       "      <td>174</td>\n",
       "      <td>860</td>\n",
       "      <td>522</td>\n",
       "      <td>2</td>\n",
       "      <td>524</td>\n",
       "      <td>0.996183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ldhb-201_969</th>\n",
       "      <td>Ldhb-201</td>\n",
       "      <td>969</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>409</td>\n",
       "      <td>412</td>\n",
       "      <td>112</td>\n",
       "      <td>1116</td>\n",
       "      <td>409</td>\n",
       "      <td>3</td>\n",
       "      <td>412</td>\n",
       "      <td>0.992718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ankrd26-201_1877</th>\n",
       "      <td>Ankrd26-201</td>\n",
       "      <td>1877</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>367</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>367</td>\n",
       "      <td>170</td>\n",
       "      <td>5215</td>\n",
       "      <td>367</td>\n",
       "      <td>0</td>\n",
       "      <td>367</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Padi6-201_909</th>\n",
       "      <td>Padi6-201</td>\n",
       "      <td>909</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>279</td>\n",
       "      <td>284</td>\n",
       "      <td>64</td>\n",
       "      <td>2112</td>\n",
       "      <td>279</td>\n",
       "      <td>1</td>\n",
       "      <td>280</td>\n",
       "      <td>0.996429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gcnt3-201_563</th>\n",
       "      <td>Gcnt3-201</td>\n",
       "      <td>563</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>1826</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gcnt3-201_737</th>\n",
       "      <td>Gcnt3-201</td>\n",
       "      <td>737</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>1826</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gcnt3-201_986</th>\n",
       "      <td>Gcnt3-201</td>\n",
       "      <td>986</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>1826</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gcnt3-201_1002</th>\n",
       "      <td>Gcnt3-201</td>\n",
       "      <td>1002</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>1826</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mid1-205_1745</th>\n",
       "      <td>Mid1-205</td>\n",
       "      <td>1745</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>345</td>\n",
       "      <td>2387</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85339 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   transcript  position REF ALT    A    C  G    T  snp_total  \\\n",
       "trans_pos                                                                      \n",
       "Padi6-201_165       Padi6-201       165   A   G  609    0  0    3        612   \n",
       "Zbed3-202_471       Zbed3-202       471   T   C    6    2  1  522        531   \n",
       "Ldhb-201_969         Ldhb-201       969   T   C    0    3  0  409        412   \n",
       "Ankrd26-201_1877  Ankrd26-201      1877   C   A    0  367  0    0        367   \n",
       "Padi6-201_909       Padi6-201       909   T   C    4    1  0  279        284   \n",
       "...                       ...       ...  ..  ..  ...  ... ..  ...        ...   \n",
       "Gcnt3-201_563       Gcnt3-201       563   T   C    0    0  0    0          0   \n",
       "Gcnt3-201_737       Gcnt3-201       737   A   G    0    0  0    0          0   \n",
       "Gcnt3-201_986       Gcnt3-201       986   T   C    0    0  0    0          0   \n",
       "Gcnt3-201_1002      Gcnt3-201      1002   A   G    0    0  0    0          0   \n",
       "Mid1-205_1745        Mid1-205      1745   A   G    0    0  0    0          0   \n",
       "\n",
       "                  CDS_start  CDS_stop  maternal_count  paternal_count  \\\n",
       "trans_pos                                                               \n",
       "Padi6-201_165            64      2112             609               0   \n",
       "Zbed3-202_471           174       860             522               2   \n",
       "Ldhb-201_969            112      1116             409               3   \n",
       "Ankrd26-201_1877        170      5215             367               0   \n",
       "Padi6-201_909            64      2112             279               1   \n",
       "...                     ...       ...             ...             ...   \n",
       "Gcnt3-201_563           513      1826               0               0   \n",
       "Gcnt3-201_737           513      1826               0               0   \n",
       "Gcnt3-201_986           513      1826               0               0   \n",
       "Gcnt3-201_1002          513      1826               0               0   \n",
       "Mid1-205_1745           345      2387               0               0   \n",
       "\n",
       "                  mat_pat_total  maternal_ratio  \n",
       "trans_pos                                        \n",
       "Padi6-201_165               609        1.000000  \n",
       "Zbed3-202_471               524        0.996183  \n",
       "Ldhb-201_969                412        0.992718  \n",
       "Ankrd26-201_1877            367        1.000000  \n",
       "Padi6-201_909               280        0.996429  \n",
       "...                         ...             ...  \n",
       "Gcnt3-201_563                 1        0.000000  \n",
       "Gcnt3-201_737                 1        0.000000  \n",
       "Gcnt3-201_986                 1        0.000000  \n",
       "Gcnt3-201_1002                1        0.000000  \n",
       "Mid1-205_1745                 1        0.000000  \n",
       "\n",
       "[85339 rows x 15 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_maternal_ratio( riboseq_snp_total_one_cell )\n",
    "find_maternal_ratio( riboseq_snp_total_two_cell ) \n",
    "find_maternal_ratio( riboseq_snp_total_four_cell ) \n",
    "find_maternal_ratio( riboseq_snp_total_eight_cell )\n",
    "\n",
    "find_maternal_ratio( riboseq_snp_total_mii )\n",
    "find_maternal_ratio( riboseq_snp_total_gv ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a549651e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>snp_total</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>CDS_stop</th>\n",
       "      <th>maternal_count</th>\n",
       "      <th>paternal_count</th>\n",
       "      <th>mat_pat_total</th>\n",
       "      <th>maternal_ratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trans_pos</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H1f8-201_69</th>\n",
       "      <td>H1f8-201</td>\n",
       "      <td>69</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>6102</td>\n",
       "      <td>6106</td>\n",
       "      <td>46</td>\n",
       "      <td>960</td>\n",
       "      <td>6102</td>\n",
       "      <td>3</td>\n",
       "      <td>6105</td>\n",
       "      <td>0.999509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ooep-201_257</th>\n",
       "      <td>Ooep-201</td>\n",
       "      <td>257</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>6037</td>\n",
       "      <td>6039</td>\n",
       "      <td>198</td>\n",
       "      <td>692</td>\n",
       "      <td>6037</td>\n",
       "      <td>2</td>\n",
       "      <td>6039</td>\n",
       "      <td>0.999669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ftdc2-201_438</th>\n",
       "      <td>Ftdc2-201</td>\n",
       "      <td>438</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>5426</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>5430</td>\n",
       "      <td>367</td>\n",
       "      <td>945</td>\n",
       "      <td>5426</td>\n",
       "      <td>3</td>\n",
       "      <td>5429</td>\n",
       "      <td>0.999447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Padi6-201_222</th>\n",
       "      <td>Padi6-201</td>\n",
       "      <td>222</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>5223</td>\n",
       "      <td>5226</td>\n",
       "      <td>64</td>\n",
       "      <td>2112</td>\n",
       "      <td>5223</td>\n",
       "      <td>0</td>\n",
       "      <td>5223</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zp3-201_141</th>\n",
       "      <td>Zp3-201</td>\n",
       "      <td>141</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5031</td>\n",
       "      <td>0</td>\n",
       "      <td>5032</td>\n",
       "      <td>36</td>\n",
       "      <td>1310</td>\n",
       "      <td>5031</td>\n",
       "      <td>0</td>\n",
       "      <td>5031</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ltbp4-201_4658</th>\n",
       "      <td>Ltbp4-201</td>\n",
       "      <td>4658</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>126</td>\n",
       "      <td>5126</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Shkbp1-201_861</th>\n",
       "      <td>Shkbp1-201</td>\n",
       "      <td>861</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>2178</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Shkbp1-201_1359</th>\n",
       "      <td>Shkbp1-201</td>\n",
       "      <td>1359</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>2178</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Shkbp1-201_1992</th>\n",
       "      <td>Shkbp1-201</td>\n",
       "      <td>1992</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>2178</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mid1-205_1745</th>\n",
       "      <td>Mid1-205</td>\n",
       "      <td>1745</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>345</td>\n",
       "      <td>2387</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85339 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 transcript  position REF ALT  A     C     G     T  snp_total  \\\n",
       "trans_pos                                                                       \n",
       "H1f8-201_69        H1f8-201        69   T   C  1     3     0  6102       6106   \n",
       "Ooep-201_257       Ooep-201       257   T   C  0     2     0  6037       6039   \n",
       "Ftdc2-201_438     Ftdc2-201       438   C   T  1  5426     0     3       5430   \n",
       "Padi6-201_222     Padi6-201       222   T   C  0     0     3  5223       5226   \n",
       "Zp3-201_141         Zp3-201       141   G   T  1     0  5031     0       5032   \n",
       "...                     ...       ...  ..  .. ..   ...   ...   ...        ...   \n",
       "Ltbp4-201_4658    Ltbp4-201      4658   A   G  0     0     0     0          0   \n",
       "Shkbp1-201_861   Shkbp1-201       861   G   A  0     0     0     0          0   \n",
       "Shkbp1-201_1359  Shkbp1-201      1359   G   T  0     0     0     0          0   \n",
       "Shkbp1-201_1992  Shkbp1-201      1992   G   A  0     0     0     0          0   \n",
       "Mid1-205_1745      Mid1-205      1745   A   G  0     0     0     0          0   \n",
       "\n",
       "                 CDS_start  CDS_stop  maternal_count  paternal_count  \\\n",
       "trans_pos                                                              \n",
       "H1f8-201_69             46       960            6102               3   \n",
       "Ooep-201_257           198       692            6037               2   \n",
       "Ftdc2-201_438          367       945            5426               3   \n",
       "Padi6-201_222           64      2112            5223               0   \n",
       "Zp3-201_141             36      1310            5031               0   \n",
       "...                    ...       ...             ...             ...   \n",
       "Ltbp4-201_4658         126      5126               0               0   \n",
       "Shkbp1-201_861          64      2178               0               0   \n",
       "Shkbp1-201_1359         64      2178               0               0   \n",
       "Shkbp1-201_1992         64      2178               0               0   \n",
       "Mid1-205_1745          345      2387               0               0   \n",
       "\n",
       "                 mat_pat_total  maternal_ratio  \n",
       "trans_pos                                       \n",
       "H1f8-201_69               6105        0.999509  \n",
       "Ooep-201_257              6039        0.999669  \n",
       "Ftdc2-201_438             5429        0.999447  \n",
       "Padi6-201_222             5223        1.000000  \n",
       "Zp3-201_141               5031        1.000000  \n",
       "...                        ...             ...  \n",
       "Ltbp4-201_4658               1        0.000000  \n",
       "Shkbp1-201_861               1        0.000000  \n",
       "Shkbp1-201_1359              1        0.000000  \n",
       "Shkbp1-201_1992              1        0.000000  \n",
       "Mid1-205_1745                1        0.000000  \n",
       "\n",
       "[85339 rows x 15 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_maternal_ratio( rnaseq_snp_total_one_cell )\n",
    "find_maternal_ratio( rnaseq_snp_total_two_cell ) \n",
    "find_maternal_ratio( rnaseq_snp_total_four_cell ) \n",
    "find_maternal_ratio( rnaseq_snp_total_eight_cell ) \n",
    "\n",
    "find_maternal_ratio( rnaseq_snp_total_mii )\n",
    "find_maternal_ratio( rnaseq_snp_total_gv )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dfdbadc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6c089db8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>snp_total</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>CDS_stop</th>\n",
       "      <th>maternal_count</th>\n",
       "      <th>paternal_count</th>\n",
       "      <th>mat_pat_total</th>\n",
       "      <th>maternal_ratio</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trans_pos</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cacna1b-201_6873</th>\n",
       "      <td>Cacna1b-201</td>\n",
       "      <td>6873</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>1</td>\n",
       "      <td>6984</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ankrd26-201_1877</th>\n",
       "      <td>Ankrd26-201</td>\n",
       "      <td>1877</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>210</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>210</td>\n",
       "      <td>170</td>\n",
       "      <td>5215</td>\n",
       "      <td>210</td>\n",
       "      <td>0</td>\n",
       "      <td>210</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Npm1-201_731</th>\n",
       "      <td>Npm1-201</td>\n",
       "      <td>731</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>112</td>\n",
       "      <td>228</td>\n",
       "      <td>1106</td>\n",
       "      <td>63</td>\n",
       "      <td>49</td>\n",
       "      <td>112</td>\n",
       "      <td>0.562500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zfp113-201_702</th>\n",
       "      <td>Zfp113-201</td>\n",
       "      <td>702</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>84</td>\n",
       "      <td>99</td>\n",
       "      <td>186</td>\n",
       "      <td>1505</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eef2-201_1179</th>\n",
       "      <td>Eef2-201</td>\n",
       "      <td>1179</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>37</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>97</td>\n",
       "      <td>2673</td>\n",
       "      <td>25</td>\n",
       "      <td>37</td>\n",
       "      <td>62</td>\n",
       "      <td>0.403226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sorl1-201_5604</th>\n",
       "      <td>Sorl1-201</td>\n",
       "      <td>5604</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>185</td>\n",
       "      <td>6832</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sc5d-201_530</th>\n",
       "      <td>Sc5d-201</td>\n",
       "      <td>530</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>141</td>\n",
       "      <td>1040</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sc5d-201_790</th>\n",
       "      <td>Sc5d-201</td>\n",
       "      <td>790</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>141</td>\n",
       "      <td>1040</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sc5d-201_824</th>\n",
       "      <td>Sc5d-201</td>\n",
       "      <td>824</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>141</td>\n",
       "      <td>1040</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mid1-205_1745</th>\n",
       "      <td>Mid1-205</td>\n",
       "      <td>1745</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>345</td>\n",
       "      <td>2387</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85339 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   transcript  position REF ALT    A    C  G   T  snp_total  \\\n",
       "trans_pos                                                                     \n",
       "Cacna1b-201_6873  Cacna1b-201      6873   A   G  213    0  0   0        213   \n",
       "Ankrd26-201_1877  Ankrd26-201      1877   C   A    0  210  0   0        210   \n",
       "Npm1-201_731         Npm1-201       731   C   T    0   63  0  49        112   \n",
       "Zfp113-201_702     Zfp113-201       702   A   G    0   13  2  84         99   \n",
       "Eef2-201_1179        Eef2-201      1179   C   A   37   25  2   0         64   \n",
       "...                       ...       ...  ..  ..  ...  ... ..  ..        ...   \n",
       "Sorl1-201_5604      Sorl1-201      5604   G   A    0    0  0   0          0   \n",
       "Sc5d-201_530         Sc5d-201       530   A   G    0    0  0   0          0   \n",
       "Sc5d-201_790         Sc5d-201       790   A   G    0    0  0   0          0   \n",
       "Sc5d-201_824         Sc5d-201       824   C   T    0    0  0   0          0   \n",
       "Mid1-205_1745        Mid1-205      1745   A   G    0    0  0   0          0   \n",
       "\n",
       "                  CDS_start  CDS_stop  maternal_count  paternal_count  \\\n",
       "trans_pos                                                               \n",
       "Cacna1b-201_6873          1      6984             213               0   \n",
       "Ankrd26-201_1877        170      5215             210               0   \n",
       "Npm1-201_731            228      1106              63              49   \n",
       "Zfp113-201_702          186      1505               0               2   \n",
       "Eef2-201_1179            97      2673              25              37   \n",
       "...                     ...       ...             ...             ...   \n",
       "Sorl1-201_5604          185      6832               0               0   \n",
       "Sc5d-201_530            141      1040               0               0   \n",
       "Sc5d-201_790            141      1040               0               0   \n",
       "Sc5d-201_824            141      1040               0               0   \n",
       "Mid1-205_1745           345      2387               0               0   \n",
       "\n",
       "                  mat_pat_total  maternal_ratio  \n",
       "trans_pos                                        \n",
       "Cacna1b-201_6873            213        1.000000  \n",
       "Ankrd26-201_1877            210        1.000000  \n",
       "Npm1-201_731                112        0.562500  \n",
       "Zfp113-201_702                2        0.000000  \n",
       "Eef2-201_1179                62        0.403226  \n",
       "...                         ...             ...  \n",
       "Sorl1-201_5604                1        0.000000  \n",
       "Sc5d-201_530                  1        0.000000  \n",
       "Sc5d-201_790                  1        0.000000  \n",
       "Sc5d-201_824                  1        0.000000  \n",
       "Mid1-205_1745                 1        0.000000  \n",
       "\n",
       "[85339 rows x 15 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "riboseq_snp_total_four_cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e3661da9",
   "metadata": {},
   "outputs": [],
   "source": [
    "proptest = robjects.r('prop.test')\n",
    "\n",
    "def perform_prop_test(df_1, df_2, error_threshold = 0.01, filter_small_counts = True):\n",
    "    \"\"\"\n",
    "    Decide whether paternal  ratios differ between df_1 and df_2\n",
    "    \n",
    "    Insufficient counts are marked with the p-value 10.\n",
    "    \n",
    "    If the difference between the proportionsare less than the error threshold, \n",
    "    they are marked with the p-value 5.\n",
    "    \"\"\"\n",
    "    \n",
    "    pvalues            = []\n",
    "    conf_midpoints     = []\n",
    "    diff_is_sufficient = []\n",
    "\n",
    "    merged_df = df_1.merge(df_2, on = \"trans_pos\")\n",
    "    \n",
    "    \n",
    "    for i, e in merged_df.iterrows():\n",
    "        #print(i, e)\n",
    "        if filter_small_counts:\n",
    "            \n",
    "            # Mark insufficient counts with p-val 10 and exclude them from the test\n",
    "            if (e[\"maternal_count_x\"]+ e[\"paternal_count_x\"]) < 10  or\\\n",
    "               (e[\"maternal_count_y\"] + e[\"paternal_count_y\"]) < 10 :\n",
    "                pvalues.append(10)\n",
    "                conf_midpoints.append(10)\n",
    "                diff_is_sufficient.append(0)\n",
    "                continue\n",
    "\n",
    "\n",
    "            if e[\"maternal_count_x\"]+ e[\"maternal_count_y\"] == 0 or\\\n",
    "               e[\"paternal_count_x\"] + e[\"paternal_count_y\"] == 0 :\n",
    "\n",
    "                #print(\"detected zero at\", i)\n",
    "                pvalues.append(10)\n",
    "                conf_midpoints.append(10)\n",
    "                diff_is_sufficient.append(0)\n",
    "                continue\n",
    "\n",
    "            if e[\"paternal_count_x\"] <=2 and e[\"paternal_count_y\"] <= 2:\n",
    "                pvalues.append(10)\n",
    "                conf_midpoints.append(10)\n",
    "                diff_is_sufficient.append(0)\n",
    "                continue\n",
    "        # A valuye 0 means the difference between the two ratios is \n",
    "        # very small\n",
    "        # 1 means, the diffrences is NOT very small.\n",
    "        # We will use it for filtering purposes\n",
    "        if e[\"maternal_count_x\"]+ e[\"paternal_count_x\"] == 0 or\\\n",
    "               e[\"maternal_count_y\"] + e[\"paternal_count_y\"] == 0 :\n",
    "\n",
    "                #print(\"detected zero at\", i)\n",
    "                pvalues.append(100)\n",
    "                conf_midpoints.append(100)\n",
    "                diff_is_sufficient.append(0)\n",
    "                continue\n",
    "        \n",
    "        r_paternal_counts = robjects.IntVector([ e[\"paternal_count_x\"], e[\"paternal_count_y\"] ])\n",
    "        r_total_counts    = robjects.IntVector([ e[\"paternal_count_x\"] + e[\"maternal_count_x\"], \n",
    "                                                 e[\"paternal_count_y\"] + e[\"maternal_count_y\"] ])\n",
    "        \n",
    "        proptest_result = proptest(r_paternal_counts, r_total_counts)\n",
    "        result_dict     = dict(proptest_result.items())\n",
    "        conf_int        = result_dict[\"conf.int\"]\n",
    "        mid_point_conf  = (conf_int[0] + conf_int[1]) / 2\n",
    "        \n",
    "\n",
    "        \n",
    "        if conf_int[1] < 0 and error_threshold > ((-1) * conf_int[1]):\n",
    "            diff_is_sufficient.append(0)\n",
    "        elif conf_int[1] >=0 and (error_threshold > conf_int[0]):\n",
    "            diff_is_sufficient.append(0)\n",
    "        else:\n",
    "            diff_is_sufficient.append(1)\n",
    "        \n",
    "        this_p_val = proptest_result[2][0]\n",
    "        pvalues.append(this_p_val)\n",
    "        \n",
    "        conf_midpoints.append(mid_point_conf)  \n",
    "        \n",
    "    result_dict = { #\"trans_pos\" : merged_df[\"trans_pos\"],\n",
    "                    \"p_val\":               pvalues,\n",
    "                    \"conf_mid\":            conf_midpoints,\n",
    "                    \"maternal_x\":          merged_df[\"maternal_count_x\"],\n",
    "                    \"paternal_x\":          merged_df[\"paternal_count_x\"],\n",
    "                    \"maternal_y\":          merged_df[\"maternal_count_y\"],\n",
    "                    \"paternal_y\":          merged_df[\"paternal_count_y\"],\n",
    "                    \"diff_is_sufficient\" : diff_is_sufficient,  }\n",
    "    \n",
    "    \n",
    "    \n",
    "    result_df =  pd.DataFrame.from_dict( result_dict )\n",
    "    \n",
    "    return result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33a3fdc2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7602c5b3",
   "metadata": {},
   "source": [
    "## Transcript Level SNPs\n",
    "\n",
    "We aggregate paternal & paternal counts of transcripts. Next, we determine transcripts chaning behavior ( in terms of paternal / maternal ratio). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a5c7f300",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transcript_level_counts(snp_df):\n",
    "    result_dict = {}\n",
    "    \n",
    "    for g in snp_df[\"transcript\"]:\n",
    "        result_dict[g] = { \"transcript\": g, \"maternal_count\": 0, \"paternal_count\": 0 }\n",
    "        #result_dict[g] = { \"maternal_count\": 0, \"paternal_count\": 0 }\n",
    "        \n",
    "    for i, e in snp_df.iterrows():\n",
    "        \n",
    "        g = e[\"transcript\"]\n",
    "        \n",
    "        result_dict[g][\"maternal_count\"] += e[\"maternal_count\"] \n",
    "        result_dict[g][\"paternal_count\"] += e[\"paternal_count\"]\n",
    "        \n",
    "    result_df = pd.DataFrame.from_dict( result_dict ).transpose()\n",
    "    result_df.index.rename(\"trans_pos\", inplace = True)\n",
    "\n",
    "    \n",
    "    return result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ace63a7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "5d231e49",
   "metadata": {},
   "outputs": [],
   "source": [
    "riboseq_t_level_snp_total_gv         = get_transcript_level_counts(riboseq_snp_total_gv)\n",
    "riboseq_t_level_snp_total_mii        = get_transcript_level_counts(riboseq_snp_total_mii )\n",
    "riboseq_t_level_snp_total_one_cell   = get_transcript_level_counts(riboseq_snp_total_one_cell )\n",
    "riboseq_t_level_snp_total_two_cell   = get_transcript_level_counts(riboseq_snp_total_two_cell )\n",
    "riboseq_t_level_snp_total_four_cell  = get_transcript_level_counts(riboseq_snp_total_four_cell )\n",
    "riboseq_t_level_snp_total_eight_cell = get_transcript_level_counts(riboseq_snp_total_eight_cell )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0ae3bef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnaseq_t_level_snp_total_mii        = get_transcript_level_counts(rnaseq_snp_total_mii )\n",
    "rnaseq_t_level_snp_total_gv         = get_transcript_level_counts(rnaseq_snp_total_gv )\n",
    "rnaseq_t_level_snp_total_one_cell   = get_transcript_level_counts(rnaseq_snp_total_one_cell )\n",
    "rnaseq_t_level_snp_total_two_cell   = get_transcript_level_counts(rnaseq_snp_total_two_cell )\n",
    "rnaseq_t_level_snp_total_four_cell  = get_transcript_level_counts(rnaseq_snp_total_four_cell )\n",
    "rnaseq_t_level_snp_total_eight_cell = get_transcript_level_counts(rnaseq_snp_total_eight_cell )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daf0961e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "72d109ca",
   "metadata": {},
   "source": [
    "### Perform PropTest\n",
    "\n",
    "We will use the results of prop-test for filtering purposes. \n",
    "\n",
    "Namely, the genes that were used in the prop-test are the **\"expressed\"** genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "23b148fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "proptest_t_level_gv_ribo_rna         = perform_prop_test( riboseq_t_level_snp_total_gv, \n",
    "                                                          rnaseq_t_level_snp_total_gv,\n",
    "                                                          error_threshold = PROPTEST_ERROR_THRESHOLD)\n",
    "\n",
    "proptest_t_level_mii_ribo_rna        = perform_prop_test( riboseq_t_level_snp_total_mii, \n",
    "                                                          rnaseq_t_level_snp_total_mii,\n",
    "                                                          error_threshold = PROPTEST_ERROR_THRESHOLD)\n",
    "\n",
    "proptest_t_level_one_cell_ribo_rna   = perform_prop_test( riboseq_t_level_snp_total_one_cell, \n",
    "                                                          rnaseq_t_level_snp_total_one_cell,\n",
    "                                                          error_threshold = PROPTEST_ERROR_THRESHOLD)\n",
    "\n",
    "proptest_t_level_two_cell_ribo_rna   = perform_prop_test( riboseq_t_level_snp_total_two_cell, \n",
    "                                                          rnaseq_t_level_snp_total_two_cell,\n",
    "                                                          error_threshold = PROPTEST_ERROR_THRESHOLD)\n",
    "\n",
    "proptest_t_level_four_cell_ribo_rna  = perform_prop_test( riboseq_t_level_snp_total_four_cell, \n",
    "                                                          rnaseq_t_level_snp_total_four_cell,\n",
    "                                                          error_threshold = PROPTEST_ERROR_THRESHOLD)\n",
    "\n",
    "proptest_t_level_eight_cell_ribo_rna = perform_prop_test( riboseq_t_level_snp_total_eight_cell, \n",
    "                                                          rnaseq_t_level_snp_total_eight_cell,\n",
    "                                                          error_threshold = PROPTEST_ERROR_THRESHOLD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67783ce0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7212a266",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p_val</th>\n",
       "      <th>conf_mid</th>\n",
       "      <th>maternal_x</th>\n",
       "      <th>paternal_x</th>\n",
       "      <th>maternal_y</th>\n",
       "      <th>paternal_y</th>\n",
       "      <th>diff_is_sufficient</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trans_pos</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cacna1b-201</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ankrd26-201</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>212</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Npm1-201</th>\n",
       "      <td>0.123346</td>\n",
       "      <td>-0.073905</td>\n",
       "      <td>76</td>\n",
       "      <td>61</td>\n",
       "      <td>527</td>\n",
       "      <td>569</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zfp113-201</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eef2-201</th>\n",
       "      <td>0.014625</td>\n",
       "      <td>0.146279</td>\n",
       "      <td>73</td>\n",
       "      <td>105</td>\n",
       "      <td>74</td>\n",
       "      <td>59</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Grik4-202</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Clmp-201</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ubash3b-201</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sc5d-201</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tbcel-201</th>\n",
       "      <td>10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14042 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 p_val   conf_mid maternal_x paternal_x maternal_y paternal_y  \\\n",
       "trans_pos                                                                       \n",
       "Cacna1b-201  10.000000  10.000000        213          0          2          0   \n",
       "Ankrd26-201  10.000000  10.000000        212          0          3          0   \n",
       "Npm1-201      0.123346  -0.073905         76         61        527        569   \n",
       "Zfp113-201   10.000000  10.000000          0          2          3          2   \n",
       "Eef2-201      0.014625   0.146279         73        105         74         59   \n",
       "...                ...        ...        ...        ...        ...        ...   \n",
       "Grik4-202    10.000000  10.000000          0          0          0          0   \n",
       "Clmp-201     10.000000  10.000000          0          0          1          0   \n",
       "Ubash3b-201  10.000000  10.000000          0          0          1          0   \n",
       "Sc5d-201     10.000000  10.000000          0          0          2          2   \n",
       "Tbcel-201    10.000000  10.000000          0          0          0          0   \n",
       "\n",
       "             diff_is_sufficient  \n",
       "trans_pos                        \n",
       "Cacna1b-201                   0  \n",
       "Ankrd26-201                   0  \n",
       "Npm1-201                      0  \n",
       "Zfp113-201                    0  \n",
       "Eef2-201                      0  \n",
       "...                         ...  \n",
       "Grik4-202                     0  \n",
       "Clmp-201                      0  \n",
       "Ubash3b-201                   0  \n",
       "Sc5d-201                      0  \n",
       "Tbcel-201                     0  \n",
       "\n",
       "[14042 rows x 7 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proptest_t_level_four_cell_ribo_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "10539f1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>maternal_x</th>\n",
       "      <th>paternal_x</th>\n",
       "      <th>maternal_y</th>\n",
       "      <th>paternal_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Npm1</th>\n",
       "      <td>76</td>\n",
       "      <td>61</td>\n",
       "      <td>527</td>\n",
       "      <td>569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eef2</th>\n",
       "      <td>73</td>\n",
       "      <td>105</td>\n",
       "      <td>74</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Timd2</th>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>198</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ddx54</th>\n",
       "      <td>12</td>\n",
       "      <td>17</td>\n",
       "      <td>34</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rps11</th>\n",
       "      <td>16</td>\n",
       "      <td>47</td>\n",
       "      <td>1435</td>\n",
       "      <td>1610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Swt1</th>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>50</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ddias</th>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>31</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hmmr</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>21</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sgo2a</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Urb1</th>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>362 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      maternal_x paternal_x maternal_y paternal_y\n",
       "gene                                             \n",
       "Npm1          76         61        527        569\n",
       "Eef2          73        105         74         59\n",
       "Timd2         45         45        198        179\n",
       "Ddx54         12         17         34         16\n",
       "Rps11         16         47       1435       1610\n",
       "...          ...        ...        ...        ...\n",
       "Swt1           8          5         50         30\n",
       "Ddias          5          6         31         63\n",
       "Hmmr          10          6         21         20\n",
       "Sgo2a         10          0         11         12\n",
       "Urb1           3          7         12         11\n",
       "\n",
       "[362 rows x 4 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fourcell_expressed_genes_df = (proptest_t_level_four_cell_ribo_rna[  proptest_t_level_four_cell_ribo_rna[\"p_val\"] <=1 ])\\\n",
    "    [ [\"maternal_x\", \"paternal_x\", \"maternal_y\", \"paternal_y\"]]\n",
    "\n",
    "gene_names_4cell = list(map(lambda x: x.split(\"-\")[0], fourcell_expressed_genes_df.index)  )\n",
    "\n",
    "fourcell_expressed_genes_df.set_index( pd.Index(gene_names_4cell).rename(\"gene\") , inplace = True)\n",
    "\n",
    "fourcell_expressed_genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9903eb57",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "45e324c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>maternal_x</th>\n",
       "      <th>paternal_x</th>\n",
       "      <th>maternal_y</th>\n",
       "      <th>paternal_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Npm1</th>\n",
       "      <td>142</td>\n",
       "      <td>158</td>\n",
       "      <td>1381</td>\n",
       "      <td>1313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hspa8</th>\n",
       "      <td>158</td>\n",
       "      <td>139</td>\n",
       "      <td>1028</td>\n",
       "      <td>1170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eef2</th>\n",
       "      <td>192</td>\n",
       "      <td>159</td>\n",
       "      <td>388</td>\n",
       "      <td>392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Timd2</th>\n",
       "      <td>186</td>\n",
       "      <td>130</td>\n",
       "      <td>744</td>\n",
       "      <td>572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rpl38</th>\n",
       "      <td>44</td>\n",
       "      <td>67</td>\n",
       "      <td>1992</td>\n",
       "      <td>1634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nol8</th>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>50</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Arhgap21</th>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>27</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wdfy3</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nup205</th>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>34</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zfc3h1</th>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>850 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         maternal_x paternal_x maternal_y paternal_y\n",
       "gene                                                \n",
       "Npm1            142        158       1381       1313\n",
       "Hspa8           158        139       1028       1170\n",
       "Eef2            192        159        388        392\n",
       "Timd2           186        130        744        572\n",
       "Rpl38            44         67       1992       1634\n",
       "...             ...        ...        ...        ...\n",
       "Nol8              7          5         50         13\n",
       "Arhgap21          2         11         27         31\n",
       "Wdfy3             4          6          8          3\n",
       "Nup205            7          4         34         50\n",
       "Zfc3h1            2          8         10          3\n",
       "\n",
       "[850 rows x 4 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eightcell_expressed_genes_df = (proptest_t_level_eight_cell_ribo_rna[  proptest_t_level_eight_cell_ribo_rna[\"p_val\"] <=1 ])\\\n",
    "    [ [\"maternal_x\", \"paternal_x\", \"maternal_y\", \"paternal_y\"]]\n",
    "\n",
    "gene_names_8cell = list(map(lambda x: x.split(\"-\")[0], eightcell_expressed_genes_df.index)  )\n",
    "\n",
    "eightcell_expressed_genes_df.set_index( pd.Index(gene_names_8cell).rename(\"gene\") , inplace = True)\n",
    "\n",
    "eightcell_expressed_genes_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4ff8919e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fourcell_expressed_genes_df.to_csv(\"fourcell_snp_counts.csv.gz\")\n",
    "eightcell_expressed_genes_df.to_csv(\"eightcell_snp_counts.csv.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "eb3e5b6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fourcell_all_snp_counts_df  = proptest_t_level_four_cell_ribo_rna[ [\"maternal_x\", \"paternal_x\", \"maternal_y\", \"paternal_y\"] ]\n",
    "\n",
    "eightcell_all_snp_counts_df = proptest_t_level_eight_cell_ribo_rna[ [\"maternal_x\", \"paternal_x\", \"maternal_y\", \"paternal_y\"] ]\n",
    "\n",
    "gene_names_4cell_all = list(map(lambda x: x.split(\"-\")[0], fourcell_all_snp_counts_df.index)  )\n",
    "gene_names_8cell_all = list(map(lambda x: x.split(\"-\")[0], eightcell_all_snp_counts_df.index)  )\n",
    "\n",
    "fourcell_all_snp_counts_df.set_index( pd.Index(gene_names_4cell_all).rename(\"gene\"), inplace = True )\n",
    "eightcell_all_snp_counts_df.set_index( pd.Index(gene_names_8cell_all).rename(\"gene\"), inplace = True )\n",
    "\n",
    "fourcell_all_snp_counts_df.to_csv(\"fourcell_all_snp_counts.csv.gz\")\n",
    "eightcell_all_snp_counts_df.to_csv(\"eightcell_all_snp_counts.csv.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "697556d0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
