---
title: "Imprinted Genes"
output: html_document
---

We determine imprinted genes first. Then we compare the translation efficiency of imprinted genes to those of background genes.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r include,echo=FALSE,warning=FALSE}
library(cowplot)
library(ggplot2)
library(dplyr)
library(Cairo)
library(edgeR)
library(MatchIt)
library(Seurat)
```

```{r constants,echo=FALSE,warning=FALSE}
CONSTANT_ADDITION = 1
```


```{r setwd,echo=FALSE,warning=FALSE}

setwd("~/projects/mice_cell_development/mouse-itp/imprinted_genes/")
#setwd("/data/projects/mice_cell_development/mouse-itp/imprinted_genes/")
```

```{r filepaths,echo=FALSE,warning=FALSE}
published_imprinted_gene_list_file = "./imprinted_gene_list_41467_2021_23510_MOESM4_ESM.csv"


# Another imprinted gene list from
# https://www.cell.com/cell/fulltext/S0092-8674(19)30106-0?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867419301060%3Fshowall%3Dtrue#%20
tucci_et_al_imprinted_gene_list_file = "./tucci_et_al_2019_mouse_imprinted_genes.csv"

four_cell_all_snp_counts_file  = "./fourcell_all_snp_counts.csv.gz"
eight_cell_all_snp_counts_file = "./eightcell_all_snp_counts.csv.gz" 

ribo_rna_counts_file = "../cds_counts.csv.gz"

chrx_genes_file = "./chrx_genes.tsv"

#region_lengths_file = "../../mouse_itp_reference/transcriptome/appris_mouse_v2_filtered_regions.bed"

region_lengths_file = "../appris_mouse_v2_filtered_regions.bed"
 
detailed_snp_file = "../snp/notebooks/snp_dataframes/rnaseq_detailed_snps.csv.gz"
GSE80810_allelic_ratio_file = "./GSE80810_allelic_ratio.csv"
```

```{r region_lengths,echo=FALSE,warning=FALSE}

# Here we obtain the cds lengths

region_lengths_pre = read.csv(region_lengths_file, sep = "\t", header = FALSE)

colnames(region_lengths_pre) = c("gene", "start", "end", "region", "score", "strand")

get_gene_name = function(long_name){
   return(strsplit(long_name,  split= "|", fixed = TRUE) )
}

region_lengths_pre$gene = unlist( lapply( get_gene_name(region_lengths_pre$gene)  , "[[", 6)  )

region_lengths_pre      = region_lengths_pre[ region_lengths_pre$region == "CDS" ,]

cds_lengths             = region_lengths_pre$end - region_lengths_pre$start

region_lengths          = data.frame(gene = region_lengths_pre$gene, cds_length = cds_lengths)

```

## Publisehd Imprinted Genes

[Santini et al ](https://www.nature.com/articles/s41467-021-23510-4#MOESM4) studies mouse embryos and provide a list of imprinted genes.
However their list doesn't indicate any direction (maternal or paternal).

We define **paternal ratio** as

$$
\frac{ \mbox{PATERNAL READS + CONSTANT} }{ \mbox{MATERNAL READS + PATERNAL READS + 2*CONSTANT}  }
$$

where constant is a positive integer. Namely, we are taking the ratio of paternal reads to the total reads after adding a constant to both maternal and paternal reads (hence the coefficient 2 in the denominatior).

```{r echo=FALSE,warning=FALSE}
four_cell_all_snp_counts  = read.csv(four_cell_all_snp_counts_file)
eight_cell_all_snp_counts = read.csv(eight_cell_all_snp_counts_file)

four_cell_all_snp_counts$paternal_ratio  = 
  (four_cell_all_snp_counts$paternal_y + CONSTANT_ADDITION  ) / (four_cell_all_snp_counts$maternal_y + four_cell_all_snp_counts$paternal_y + 2* CONSTANT_ADDITION)

eight_cell_all_snp_counts$paternal_ratio = 
  (eight_cell_all_snp_counts$paternal_y + CONSTANT_ADDITION) / (eight_cell_all_snp_counts$maternal_y + eight_cell_all_snp_counts$paternal_y + 2*CONSTANT_ADDITION)

published_nature_gen_df = read.csv(published_imprinted_gene_list_file)

published_nature_gen_df
```

[Tucci et. al.](https://www.cell.com/cell/fulltext/S0092-8674(19)30106-0?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867419301060%3Fshowall%3Dtrue#%20)
however, indicates the direction, though their list is less comprehensive.

```{r echo=FALSE,warning=FALSE}

tucci_et_al_imprinted_gene_df_pre = read.csv(tucci_et_al_imprinted_gene_list_file)

# get rid of the isoform-psecific entries
tucci_et_al_imprinted_gene_df = tucci_et_al_imprinted_gene_df_pre[tucci_et_al_imprinted_gene_df_pre$ExpressedAllele %in% c("P", "M") ,]

tucci_et_al_imprinted_gene_df
```

## Distribution of SNP RNA Counts

We sum the reads overlapping with SNPs for each gene. Then we aggregate the read counts across replicates.
So, we obtain a single count for each gene in each stage of the embryo.
We want to be confident with our imprinted gene picks. So we filter out genes having less than 10 counts.

Please note that these are different (less) than regular RNA-Seq counts as we use the reads overlapping with SNPs only.

On the other hand, for TE calculations, we use all the reads. In either case, we user the reads mapping to the CDS region only.

```{r echo=FALSE,warning=FALSE}

fourcel_rna_total = four_cell_all_snp_counts$paternal_y + four_cell_all_snp_counts$maternal_y

eight_cell_rna_total = eight_cell_all_snp_counts$paternal_y + eight_cell_all_snp_counts$maternal_y

hist( log2( fourcel_rna_total[fourcel_rna_total>=0 ] ) )

hist( log2( eight_cell_rna_total[ eight_cell_rna_total >= 0 ] ) )

hist( log2( fourcel_rna_total[fourcel_rna_total>=10 ] ) )

hist( log2( eight_cell_rna_total[ eight_cell_rna_total >= 10 ] ) )

```
We pick genes with total count of at least 10.

```{r echo=FALSE,warning=FALSE}

four_cell_snp_filtered_pre  = four_cell_all_snp_counts[ (four_cell_all_snp_counts$paternal_y + four_cell_all_snp_counts$maternal_y) >= 10, ]

eight_cell_snp_filtered_pre = eight_cell_all_snp_counts[ (eight_cell_all_snp_counts$paternal_y + eight_cell_all_snp_counts$maternal_y) >= 10,]

print("Four cell SNP (>=10)  genes:")
print(dim(four_cell_snp_filtered_pre)[1]  )
print("Eight cell SNP (>=10)  genes:")
print(dim(eight_cell_snp_filtered_pre)[1])

hist( ( four_cell_snp_filtered_pre$paternal_ratio )  )
hist( ( eight_cell_snp_filtered_pre$paternal_ratio )  )
```
### Genes from the X Chromosome

Note that we are excluding genes coming from the X chromosome

```{r echo=FALSE,warning=FALSE}

chrx_df = read.csv(chrx_genes_file, sep = "\t", header = F)

chrx_genes = chrx_df[, 1]

chrx_eight_cell_intersection_count = length( intersect(eight_cell_snp_filtered_pre$gene, chrx_genes) )
chrx_four_cell_intersection_count  = length( intersect(four_cell_snp_filtered_pre$gene, chrx_genes) )

print("Genes of the chrX:")
print( c("Intersection with the 4cell expressed genes is", chrx_four_cell_intersection_count)  )
print( c("Intersection with the 8cell expressed genes is", chrx_eight_cell_intersection_count)  )


eight_cell_snp_filtered = eight_cell_snp_filtered_pre[!(eight_cell_snp_filtered_pre$gene %in% chrx_genes ) , ]
four_cell_snp_filtered  = four_cell_snp_filtered_pre[ !(four_cell_snp_filtered_pre$gene %in% chrx_genes), ]


```

## A simple way of finding Imprinted Genes

After combining the reads of the replicates, we initially determine the maternal and paternal reads from the 0.5% tails of the paternal ratio distribution.
Later we control for replicate-to-replicate variability and refine this list.

```{r echo=FALSE,warning=FALSE}

TAIL_SIZE   = 0.05
quant_probs = probs = c(TAIL_SIZE, 1 - TAIL_SIZE)

fourcell_quantiles = quantile( four_cell_snp_filtered$paternal_ratio , quant_probs  )

eightcell_quantiles = quantile( eight_cell_snp_filtered$paternal_ratio, quant_probs  )

print("Quantiles are (4 and 8 cells)")
print(fourcell_quantiles)
print(eightcell_quantiles)

fourcell_paternal_genes = four_cell_snp_filtered[ four_cell_snp_filtered$paternal_ratio >= fourcell_quantiles[2] , ]$gene
fourcell_maternal_genes = four_cell_snp_filtered[ four_cell_snp_filtered$paternal_ratio <= fourcell_quantiles[1] , ]$gene

print("Four cell paternal and maternal genes are:")
print( c(length(fourcell_paternal_genes), length(fourcell_maternal_genes) )  )


eightcell_paternal_genes = eight_cell_snp_filtered[ eight_cell_snp_filtered$paternal_ratio >= eightcell_quantiles[2] , ]$gene
eightcell_maternal_genes = eight_cell_snp_filtered[ eight_cell_snp_filtered$paternal_ratio <= eightcell_quantiles[1] , ]$gene

print("Eight cell paternal and maternal genes are:")
print( c(length(eightcell_paternal_genes), length(eightcell_maternal_genes) )  )

```





## Comparison of Our Imprinted Genes to the Published Ones

There are similarities and differences worth noting with the papers we compare our data to.

Our data come from 4 or 8 cell stages of the embryo whereas published data come from 32-cell stage. 
It is already known that imprinted genes show different patterns during the stages of the development. Hence an overlap is not expected.

From [Santini et. al.](https://www.nature.com/articles/s41467-021-23510-4#Abs1):
**"Hence, statistically significant parent-of-origin-specific gene expression was detected in blastocysts for only a quarter of all published imprinted genes, indicating a strong impact of tissue and cell type in defining imprinting patterns during development."**

```{r echo=FALSE,warning=FALSE}

nature_intersection_df = merge(published_nature_gen_df, region_lengths, by, by.x = "Gene.symbol", by.y = "gene" )
published_nature_gen_df
nature_intersection_df

fourcell_common_genes_with_nature = intersect( nature_intersection_df$Gene.symbol, union( fourcell_maternal_genes, fourcell_paternal_genes  )   )

print("4 Cell: Common imprinted genes with the Santini et. al. paper:")
print(fourcell_common_genes_with_nature)

print("8 Cell: Common imprinted genes with the Santini et. al. paper:")
eightcell_common_genes_with_nature = intersect( nature_intersection_df$Gene.symbol, union( eightcell_maternal_genes, eightcell_paternal_genes  )   )

eightcell_common_genes_with_nature

#published_nature_gen_df
#fourcell_tucci_merged = merge(four_cell_snp_filtered, tucci_et_al_imprinted_gene_df, by.x = "gene", by.y = "Gene")
#fourcell_tucci_merged$gene

#eightcell_tucci_merged = merge(eight_cell_snp_filtered, tucci_et_al_imprinted_gene_df, by.x = "gene", by.y = "Gene")
#eightcell_tucci_merged$gene
```


```{r echo=FALSE,warning=FALSE}

a = merge(published_nature_gen_df, eight_cell_snp_filtered, by.x = "Gene.symbol", by.y = "gene")
a

```


## Translation Efficiency of Imprinted Genes 

```{r echo=FALSE,warning=FALSE}
ribo_rna_counts = read.csv(ribo_rna_counts_file )

ribo_rna_gene_names        = unlist(lapply( (strsplit(ribo_rna_counts$transcript, split = "-" ) ), "[[", 1  ) )

ribo_rna_counts$transcript = ribo_rna_gene_names

ribo_rna_counts
```


We apply CLR normalization to our rna and ribo data
```{r echo=TRUE,warning=FALSE}
set.seed(3)

clr_normalized_counts_pre = NormalizeData(ribo_rna_counts[,-1], 
                                          scale.factor = 10000, margin = 2, 
                                          normalization.method= "CLR",verbose = TRUE)

clr_normalized_counts = data.frame(transcript = ribo_rna_counts$transcript , clr_normalized_counts_pre)

colSums(clr_normalized_counts_pre)
print("\n\n##################\n\n")
colSums(ribo_rna_counts[,-1])
```

```{r echo=FALSE,warning=FALSE}

rnaseq_fourcell_averages_clr  = (clr_normalized_counts$X4cell.1.RNAseq + clr_normalized_counts$X4cell.2.RNAseq) / 2
rnaseq_eightcell_averages_clr = (clr_normalized_counts$X8cell.1.RNAseq + clr_normalized_counts$X8cell.2.RNAseq + clr_normalized_counts$X8cell.3.RNAseq + clr_normalized_counts$X8cell.4.RNAseq) / 4
ribo_fourcell_averages_clr    = (clr_normalized_counts$X4cell.1 + clr_normalized_counts$X4cell.2 + clr_normalized_counts$X4cell.3) / 3
ribo_eightcell_averages_clr   = (clr_normalized_counts$X8cell.1 + clr_normalized_counts$X8cell.2 + clr_normalized_counts$X8cell.3 + clr_normalized_counts$X8cell.4) / 4

fourcell_TE_clr  = ribo_fourcell_averages_clr / rnaseq_fourcell_averages_clr
eightcell_TE_clr = ribo_eightcell_averages_clr / rnaseq_eightcell_averages_clr


TE_df_clr = data.frame( transcript = ribo_rna_counts$transcript, fourcell_TE_clr = fourcell_TE_clr, eightcell_TE_clr = eightcell_TE_clr  )

hist( log2(fourcell_TE_clr) )

hist( log2(eightcell_TE_clr) )

```

```{r echo=FALSE,warning=FALSE}



fourcell_background_genes = setdiff( TE_df_clr$transcript , union( fourcell_maternal_genes, fourcell_paternal_genes  ) )

eightcell_background_genes = setdiff( TE_df_clr$transcript , union( eightcell_maternal_genes, eightcell_paternal_genes  ) )


fourcell_paternal_TE_clr   = TE_df_clr[ TE_df_clr$transcript %in% fourcell_paternal_genes,]$fourcell_TE_clr
fourcell_maternal_TE_clr   = TE_df_clr[ TE_df_clr$transcript %in% fourcell_maternal_genes,]$fourcell_TE_clr
fourcell_background_TE_clr = TE_df_clr[ TE_df_clr$transcript %in% fourcell_background_genes,]$fourcell_TE_clr


fourcell_paternal_rna_clr   = rnaseq_fourcell_averages_clr[ TE_df_clr$transcript %in% fourcell_paternal_genes    ]
fourcell_maternal_rna_clr   = rnaseq_fourcell_averages_clr[ TE_df_clr$transcript %in% fourcell_maternal_genes    ]
fourcell_background_rna_clr = rnaseq_fourcell_averages_clr[ TE_df_clr$transcript %in% fourcell_background_genes  ]

fourcell_arranged_TE_clr = c( fourcell_background_TE_clr, fourcell_paternal_TE_clr, fourcell_maternal_TE_clr )

fourcell_arranged_rna_clr = c(fourcell_background_rna_clr, fourcell_paternal_rna_clr, fourcell_maternal_rna_clr )

fourcell_TE_labels_clr = c(
  rep("background", length(fourcell_background_TE_clr)) ,
  rep("paternal", length(fourcell_paternal_TE_clr) ),
  rep("maternal",  length(fourcell_maternal_TE_clr) )
)

fourcell_TE_imprinted_clr = c(
  rep(0, length(fourcell_background_TE_clr)) ,
  rep(1, length(fourcell_paternal_TE_clr) ),
  rep(1,  length(fourcell_maternal_TE_clr) )
)




fourcell_paternal_genes_clr   = TE_df_clr[ TE_df_clr$transcript %in% fourcell_paternal_genes,]$transcript
fourcell_maternal_genes_clr   = TE_df_clr[ TE_df_clr$transcript %in% fourcell_maternal_genes,]$transcript
fourcell_background_genes_clr = TE_df_clr[ TE_df_clr$transcript %in% fourcell_background_genes,]$transcript

fourcell_arranged_genes_clr = c(
  fourcell_background_genes_clr,
  fourcell_paternal_genes_clr,
  fourcell_maternal_genes_clr
)


fourcell_TE_arranged_df_clr = data.frame( type       = fourcell_TE_labels_clr,
                                      imprinted  = fourcell_TE_imprinted_clr,
                                      transcript = fourcell_arranged_genes_clr,
                                      TE         = fourcell_arranged_TE_clr,
                                      rna        = fourcell_arranged_rna_clr )





sample_fourcell_clr = matchit(imprinted~rna, data = fourcell_TE_arranged_df_clr )

sample_fourcell_clr


matched_fourcell_clr = match.data(sample_fourcell_clr)

matched_fourcell_clr

boxplot( log2(matched_fourcell_clr$rna) ~ matched_fourcell_clr$imprinted )
boxplot( (matched_fourcell_clr$rna) ~ matched_fourcell_clr$imprinted )

boxplot( log2(matched_fourcell_clr$rna) ~ matched_fourcell_clr$type )

boxplot( log2(matched_fourcell_clr$TE) ~ matched_fourcell_clr$type )

boxplot( (matched_fourcell_clr$TE) ~ matched_fourcell_clr$type )



```

```{r echo=FALSE,warning=FALSE}



wilcox.test( matched_fourcell_clr[matched_fourcell_clr$type == "background",]$TE, matched_fourcell_clr[matched_fourcell_clr$type == "maternal",]$TE  )

wilcox.test( matched_fourcell_clr[matched_fourcell_clr$type == "background",]$TE, matched_fourcell_clr[matched_fourcell_clr$type == "paternal",]$TE  )

wilcox.test( matched_fourcell_clr[matched_fourcell_clr$type == "paternal",]$TE, matched_fourcell_clr[matched_fourcell_clr$type == "maternal",]$TE  )

```


```{r echo=FALSE,warning=FALSE}

fourcell_TE_and_lengths = merge(matched_fourcell_clr, region_lengths, by.x = "transcript", by.y = "gene")

hist( log2(fourcell_TE_and_lengths$cds_length) )

hist( log2(cds_lengths) )

median( fourcell_TE_and_lengths$cds_length )

median( cds_lengths )

boxplot( (fourcell_TE_and_lengths$cds_length) ~ fourcell_TE_and_lengths$type )


hist( log2( fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "maternal",]$cds_length  ) )

hist( log2( fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "background",]$cds_length  ) )


plot(density( log2( fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "maternal",]$cds_length  )   ) )

lines(density( log2( fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "background",]$cds_length  )   ), col = "red" )
```


```{r echo=FALSE,warning=FALSE}

wilcox.test(  fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "maternal",]$cds_length,  fourcell_TE_and_lengths[fourcell_TE_and_lengths$type == "background",]$cds_length  )

wilcox.test( fourcell_TE_and_lengths$cds_length, cds_lengths  )
```





```{r echo=FALSE,warning=FALSE}

eightcell_background_genes = setdiff( TE_df_clr$transcript , union( eightcell_maternal_genes, eightcell_paternal_genes  ) )

eightcell_background_genes = setdiff( TE_df_clr$transcript , union( eightcell_maternal_genes, eightcell_paternal_genes  ) )


eightcell_paternal_TE_clr   = TE_df_clr[ TE_df_clr$transcript %in% eightcell_paternal_genes,]$eightcell_TE_clr
eightcell_maternal_TE_clr   = TE_df_clr[ TE_df_clr$transcript %in% eightcell_maternal_genes,]$eightcell_TE_clr
eightcell_background_TE_clr = TE_df_clr[ TE_df_clr$transcript %in% eightcell_background_genes,]$eightcell_TE_clr


eightcell_paternal_rna_clr   = rnaseq_eightcell_averages_clr[ TE_df_clr$transcript %in% eightcell_paternal_genes    ]
eightcell_maternal_rna_clr   = rnaseq_eightcell_averages_clr[ TE_df_clr$transcript %in% eightcell_maternal_genes    ]
eightcell_background_rna_clr = rnaseq_eightcell_averages_clr[ TE_df_clr$transcript %in% eightcell_background_genes  ]

eightcell_arranged_TE_clr = c( eightcell_background_TE_clr, eightcell_paternal_TE_clr, eightcell_maternal_TE_clr )

eightcell_arranged_rna_clr = c(eightcell_background_rna_clr, eightcell_paternal_rna_clr, eightcell_maternal_rna_clr )

eightcell_TE_labels_clr = c(
  rep("background", length(eightcell_background_TE_clr)) ,
  rep("paternal", length(eightcell_paternal_TE_clr) ),
  rep("maternal",  length(eightcell_maternal_TE_clr) )
)

eightcell_TE_imprinted_clr = c(
  rep(0, length(eightcell_background_TE_clr)) ,
  rep(1, length(eightcell_paternal_TE_clr) ),
  rep(1,  length(eightcell_maternal_TE_clr) )
)




eightcell_paternal_genes_clr   = TE_df_clr[ TE_df_clr$transcript %in% eightcell_paternal_genes,]$transcript
eightcell_maternal_genes_clr   = TE_df_clr[ TE_df_clr$transcript %in% eightcell_maternal_genes,]$transcript
eightcell_background_genes_clr = TE_df_clr[ TE_df_clr$transcript %in% eightcell_background_genes,]$transcript

eightcell_arranged_genes_clr = c(
  eightcell_background_genes_clr,
  eightcell_paternal_genes_clr,
  eightcell_maternal_genes_clr
)

eightcell_TE_arranged_df_clr = data.frame( type       = eightcell_TE_labels_clr,
                                      imprinted  = eightcell_TE_imprinted_clr,
                                      transcript = eightcell_arranged_genes_clr,
                                      TE         = eightcell_arranged_TE_clr,
                                      rna        = eightcell_arranged_rna_clr )





sample_eightcell_clr = matchit(imprinted~rna, data = eightcell_TE_arranged_df_clr )

sample_eightcell_clr


matched_eightcell_clr = match.data(sample_eightcell_clr)

matched_eightcell_clr

boxplot( log2(matched_eightcell_clr$rna) ~ matched_eightcell_clr$imprinted )

boxplot( log2(matched_eightcell_clr$rna) ~ matched_eightcell_clr$type )

boxplot( log2(matched_eightcell_clr$TE) ~ matched_eightcell_clr$type )

boxplot( (matched_eightcell_clr$TE) ~ matched_eightcell_clr$type )

```


```{r echo=FALSE,warning=FALSE}



wilcox.test( matched_eightcell_clr[matched_eightcell_clr$type == "background",]$TE, matched_eightcell_clr[matched_eightcell_clr$type == "maternal",]$TE  )

wilcox.test( matched_eightcell_clr[matched_eightcell_clr$type == "background",]$TE, matched_eightcell_clr[matched_eightcell_clr$type == "paternal",]$TE  )

wilcox.test( matched_eightcell_clr[matched_eightcell_clr$type == "paternal",]$TE, matched_eightcell_clr[matched_eightcell_clr$type == "maternal",]$TE  )

```




























## Imprinted genes after Bootsrapping 

We obtain the snp counts for individual experiments first. Next, we aggregate the counts by position and obtain counts for each experiment per gene.

```{r echo=FALSE,warning=FALSE}

experiments_fourcell  = c("20210607-RNAseq-4cell-cross-A", "20210607-RNAseq-4cell-cross-C")
experiments_eightcell = c("20210607-RNAseq-8cell-cross-A", "20210607-RNAseq-8cell-cross-B", "20210607-RNAseq-8cell-cross-C", "20210607-RNAseq-8cell-cross-D") 

detailed_snps = read.csv(detailed_snp_file, header = T)

detailed_snps = detailed_snps[, 1:5]

detailed_snps = detailed_snps[detailed_snps$experiment %in% c(experiments_fourcell, experiments_eightcell), ]

detailed_snps = 
  detailed_snps %>% group_by( experiment, transcript  ) %>%
    summarise(paternal_total = sum(paternal), maternal_total = sum(maternal))

detailed_snps = detailed_snps[ order(detailed_snps$transcript)  ,]

colnames(detailed_snps) = c("experiment", "gene", "paternal", "maternal")

detailed_snps$gene = unlist(lapply( strsplit(detailed_snps$gene, split = "-"), "[[", 1 ) )

detailed_snps[1000:1020,]

```


```{r echo=FALSE,warning=FALSE}

detailed_fourcell_snps = detailed_snps[ detailed_snps$experiment %in% experiments_fourcell, ]
detailed_fourcell_snps = detailed_fourcell_snps[ detailed_fourcell_snps$gene %in% four_cell_snp_filtered$gene,  ]

detailed_eightcell_snps = detailed_snps[ detailed_snps$experiment %in% experiments_eightcell,  ]
detailed_eightcell_snps = detailed_eightcell_snps[ detailed_eightcell_snps$gene %in% eight_cell_snp_filtered$gene,  ]



compute_paternal_ratio = function(unit_df, paternal_threshold){
  paternal_ratios        = (unit_df$paternal + CONSTANT_ADDITION) / (unit_df$paternal + unit_df$maternal + (2 * CONSTANT_ADDITION) )
  unit_df$paternal_ratio = paternal_ratios
  unit_df$is_paternal     = as.integer( paternal_ratios > paternal_threshold )
  unit_df$is_maternal     = as.integer( paternal_ratios < (1 - paternal_threshold) )
  
  return(unit_df)
}

unit_sample = function( snp_df , paternal_threshold ){
  experiments         = unique(snp_df$experiment)
  sampled_experiments = sample( experiments, replace = T, size = length( experiments  ) )
  
  initial_df = snp_df[ snp_df$experiment == sampled_experiments[1], -1  ]
  
  #print(sampled_experiments)
  
  for (e in sampled_experiments[-1]){
    this_df             = snp_df[ snp_df$experiment == e,   ]
    initial_df$paternal = initial_df$paternal + this_df$paternal
    initial_df$maternal = initial_df$maternal + this_df$maternal
  }
  
  result = compute_paternal_ratio(initial_df, paternal_threshold)
  
  return(result)
}

bulk_sample = function( snp_df, paternal_threshold, rounds = 100  ){
  experiments         = unique(snp_df$experiment)
  
  genes = snp_df[ snp_df$experiment == experiments[1],   ]$gene
  
  paternal_hits = rep(0, length(genes))
  maternal_hits = rep(0, length(genes))
  
  for(i in 1:rounds){
    this_sampling_df = unit_sample(snp_df, paternal_threshold = 0.7)
    paternal_hits = paternal_hits + this_sampling_df$is_paternal
    maternal_hits = maternal_hits + this_sampling_df$is_maternal
  }
  
  result = data.frame( gene = genes, paternal = paternal_hits, maternal = maternal_hits  )
  
  
  return(result)
  
}

#a = unit_sample(detailed_eightcell_snps)
#compute_paternal_ratio(a)


#unit_sample(detailed_eightcell_snps, paternal_threshold = 0.7)
```


```{r echo=FALSE,warning=FALSE}

set.seed(3)

eightcell_bootstrap_df = bulk_sample(detailed_eightcell_snps, paternal_threshold = 0.7, rounds = 1000)
fourcell_bootstrap_df  = bulk_sample(detailed_fourcell_snps, paternal_threshold  = 0.7, rounds = 1000) 

```



```{r echo=FALSE,warning=FALSE}

hist( eightcell_bootstrap_df$maternal, breaks = seq(0,1000, 50)   )

hist( fourcell_bootstrap_df$maternal   )
```


```{r echo=FALSE,warning=FALSE}

BOOTSTRAP_THRESHOLD = 800

eightcell_paternal_genes_bootstrap   = eightcell_bootstrap_df[eightcell_bootstrap_df$paternal > BOOTSTRAP_THRESHOLD, ]$gene 
eightcell_maternal_genes_bootstrap   = eightcell_bootstrap_df[eightcell_bootstrap_df$maternal > BOOTSTRAP_THRESHOLD, ]$gene 
eightcell_background_genes_bootstrap = setdiff( eightcell_bootstrap_df$gene , c(eightcell_paternal_genes_bootstrap, eightcell_maternal_genes_bootstrap) )

fourcell_paternal_genes_bootstrap   = fourcell_bootstrap_df[fourcell_bootstrap_df$paternal > BOOTSTRAP_THRESHOLD, ]$gene
fourcell_maternal_genes_bootstrap   = fourcell_bootstrap_df[fourcell_bootstrap_df$maternal > BOOTSTRAP_THRESHOLD, ]$gene
fourcell_background_genes_bootstrap = setdiff( fourcell_bootstrap_df$gene , c(fourcell_paternal_genes_bootstrap, fourcell_maternal_genes_bootstrap) )


print( c(  "Paternal Genes from 4cells and 8 cells are:", length(fourcell_paternal_genes_bootstrap), length(eightcell_paternal_genes_bootstrap) ) )
print( c(  "Maternal Genes from 4cells and 8 cells are:", length(fourcell_maternal_genes_bootstrap), length(eightcell_maternal_genes_bootstrap) ) )

```


```{r echo=FALSE,warning=FALSE}


mymain = function(TE_df, rna_averages, gene_lengths, picker_name , paternal_genes, maternal_genes, background_genes){
  paternal_TE = TE_df[ TE_df$transcript %in% paternal_genes,   ][picker_name][,1]
  maternal_TE = TE_df[ TE_df$transcript %in% maternal_genes,   ][picker_name][,1]
  bg_TE       = TE_df[ TE_df$transcript %in% background_genes, ][picker_name][,1]
  
  paternal_rna   = rna_averages[ TE_df$transcript %in% paternal_genes ]
  maternal_rna   = rna_averages[ TE_df$transcript %in% maternal_genes ]
  background_rna = rna_averages[ TE_df$transcript %in% background_genes ]
  
  arranged_TE =  c( bg_TE, paternal_TE, maternal_TE )
  arrenged_rna = c( background_rna, paternal_rna, maternal_rna)
  
  TE_labels = c(
      rep("background", length(bg_TE)) ,
      rep("paternal", length(paternal_TE) ),
      rep("maternal",  length(maternal_TE) )
  )
  
  TE_imprinted = c(
    rep(0, length(bg_TE)) ,
    rep(1, length(paternal_TE) ),
    rep(1,  length(maternal_TE) )
  )
  
  paternal_genes_TE   = TE_df[TE_df$transcript %in% paternal_genes,]$transcript
  maternal_genes_TE   = TE_df[TE_df$transcript %in% maternal_genes,]$transcript
  background_genes_TE = TE_df[TE_df$transcript %in% background_genes,]$transcript
  
  TE_arranged_genes = c(background_genes_TE, paternal_genes_TE, maternal_genes_TE)
  
  TE_arranged_df_pre = data.frame(  type       = TE_labels,
                                imprinted  = TE_imprinted,
                                gene       = TE_arranged_genes,
                                TE         = arranged_TE,
                                rna        = arrenged_rna )
  
  TE_arranged_df = merge(TE_arranged_df_pre, gene_lengths, by.x = "gene", by.y= "gene")
  
  return(TE_arranged_df)
  
}


```



```{r echo=FALSE,warning=FALSE}

set.seed(3)

fourcell_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_fourcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "fourcell_TE_clr"  , 
       paternal_genes   = fourcell_paternal_genes_bootstrap, 
       maternal_genes   = fourcell_maternal_genes_bootstrap, 
       background_genes = fourcell_background_genes_bootstrap)

fourcell_TE



sample_fourcell = matchit(imprinted~rna + cds_length, data = fourcell_TE )

sample_fourcell


matched_fourcell = match.data(sample_fourcell)

matched_fourcell

boxplot( log2(fourcell_TE$rna) ~ fourcell_TE$imprinted )

boxplot( log2(matched_fourcell$rna) ~ matched_fourcell$imprinted )

boxplot( log2(matched_fourcell$rna) ~ matched_fourcell$type )

boxplot( fourcell_TE$TE ~ fourcell_TE$type )

boxplot( matched_fourcell$TE ~ matched_fourcell$type )


plot(density( log2( matched_fourcell[matched_fourcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_fourcell[matched_fourcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_fourcell[matched_fourcell$type == "background",]$cds_length  )   ), col = "red" )

```

```{r echo=FALSE,warning=FALSE}
wilcox.test( matched_fourcell[matched_fourcell$type == "background", ]$TE, matched_fourcell[matched_fourcell$type == "maternal", ]$TE  )

wilcox.test( matched_fourcell[matched_fourcell$type == "background", ]$TE, matched_fourcell[matched_fourcell$type == "paternal", ]$TE  )

wilcox.test( matched_fourcell[matched_fourcell$type == "background", ]$TE, c(matched_fourcell[matched_fourcell$type == "paternal", ]$TE, matched_fourcell[matched_fourcell$type == "maternal", ]$TE)  )

wilcox.test( matched_fourcell[matched_fourcell$type == "maternal", ]$TE, matched_fourcell[matched_fourcell$type == "paternal", ]$TE  )
```

```{r echo=FALSE,warning=FALSE}
wilcox.test( fourcell_TE[fourcell_TE$type == "background", ]$TE, fourcell_TE[fourcell_TE$type == "maternal", ]$TE  )

wilcox.test( fourcell_TE[fourcell_TE$type == "background", ]$TE, fourcell_TE[fourcell_TE$type == "paternal", ]$TE  )

wilcox.test( fourcell_TE[fourcell_TE$type == "background", ]$TE, c(fourcell_TE[fourcell_TE$type == "paternal", ]$TE, fourcell_TE[fourcell_TE$type == "maternal", ]$TE)  )

wilcox.test( fourcell_TE[fourcell_TE$type == "maternal", ]$TE, fourcell_TE[fourcell_TE$type == "paternal", ]$TE  )
```



```{r echo=FALSE,warning=FALSE}

four_cell_snp_filtered$total_rna_snp = four_cell_snp_filtered$maternal_y + four_cell_snp_filtered$paternal_y


fourcell_imprented_totals = four_cell_snp_filtered[four_cell_snp_filtered$gene %in% fourcell_TE[fourcell_TE$imprinted == 1, ]$gene, ]$total_rna_snp
fourcell_background_totals = four_cell_snp_filtered[four_cell_snp_filtered$gene %in% fourcell_TE[fourcell_TE$imprinted == 0, ]$gene, ]$total_rna_snp

plot(  density(  log2(fourcell_imprented_totals) ) )
lines(  density( log2(fourcell_background_totals) ), col="red" )

hist( log2(fourcell_imprented_totals)) 
lines( hist( log2(fourcell_imprented_totals), plot = F) ) 

four_cell_snp_filtered$paternal_ratio = four_cell_snp_filtered$paternal_y / four_cell_snp_filtered$total_rna_snp
four_cell_snp_filtered[four_cell_snp_filtered$paternal_ratio == 0,]
four_cell_snp_filtered[four_cell_snp_filtered$paternal_ratio == 1,]

```


```{r echo=FALSE,warning=FALSE}

eightcell_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_eightcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "eightcell_TE_clr"  , 
       paternal_genes   = eightcell_paternal_genes_bootstrap, 
       maternal_genes   = eightcell_maternal_genes_bootstrap, 
       background_genes = eightcell_background_genes_bootstrap)

eightcell_TE



sample_eightcell = matchit(imprinted~rna + cds_length, data = eightcell_TE )

sample_eightcell


matched_eightcell = match.data(sample_eightcell)

matched_eightcell

boxplot( log2(matched_eightcell$rna) ~ matched_eightcell$imprinted )

boxplot( log2(matched_eightcell$rna) ~ matched_eightcell$type )

boxplot( matched_eightcell$TE ~ matched_eightcell$type )


plot(density( log2( matched_eightcell[matched_eightcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_eightcell[matched_eightcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_eightcell[matched_eightcell$type == "background",]$cds_length  )   ), col = "red" )

```



```{r echo=FALSE,warning=FALSE}
wilcox.test( matched_eightcell[matched_eightcell$type == "background", ]$TE, matched_eightcell[matched_eightcell$type == "maternal", ]$TE  )

wilcox.test( matched_eightcell[matched_eightcell$type == "background", ]$TE, matched_eightcell[matched_eightcell$type == "paternal", ]$TE  )

wilcox.test( matched_eightcell[matched_eightcell$type == "background", ]$TE, c(matched_eightcell[matched_eightcell$type == "paternal", ]$TE, matched_eightcell[matched_eightcell$type == "maternal", ]$TE )  )

wilcox.test( matched_eightcell[matched_eightcell$type == "maternal", ]$TE, matched_eightcell[matched_eightcell$type == "paternal", ]$TE  )
```

```{r echo=FALSE,warning=FALSE}
wilcox.test( eightcell_TE[eightcell_TE$type == "background", ]$TE, eightcell_TE[eightcell_TE$type == "maternal", ]$TE  )

wilcox.test( eightcell_TE[eightcell_TE$type == "background", ]$TE, eightcell_TE[eightcell_TE$type == "paternal", ]$TE  )

wilcox.test( eightcell_TE[eightcell_TE$type == "background", ]$TE, c(eightcell_TE[eightcell_TE$type == "paternal", ]$TE, eightcell_TE[eightcell_TE$type == "maternal", ]$TE )  )

wilcox.test( matched_eightcell[matched_eightcell$type == "maternal", ]$TE, matched_eightcell[matched_eightcell$type == "paternal", ]$TE  )
```


## Mono Allelic Genes

Among our imprinted genes, 
If a gene has (zero paternal counts and more than one paternal counts) or
(has more than one paternal counts and zero maternal counts), it is called a mono alelic gene.
Here we investigate the translation efficiency of mono allelic genes.




```{r echo=FALSE,warning=FALSE}

mono_allelic_eightcell_paternal_df = eight_cell_snp_filtered[ eight_cell_snp_filtered$maternal_y == 0 & eight_cell_snp_filtered$paternal_y > 0 & eight_cell_snp_filtered$gene %in% eightcell_paternal_genes_bootstrap,  ]

mono_allelic_eightcell_paternal_df

mono_allelic_eightcell_paternal_genes = eight_cell_snp_filtered[ eight_cell_snp_filtered$maternal_y == 0 & eight_cell_snp_filtered$paternal_y > 0 & eight_cell_snp_filtered$gene %in% eightcell_paternal_genes_bootstrap,  ]$gene


mono_allelic_eightcell_maternal_genes = eight_cell_snp_filtered[ eight_cell_snp_filtered$paternal_y == 0 & eight_cell_snp_filtered$maternal_y > 0 & eight_cell_snp_filtered$gene %in% eightcell_maternal_genes_bootstrap,  ]$gene

print( c("Eight cell Monoallelic paternal and maternal genes", length(mono_allelic_eightcell_paternal_genes), length(mono_allelic_eightcell_maternal_genes))  )


mono_allelic_fourcell_paternal_genes = four_cell_snp_filtered[ four_cell_snp_filtered$maternal_y == 0 & four_cell_snp_filtered$paternal_y > 0 & four_cell_snp_filtered$gene %in% fourcell_paternal_genes_bootstrap,  ]$gene


mono_allelic_fourcell_maternal_genes = four_cell_snp_filtered[ four_cell_snp_filtered$paternal_y == 0 & four_cell_snp_filtered$maternal_y > 0 & four_cell_snp_filtered$gene %in% fourcell_maternal_genes_bootstrap,  ]$gene

print( c("four cell Monoallelic paternal and maternal genes", length(mono_allelic_fourcell_paternal_genes), length(mono_allelic_fourcell_maternal_genes))  )

mono_allelic_eightcell_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_eightcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "eightcell_TE_clr"  , 
       paternal_genes   = mono_allelic_eightcell_paternal_genes, 
       maternal_genes   = mono_allelic_eightcell_maternal_genes, 
       background_genes = eightcell_background_genes_bootstrap)

mono_allelic_eightcell_TE


mono_allelic_fourcell_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_fourcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "fourcell_TE_clr"  , 
       paternal_genes   = mono_allelic_fourcell_paternal_genes, 
       maternal_genes   = mono_allelic_fourcell_maternal_genes, 
       background_genes = fourcell_background_genes_bootstrap)

mono_allelic_fourcell_TE

```


```{r echo=FALSE,warning=FALSE}

length(mono_allelic_eightcell_paternal_genes)
length(mono_allelic_fourcell_maternal_genes)
```

```{r echo=FALSE,warning=FALSE}

length(eightcell_paternal_genes_bootstrap)
length(fourcell_paternal_genes_bootstrap)
length(intersect( eightcell_paternal_genes_bootstrap, fourcell_paternal_genes_bootstrap  ))

length(eightcell_maternal_genes_bootstrap)
length(fourcell_maternal_genes_bootstrap)
length(intersect( eightcell_maternal_genes_bootstrap, fourcell_maternal_genes_bootstrap  ))

```

### Eight Cell Mono Allelic Genes

```{r echo=FALSE,warning=FALSE}

sample_mono_allelic_eightcell = matchit(imprinted~rna + cds_length, data = mono_allelic_eightcell_TE )

sample_mono_allelic_eightcell


matched_mono_allelic_eightcell = match.data(sample_mono_allelic_eightcell)

matched_mono_allelic_eightcell

boxplot( log2(matched_mono_allelic_eightcell$rna) ~ matched_mono_allelic_eightcell$imprinted )

boxplot( log2(matched_mono_allelic_eightcell$rna) ~ matched_mono_allelic_eightcell$type )

boxplot( mono_allelic_eightcell_TE$TE ~ mono_allelic_eightcell_TE$type )

boxplot( matched_mono_allelic_eightcell$TE ~ matched_mono_allelic_eightcell$type )


plot(density( log2( matched_mono_allelic_eightcell[matched_mono_allelic_eightcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_mono_allelic_eightcell[matched_mono_allelic_eightcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_mono_allelic_eightcell[matched_mono_allelic_eightcell$type == "background",]$cds_length  )   ), col = "red" )



```




```{r echo=FALSE,warning=FALSE}

wilcox.test( matched_mono_allelic_eightcell[matched_mono_allelic_eightcell$imprinted == 0, ]$TE , matched_mono_allelic_eightcell[matched_mono_allelic_eightcell$imprinted == 1, ]$TE  )

```

### Four Cell Mono Allelic Genes

```{r echo=FALSE,warning=FALSE}

sample_mono_allelic_fourcell = matchit(imprinted~rna + cds_length, data = mono_allelic_fourcell_TE )

sample_mono_allelic_fourcell


matched_mono_allelic_fourcell = match.data(sample_mono_allelic_fourcell)

matched_mono_allelic_fourcell

boxplot( log2(matched_mono_allelic_fourcell$rna) ~ matched_mono_allelic_fourcell$imprinted )

boxplot( log2(matched_mono_allelic_fourcell$rna) ~ matched_mono_allelic_fourcell$type )

boxplot( mono_allelic_fourcell_TE$TE ~ mono_allelic_fourcell_TE$type, varwidth = T )

boxplot( matched_mono_allelic_fourcell$TE ~ matched_mono_allelic_fourcell$type )


plot(density( log2( matched_mono_allelic_fourcell[matched_mono_allelic_fourcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_mono_allelic_fourcell[matched_mono_allelic_fourcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_mono_allelic_fourcell[matched_mono_allelic_fourcell$type == "background",]$cds_length  )   ), col = "red" )



```


```{r echo=FALSE,warning=FALSE}

wilcox.test( matched_mono_allelic_fourcell[matched_mono_allelic_fourcell$imprinted == 0, ]$TE , matched_mono_allelic_fourcell[matched_mono_allelic_fourcell$imprinted == 1, ]$TE  )

```


### Multiple SNP Concordence

Now we check that multiple SNPS 

```{r echo=FALSE,warning=FALSE}

raw_snps = read.csv(detailed_snp_file)

raw_snps_four_cell = raw_snps[raw_snps$experiment %in% experiments_fourcell, ]
raw_snps_eight_cell = raw_snps[raw_snps$experiment %in% experiments_eightcell, ]

compute_support_ratio = function(this_df){
  result = 
    this_df %>% group_by(experiment, transcript) %>%
    mutate(support_ratio_of_replicate = sum(support) / snp_positions)
  
  result = 
    result %>% group_by(transcript, position) %>%
    mutate(support_ratio_combined = sum(support_ratio_of_replicate))
  
  return(result)
}

find_well_supported_genes = function(this_df, threshold){
  result = this_df[this_df$support_ratio_combined > threshold,]$transcript 
  return( unique(result) )
} 

```

```{r echo=FALSE,warning=FALSE}

snps_pos_four_cell_pre = raw_snps_four_cell[, 1:5]

snps_pos_four_cell_pre$transcript = unlist( lapply(strsplit(snps_pos_four_cell_pre$transcript, split = "-" ) , "[[", 1))

snps_pos_four_cell_pre = snps_pos_four_cell_pre[ snps_pos_four_cell_pre$paternal + snps_pos_four_cell_pre$maternal > 0 , ]

snps_pos_four_cell_pre = 
  snps_pos_four_cell_pre %>% 
  group_by(experiment, transcript) %>%
  mutate(snp_positions = length(position))


snps_pos_four_cell_pre = snps_pos_four_cell_pre[snps_pos_four_cell_pre$snp_positions > 1, ]

snps_pos_four_cell_pre$paternal_ratio = (snps_pos_four_cell_pre$paternal + 1) / (snps_pos_four_cell_pre$paternal + snps_pos_four_cell_pre$maternal + 2)

snps_pos_four_cell_paternal = snps_pos_four_cell_pre[ snps_pos_four_cell_pre$transcript %in% fourcell_paternal_genes_bootstrap,  ]
snps_pos_four_cell_maternal = snps_pos_four_cell_pre[ snps_pos_four_cell_pre$transcript %in% fourcell_maternal_genes_bootstrap,  ]

snps_pos_four_cell_paternal$support = as.integer( snps_pos_four_cell_paternal$paternal_ratio > 0.5 )
snps_pos_four_cell_maternal$support = as.integer( snps_pos_four_cell_maternal$paternal_ratio < 0.5 )



snps_pos_four_cell_paternal = compute_support_ratio(snps_pos_four_cell_paternal)
snps_pos_four_cell_maternal = compute_support_ratio(snps_pos_four_cell_maternal)

snps_pos_four_cell_paternal
snps_pos_four_cell_maternal


# There needs to be 0.6 support from each replicate hence the thereshold 1.2
multiple_snp_supported_fourcell_paternal_genes = find_well_supported_genes(snps_pos_four_cell_paternal, 1.2)

multiple_snp_supported_fourcell_maternal_genes = find_well_supported_genes(snps_pos_four_cell_maternal, 1.2)
#snps_pos_four_cell_paternal$paternal_ratio = (snps_pos_four_cell_paternal$paternal + 1) / (snps_pos_four_cell_maternal$paternal + snps_pos_four_cell_paternal$maternal + 2)

multiple_snp_supported_fourcell_paternal_genes
multiple_snp_supported_fourcell_maternal_genes
```


```{r echo=FALSE,warning=FALSE}
length(unique(snps_pos_four_cell_maternal$transcript) )
length(  multiple_snp_supported_fourcell_maternal_genes )
```

```{r echo=FALSE,warning=FALSE}
snps_pos_eight_cell_pre = raw_snps_eight_cell[, 1:5]

snps_pos_eight_cell_pre$transcript = unlist( lapply(strsplit(snps_pos_eight_cell_pre$transcript, split = "-" ) , "[[", 1))

snps_pos_eight_cell_pre = snps_pos_eight_cell_pre[ snps_pos_eight_cell_pre$paternal + snps_pos_eight_cell_pre$maternal > 0 , ]

snps_pos_eight_cell_pre = 
  snps_pos_eight_cell_pre %>% 
  group_by(experiment, transcript) %>%
  mutate(snp_positions = length(position))

snps_pos_eight_cell_pre = snps_pos_eight_cell_pre[snps_pos_eight_cell_pre$snp_positions > 1, ]

snps_pos_eight_cell_pre$paternal_ratio = (snps_pos_eight_cell_pre$paternal + 1) / (snps_pos_eight_cell_pre$paternal + snps_pos_eight_cell_pre$maternal + 2)

snps_pos_eight_cell_paternal = snps_pos_eight_cell_pre[ snps_pos_eight_cell_pre$transcript %in% eightcell_paternal_genes_bootstrap,  ]
snps_pos_eight_cell_maternal = snps_pos_eight_cell_pre[ snps_pos_eight_cell_pre$transcript %in% eightcell_maternal_genes_bootstrap,  ]

snps_pos_eight_cell_paternal$support = as.integer( snps_pos_eight_cell_paternal$paternal_ratio > 0.5 )
snps_pos_eight_cell_maternal$support = as.integer( snps_pos_eight_cell_maternal$paternal_ratio < 0.5 )



snps_pos_eight_cell_paternal = compute_support_ratio(snps_pos_eight_cell_paternal)
snps_pos_eight_cell_maternal = compute_support_ratio(snps_pos_eight_cell_maternal)

snps_pos_eight_cell_paternal
snps_pos_eight_cell_maternal


# There needs to be 0.6 support from each replicate hence the thereshold 2.4
multiple_snp_supported_eightcell_paternal_genes = find_well_supported_genes(snps_pos_eight_cell_paternal, 2.4)

multiple_snp_supported_eightcell_maternal_genes = find_well_supported_genes(snps_pos_eight_cell_maternal, 2.4)

multiple_snp_supported_eightcell_paternal_genes
multiple_snp_supported_eightcell_maternal_genes
```


```{r echo=FALSE,warning=FALSE}

set.seed(3)

fourcell_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_fourcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "fourcell_TE_clr"  , 
       paternal_genes   = multiple_snp_supported_fourcell_paternal_genes, 
       maternal_genes   = multiple_snp_supported_fourcell_maternal_genes, 
       background_genes = fourcell_background_genes_bootstrap)

fourcell_TE



sample_fourcell = matchit(imprinted~rna + cds_length, data = fourcell_TE )

sample_fourcell


matched_fourcell = match.data(sample_fourcell)

matched_fourcell

boxplot( log2(fourcell_TE$rna) ~ fourcell_TE$imprinted )

boxplot( log2(matched_fourcell$rna) ~ matched_fourcell$imprinted )

boxplot( log2(matched_fourcell$rna) ~ matched_fourcell$type )

boxplot( fourcell_TE$TE ~ fourcell_TE$type )

boxplot( matched_fourcell$TE ~ matched_fourcell$type )


plot(density( log2( matched_fourcell[matched_fourcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.6) )

lines(density( log2( matched_fourcell[matched_fourcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_fourcell[matched_fourcell$type == "background",]$cds_length  )   ), col = "red" )
```


```{r echo=FALSE,warning=FALSE}
wilcox.test(  fourcell_TE[fourcell_TE$imprinted == 0, ]$TE, fourcell_TE[fourcell_TE$imprinted == 1, ]$TE )

wilcox.test(  matched_fourcell[matched_fourcell$imprinted == 0, ]$TE, matched_fourcell[matched_fourcell$imprinted == 1, ]$TE )

wilcox.test(  fourcell_TE[fourcell_TE$type == "background", ]$TE, fourcell_TE[fourcell_TE$type == "paternal", ]$TE )

wilcox.test(  fourcell_TE[fourcell_TE$type == "background", ]$TE, fourcell_TE[fourcell_TE$type == "maternal", ]$TE )

wilcox.test(  fourcell_TE[fourcell_TE$type == "maternal", ]$TE, fourcell_TE[fourcell_TE$type == "paternal", ]$TE )
```

```{r echo=FALSE,warning=FALSE}

set.seed(3)

eightcell_TE = 
  mymain(
       TE_df            = TE_df_clr, 
       rna_averages     = rnaseq_eightcell_averages_clr, 
       gene_lengths     = region_lengths, 
       picker_name      = "eightcell_TE_clr"  , 
       paternal_genes   = multiple_snp_supported_eightcell_paternal_genes, 
       maternal_genes   = multiple_snp_supported_eightcell_maternal_genes, 
       background_genes = eightcell_background_genes_bootstrap)

eightcell_TE



sample_eightcell = matchit(imprinted~rna + cds_length, data = eightcell_TE )

sample_eightcell


matched_eightcell = match.data(sample_eightcell)

matched_eightcell

boxplot( log2(eightcell_TE$rna) ~ eightcell_TE$imprinted )

boxplot( log2(matched_eightcell$rna) ~ matched_eightcell$imprinted )

boxplot( log2(matched_eightcell$rna) ~ matched_eightcell$type )

boxplot( eightcell_TE$TE ~ eightcell_TE$type )

boxplot( matched_eightcell$TE ~ matched_eightcell$type )


plot(density( log2( matched_eightcell[matched_eightcell$type == "maternal",]$cds_length  )   ), ylim = c(0, 0.8) )

lines(density( log2( matched_eightcell[matched_eightcell$type == "paternal",]$cds_length  )   ), col = "blue" )

lines(density( log2( matched_eightcell[matched_eightcell$type == "background",]$cds_length  )   ), col = "red" )
```

```{r echo=FALSE,warning=FALSE}
wilcox.test(  eightcell_TE[eightcell_TE$imprinted == 0, ]$TE, eightcell_TE[eightcell_TE$imprinted == 1, ]$TE )

wilcox.test(  matched_eightcell[matched_eightcell$imprinted == 0, ]$TE, matched_eightcell[matched_eightcell$imprinted == 1, ]$TE )
```


## Comparison with GSE80810

```{r echo=FALSE,warning=FALSE}
GSE80810_allelic_ratios = read.csv(GSE80810_allelic_ratio_file)

a = colnames(GSE80810_allelic_ratios)
GSE80810_allelic_ratios = GSE80810_allelic_ratios[, c(1,16:20,26:44)]

GSE80810_8C_paternal_intersection = GSE80810_allelic_ratios[GSE80810_allelic_ratios$Genes %in% eightcell_paternal_genes_bootstrap,-c(1,2,3,4,5,6)]
GSE80810_8C_paternal_averages = apply(GSE80810_8C_paternal_intersection , MARGIN = 1, FUN = mean, na.rm = 1) 
GSE80810_8C_paternal_agreements = sum(as.integer(GSE80810_8C_paternal_averages > 0.5 ) , na.rm = T)

print(c("8 cell paternal", GSE80810_8C_paternal_agreements, "agreements out of ", dim(GSE80810_8C_paternal_intersection)[1]))



GSE80810_8C_maternal_intersection = GSE80810_allelic_ratios[GSE80810_allelic_ratios$Genes %in% eightcell_maternal_genes_bootstrap,-c(1,2,3,4,5,6)]
GSE80810_8C_maternal_averages = apply(GSE80810_8C_maternal_intersection , MARGIN = 1, FUN = mean, na.rm = 1)
GSE80810_8C_maternal_agreements = sum(as.integer(GSE80810_8C_maternal_averages < 0.5 ) , na.rm = T)

print(c("8 cell maternal", GSE80810_8C_maternal_agreements, "agreements out of ", dim(GSE80810_8C_maternal_intersection)[1]))



GSE80810_4C_paternal_intersection = GSE80810_allelic_ratios[GSE80810_allelic_ratios$Genes %in% fourcell_paternal_genes_bootstrap,-c(1,2,3,4,5,6)]
GSE80810_4C_paternal_averages = apply(GSE80810_4C_paternal_intersection , MARGIN = 1, FUN = mean, na.rm = 1) 
GSE80810_4C_paternal_agreements = sum(as.integer(GSE80810_4C_paternal_averages > 0.5 ) , na.rm = T)

print(c("4 cell paternal", GSE80810_4C_paternal_agreements, "agreements out of ", dim(GSE80810_4C_paternal_intersection)[1]))



GSE80810_4C_maternal_intersection = GSE80810_allelic_ratios[GSE80810_allelic_ratios$Genes %in% fourcell_maternal_genes_bootstrap,-c(1,2,3,4,5,6)]
GSE80810_4C_maternal_averages = apply(GSE80810_4C_maternal_intersection , MARGIN = 1, FUN = mean, na.rm = 1) 
GSE80810_4C_maternal_agreements = sum(as.integer(GSE80810_4C_maternal_averages < 0.5 ) , na.rm = T)

print(c("4 cell maternal", GSE80810_4C_maternal_agreements, "agreements out of ", dim(GSE80810_4C_maternal_intersection)[1]))

```

```{r echo=FALSE,warning=FALSE}
```


```{r echo=FALSE,warning=FALSE}
```


```{r echo=FALSE,warning=FALSE}
```


```{r echo=FALSE,warning=FALSE}
```


```{r echo=FALSE,warning=FALSE}
```


```{r echo=FALSE,warning=FALSE}
```


```{r echo=FALSE,warning=FALSE}
```


```{r echo=FALSE,warning=FALSE}
```


```{r echo=FALSE,warning=FALSE}
```


```{r echo=FALSE,warning=FALSE}
```