---
title: "transite RBP analysis"
author: "Ian Hoskins"
date: "1/11/2022"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Transite RBP motif analysis by matrix TSMA

Transite also finds enriched RBP motifs in sequence sets. One benefit is that it does not involve subsampling of the control sets but uses the entire background of sequences for enrichment analysis. One drawback is it does not control for other features correlated with translational efficiency. Consider using the matrix SPMA method for gene rank-based differential enrichment analysis, or the k-merized versions of TSMA, SPMA to search for enriched k-mers in unbiased fashion.

Things to do: 
- Update background to include expressed genes
- Repeat analysis of heptamers?
```{r Transite matrix TSMA}

transite_res_dir<- "~/Desktop/transite/transite"
cache_dir<- paste(transite_res_dir, "transite_cache", sep="/")

library(data.table)
library(ggplot2)
library(ggrepel)
library(cowplot)
library(ggsci)

library(Biostrings)
library(transite)

# Use stage, TE direction specific thresholds?
#transite_fdr_thresh<- 0.001

# Get the sequences of foreground and background sets for relevant combinations where we observed
# hits
transite_TE_up_GV_vs_MII_UTR5_dir<- paste(transite_res_dir, "TE_up_GV_vs_MII", "UTR5", sep="/")
transite_TE_up_GV_vs_MII_UTR3_dir<- paste(transite_res_dir, "TE_up_GV_vs_MII", "UTR3", sep="/")
transite_TE_up_GV_vs_MII_CDS_dir<- paste(transite_res_dir, "TE_up_GV_vs_MII", "CDS", sep="/")

transite_TE_down_GV_vs_MII_UTR5_dir<- paste(transite_res_dir, "TE_down_GV_vs_MII", "UTR5", sep="/")
transite_TE_down_GV_vs_MII_UTR3_dir<- paste(transite_res_dir, "TE_down_GV_vs_MII", "UTR3", sep="/")
transite_TE_down_GV_vs_MII_CDS_dir<- paste(transite_res_dir, "TE_down_GV_vs_MII", "CDS", sep="/")

transite_TE_up_MII_vs_1_UTR5_dir<- paste(transite_res_dir, "TE_up_MII_vs_1", "UTR5", sep="/")
transite_TE_up_MII_vs_1_UTR3_dir<- paste(transite_res_dir, "TE_up_MII_vs_1", "UTR3", sep="/")
transite_TE_up_MII_vs_1_CDS_dir<- paste(transite_res_dir, "TE_up_MII_vs_1", "CDS", sep="/")

transite_TE_down_MII_vs_1_UTR5_dir<- paste(transite_res_dir, "TE_down_MII_vs_1", "UTR5", sep="/")
transite_TE_down_MII_vs_1_UTR3_dir<- paste(transite_res_dir, "TE_down_MII_vs_1", "UTR3", sep="/")
transite_TE_down_MII_vs_1_CDS_dir<- paste(transite_res_dir, "TE_down_MII_vs_1", "CDS", sep="/")


transite_TE_up_1_vs_2_UTR5_dir<- paste(transite_res_dir, "TE_up_1_vs_2", "UTR5", sep="/")
transite_TE_up_1_vs_2_UTR3_dir<- paste(transite_res_dir, "TE_up_1_vs_2", "UTR3", sep="/")
transite_TE_up_1_vs_2_CDS_dir<- paste(transite_res_dir, "TE_up_1_vs_2", "CDS", sep="/")

transite_TE_down_1_vs_2_UTR5_dir<- paste(transite_res_dir, "TE_down_1_vs_2", "UTR5", sep="/")
transite_TE_down_1_vs_2_UTR3_dir<- paste(transite_res_dir, "TE_down_1_vs_2", "UTR3", sep="/")
transite_TE_down_1_vs_2_CDS_dir<- paste(transite_res_dir, "TE_down_1_vs_2", "CDS", sep="/")




# Get backgrounds
transite_UTR5_back<- paste(transite_res_dir, "UTR5_background.fa", sep="/")
transite_CDS_back<- paste(transite_res_dir, "CDS_background.fa", sep="/")
transite_UTR3_back<- paste(transite_res_dir, "UTR3_background.fa", sep="/")

transite_UTR5_back_seqs<- readRNAStringSet(transite_UTR5_back, format="fasta", use.names=TRUE)
transite_CDS_back_seqs<- readRNAStringSet(transite_CDS_back, format="fasta", use.names=TRUE)
transite_UTR3_back_seqs<- readRNAStringSet(transite_UTR3_back, format="fasta", use.names=TRUE)

# Remove sequences that have Ns
transite_UTR5_back_seqs<- transite_UTR5_back_seqs[!grepl("N", as.character(transite_UTR5_back_seqs))]
transite_CDS_back_seqs<- transite_CDS_back_seqs[!grepl("N", as.character(transite_CDS_back_seqs))]
transite_UTR3_back_seqs<- transite_UTR3_back_seqs[!grepl("N", as.character(transite_UTR3_back_seqs))]

## Restrict to Expressed set

background_genes_GV_MII = read.table('~/Desktop/transite/gene_ids/TE_background_GV_vs_MII_genes.txt')
background_genes_MII_1 = read.table('~/Desktop/transite/gene_ids/TE_background_MII_vs_1_genes.txt')
background_genes_1_2 = read.table('~/Desktop/transite/gene_ids/TE_background_1_vs_2_genes.txt')


expr_GV_MII_CDS =  sapply (base::strsplit( as.character(transite_CDS_back_seqs@ranges@NAMES), split = "|", fixed = T) , "[[", 6) %in% background_genes_GV_MII$V1
expr_GV_MII_UTR3 =  sapply (base::strsplit( as.character(transite_UTR3_back_seqs@ranges@NAMES), split = "|", fixed = T) , "[[", 6) %in% background_genes_GV_MII$V1
expr_GV_MII_UTR5 =  sapply (base::strsplit( as.character(transite_UTR5_back_seqs@ranges@NAMES), split = "|", fixed = T) , "[[", 6) %in% background_genes_GV_MII$V1

expr_MII_1_CDS =  sapply (base::strsplit( as.character(transite_CDS_back_seqs@ranges@NAMES), split = "|", fixed = T) , "[[", 6) %in% background_genes_MII_1$V1
expr_MII_1_UTR3 =  sapply (base::strsplit( as.character(transite_UTR3_back_seqs@ranges@NAMES), split = "|", fixed = T) , "[[", 6) %in% background_genes_MII_1$V1
expr_MII_1_UTR5 =  sapply (base::strsplit( as.character(transite_UTR5_back_seqs@ranges@NAMES), split = "|", fixed = T) , "[[", 6) %in% background_genes_MII_1$V1

expr_1_2_CDS =  sapply (base::strsplit( as.character(transite_CDS_back_seqs@ranges@NAMES), split = "|", fixed = T) , "[[", 6) %in% background_genes_1_2$V1
expr_1_2_UTR3 =  sapply (base::strsplit( as.character(transite_UTR3_back_seqs@ranges@NAMES), split = "|", fixed = T) , "[[", 6) %in% background_genes_1_2$V1
expr_1_2_UTR5 =  sapply (base::strsplit( as.character(transite_UTR5_back_seqs@ranges@NAMES), split = "|", fixed = T) , "[[", 6) %in% background_genes_1_2$V1


### UTR5
transite_TE_up_GV_vs_MII_UTR5_fore<- paste(transite_TE_up_GV_vs_MII_UTR5_dir, "TE_up_GV_vs_MII_genes.foreground.fa", sep="/")

transite_TE_up_GV_vs_MII_UTR5_fore_seqs<- readRNAStringSet(
  transite_TE_up_GV_vs_MII_UTR5_fore, format="fasta", use.names=TRUE)

transite_TE_down_GV_vs_MII_UTR5_fore<- paste(transite_TE_down_GV_vs_MII_UTR5_dir, "TE_down_GV_vs_MII_genes.foreground.fa", sep="/")

transite_TE_down_GV_vs_MII_UTR5_fore_seqs<- readRNAStringSet(
  transite_TE_down_GV_vs_MII_UTR5_fore, format="fasta", use.names=TRUE)


transite_TE_up_MII_vs_1_UTR5_fore<- paste(transite_TE_up_MII_vs_1_UTR5_dir, "TE_up_MII_vs_1_genes.foreground.fa", sep="/")

transite_TE_up_MII_vs_1_UTR5_fore_seqs<- readRNAStringSet(
  transite_TE_up_MII_vs_1_UTR5_fore, format="fasta", use.names=TRUE)

transite_TE_down_MII_vs_1_UTR5_fore<- paste(transite_TE_down_MII_vs_1_UTR5_dir, "TE_down_MII_vs_1_genes.foreground.fa", sep="/")

transite_TE_down_MII_vs_1_UTR5_fore_seqs<- readRNAStringSet(
  transite_TE_down_MII_vs_1_UTR5_fore, format="fasta", use.names=TRUE)


transite_TE_up_1_vs_2_UTR5_fore<- paste(transite_TE_up_1_vs_2_UTR5_dir, "TE_up_1_vs_2_genes.foreground.fa", sep="/")

transite_TE_up_1_vs_2_UTR5_fore_seqs<- readRNAStringSet(
  transite_TE_up_1_vs_2_UTR5_fore, format="fasta", use.names=TRUE)

transite_TE_down_1_vs_2_UTR5_fore<- paste(transite_TE_down_1_vs_2_UTR5_dir, "TE_down_1_vs_2_genes.foreground.fa", sep="/")

transite_TE_down_1_vs_2_UTR5_fore_seqs<- readRNAStringSet(
  transite_TE_down_1_vs_2_UTR5_fore, format="fasta", use.names=TRUE)


suppressWarnings(transite_UTR5_res_GV_MII <- run_kmer_tsma(
  foreground_sets=list(as.character(transite_TE_up_GV_vs_MII_UTR5_fore_seqs),
                       as.character(transite_TE_down_GV_vs_MII_UTR5_fore_seqs)),
  background_set=as.character(transite_UTR5_back_seqs)[expr_GV_MII_UTR5], n_cores = 64))

suppressWarnings(transite_UTR5_res_MII_1 <- run_kmer_tsma(
  foreground_sets=list(as.character(transite_TE_up_MII_vs_1_UTR5_fore_seqs),
                       as.character(transite_TE_down_MII_vs_1_UTR5_fore_seqs)),
  background_set=as.character(transite_UTR5_back_seqs)[expr_MII_1_UTR5], n_cores = 64))

suppressWarnings(transite_UTR5_res_1_2 <- run_kmer_tsma(
  foreground_sets=list(as.character(transite_TE_up_1_vs_2_UTR5_fore_seqs),
                       as.character(transite_TE_down_1_vs_2_UTR5_fore_seqs)),
  background_set=as.character(transite_UTR5_back_seqs)[expr_1_2_UTR5], n_cores = 64))




### UTR3
transite_TE_up_GV_vs_MII_UTR3_fore<- paste(transite_TE_up_GV_vs_MII_UTR3_dir, "TE_up_GV_vs_MII_genes.foreground.fa", sep="/")

transite_TE_up_GV_vs_MII_UTR3_fore_seqs<- readRNAStringSet(
  transite_TE_up_GV_vs_MII_UTR3_fore, format="fasta", use.names=TRUE)

transite_TE_down_GV_vs_MII_UTR3_fore<- paste(transite_TE_down_GV_vs_MII_UTR3_dir, "TE_down_GV_vs_MII_genes.foreground.fa", sep="/")

transite_TE_down_GV_vs_MII_UTR3_fore_seqs<- readRNAStringSet(
  transite_TE_down_GV_vs_MII_UTR3_fore, format="fasta", use.names=TRUE)


transite_TE_up_MII_vs_1_UTR3_fore<- paste(transite_TE_up_MII_vs_1_UTR3_dir, "TE_up_MII_vs_1_genes.foreground.fa", sep="/")

transite_TE_up_MII_vs_1_UTR3_fore_seqs<- readRNAStringSet(
  transite_TE_up_MII_vs_1_UTR3_fore, format="fasta", use.names=TRUE)

transite_TE_down_MII_vs_1_UTR3_fore<- paste(transite_TE_down_MII_vs_1_UTR3_dir, "TE_down_MII_vs_1_genes.foreground.fa", sep="/")

transite_TE_down_MII_vs_1_UTR3_fore_seqs<- readRNAStringSet(
  transite_TE_down_MII_vs_1_UTR3_fore, format="fasta", use.names=TRUE)


transite_TE_up_1_vs_2_UTR3_fore<- paste(transite_TE_up_1_vs_2_UTR3_dir, "TE_up_1_vs_2_genes.foreground.fa", sep="/")

transite_TE_up_1_vs_2_UTR3_fore_seqs<- readRNAStringSet(
  transite_TE_up_1_vs_2_UTR3_fore, format="fasta", use.names=TRUE)

transite_TE_down_1_vs_2_UTR3_fore<- paste(transite_TE_down_1_vs_2_UTR3_dir, "TE_down_1_vs_2_genes.foreground.fa", sep="/")

transite_TE_down_1_vs_2_UTR3_fore_seqs<- readRNAStringSet(
  transite_TE_down_1_vs_2_UTR3_fore, format="fasta", use.names=TRUE)

suppressWarnings(transite_UTR3_res_GV_MII <- run_kmer_tsma(
  foreground_sets=list(as.character(transite_TE_up_GV_vs_MII_UTR3_fore_seqs),
                       as.character(transite_TE_down_GV_vs_MII_UTR3_fore_seqs)),
  background_set=as.character(transite_UTR3_back_seqs)[expr_GV_MII_UTR3], n_cores = 64))

suppressWarnings(transite_UTR3_res_MII_1 <- run_kmer_tsma(
  foreground_sets=list(as.character(transite_TE_up_MII_vs_1_UTR3_fore_seqs),
                       as.character(transite_TE_down_MII_vs_1_UTR3_fore_seqs)),
  background_set=as.character(transite_UTR3_back_seqs)[expr_MII_1_UTR3], n_cores = 64))

suppressWarnings(transite_UTR3_res_1_2 <- run_kmer_tsma(
  foreground_sets=list(as.character(transite_TE_up_1_vs_2_UTR3_fore_seqs),
                       as.character(transite_TE_down_1_vs_2_UTR3_fore_seqs)),
  background_set=as.character(transite_UTR3_back_seqs)[expr_1_2_UTR3], n_cores = 64))



### CDS
transite_TE_up_GV_vs_MII_CDS_fore<- paste(transite_TE_up_GV_vs_MII_CDS_dir, "TE_up_GV_vs_MII_genes.foreground.fa", sep="/")

transite_TE_up_GV_vs_MII_CDS_fore_seqs<- readRNAStringSet(
  transite_TE_up_GV_vs_MII_CDS_fore, format="fasta", use.names=TRUE)

transite_TE_down_GV_vs_MII_CDS_fore<- paste(transite_TE_down_GV_vs_MII_CDS_dir, "TE_down_GV_vs_MII_genes.foreground.fa", sep="/")

transite_TE_down_GV_vs_MII_CDS_fore_seqs<- readRNAStringSet(
  transite_TE_down_GV_vs_MII_CDS_fore, format="fasta", use.names=TRUE)


transite_TE_up_MII_vs_1_CDS_fore<- paste(transite_TE_up_MII_vs_1_CDS_dir, "TE_up_MII_vs_1_genes.foreground.fa", sep="/")

transite_TE_up_MII_vs_1_CDS_fore_seqs<- readRNAStringSet(
  transite_TE_up_MII_vs_1_CDS_fore, format="fasta", use.names=TRUE)

transite_TE_down_MII_vs_1_CDS_fore<- paste(transite_TE_down_MII_vs_1_CDS_dir, "TE_down_MII_vs_1_genes.foreground.fa", sep="/")

transite_TE_down_MII_vs_1_CDS_fore_seqs<- readRNAStringSet(
  transite_TE_down_MII_vs_1_CDS_fore, format="fasta", use.names=TRUE)


transite_TE_up_1_vs_2_CDS_fore<- paste(transite_TE_up_1_vs_2_CDS_dir, "TE_up_1_vs_2_genes.foreground.fa", sep="/")

transite_TE_up_1_vs_2_CDS_fore_seqs<- readRNAStringSet(
  transite_TE_up_1_vs_2_CDS_fore, format="fasta", use.names=TRUE)

transite_TE_down_1_vs_2_CDS_fore<- paste(transite_TE_down_1_vs_2_CDS_dir, "TE_down_1_vs_2_genes.foreground.fa", sep="/")

transite_TE_down_1_vs_2_CDS_fore_seqs<- readRNAStringSet(
  transite_TE_down_1_vs_2_CDS_fore, format="fasta", use.names=TRUE)

suppressWarnings(transite_CDS_res_GV_MII <- run_kmer_tsma(
  foreground_sets=list(as.character(transite_TE_up_GV_vs_MII_CDS_fore_seqs),
                       as.character(transite_TE_down_GV_vs_MII_CDS_fore_seqs)),
  background_set=as.character(transite_CDS_back_seqs)[expr_GV_MII_CDS], n_cores = 64))

suppressWarnings(transite_CDS_res_MII_1 <- run_kmer_tsma(
  foreground_sets=list(as.character(transite_TE_up_MII_vs_1_CDS_fore_seqs),
                       as.character(transite_TE_down_MII_vs_1_CDS_fore_seqs)),
  background_set=as.character(transite_CDS_back_seqs)[expr_MII_1_CDS], n_cores = 64))

suppressWarnings(transite_CDS_res_1_2 <- run_kmer_tsma(
  foreground_sets=list(as.character(transite_TE_up_1_vs_2_CDS_fore_seqs),
                       as.character(transite_TE_down_1_vs_2_CDS_fore_seqs)),
  background_set=as.character(transite_CDS_back_seqs)[expr_1_2_CDS], n_cores = 64))



## Save hexamer results
# save(transite_CDS_res, transite_UTR5_res, transite_UTR3_res,  file = "~/Desktop/transite/CDS_results_hexamer.RData")

## We might want to also run heptamer version to compare the two. 

######################################
## Analysis of results
######################################

fdr_threshold = 0.01

## Each result is a list of 2 lists. Below is an example for one of these
transite_UTR5_res_GV_MII_dt<- as.data.table(transite_UTR5_res_GV_MII[[2]]$enrichment_df)
transite_UTR5_res_GV_MII_dt[,Sig:=FALSE]
transite_UTR5_res_GV_MII_dt[adj_p_value<fdr_threshold, Sig:=TRUE]

transite_UTR5_res_GV_MII_gg <- ggplot(
  data=transite_UTR5_res_GV_MII_dt, 
  aes(x=enrichment, y=-1*log10(adj_p_value), col=Sig))

transite_UTR5_res_GV_MII_gg +
  geom_vline(xintercept=c(1) , col="grey", linetype=2)+
  geom_point()+
  scale_color_simpsons()+
  scale_x_continuous(breaks=seq(0,2.5,0.5))+
  labs(x="RBP motif enrichment", 
       y="- log10 FDR", col="FDR < 0.01")+
  theme_cowplot()

which(transite_UTR5_res_GV_MII[[2]]$motif_df$adj_p_value_estimate < 0.01)
transite_UTR5_res_GV_MII[[1]]$volcano_plots[[16]]


write.table(as.data.frame(transite_UTR5_res_GV_MII[[1]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR5_res_GV_MII_TE_up.csv' , row.names = F )
write.table(as.data.frame(transite_UTR5_res_GV_MII[[2]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR5_res_GV_MII_TE_down.csv' , row.names = F )
write.table(as.data.frame(transite_UTR5_res_MII_1[[1]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR5_res_MII_1_TE_up.csv' , row.names = F )
write.table(as.data.frame(transite_UTR5_res_MII_1[[2]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR5_res_MII_1_TE_down.csv' , row.names = F )
write.table(as.data.frame(transite_UTR5_res_1_2[[1]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR5_res_1_2_TE_up.csv' , row.names = F )
write.table(as.data.frame(transite_UTR5_res_1_2[[2]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR5_res_1_2_TE_down.csv' , row.names = F )


write.table(as.data.frame(transite_UTR3_res_GV_MII[[1]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR3_res_GV_MII_TE_up.csv' , row.names = F )
write.table(as.data.frame(transite_UTR3_res_GV_MII[[2]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR3_res_GV_MII_TE_down.csv' , row.names = F )
write.table(as.data.frame(transite_UTR3_res_MII_1[[1]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR3_res_MII_1_TE_up.csv' , row.names = F )
write.table(as.data.frame(transite_UTR3_res_MII_1[[2]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR3_res_MII_1_TE_down.csv' , row.names = F )
write.table(as.data.frame(transite_UTR3_res_1_2[[1]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR3_res_1_2_TE_up.csv' , row.names = F )
write.table(as.data.frame(transite_UTR3_res_1_2[[2]]$enrichment_df), 
            file = '~/Desktop/transite/transite_UTR3_res_1_2_TE_down.csv' , row.names = F )


write.table(as.data.frame(transite_CDS_res_GV_MII[[1]]$enrichment_df), 
            file = '~/Desktop/transite/transite_CDS_res_GV_MII_TE_up.csv' , row.names = F )
write.table(as.data.frame(transite_CDS_res_GV_MII[[2]]$enrichment_df), 
            file = '~/Desktop/transite/transite_CDS_res_GV_MII_TE_down.csv' , row.names = F )
write.table(as.data.frame(transite_CDS_res_MII_1[[1]]$enrichment_df), 
            file = '~/Desktop/transite/transite_CDS_res_MII_1_TE_up.csv' , row.names = F )
write.table(as.data.frame(transite_CDS_res_MII_1[[2]]$enrichment_df), 
            file = '~/Desktop/transite/transite_CDS_res_MII_1_TE_down.csv' , row.names = F )
write.table(as.data.frame(transite_CDS_res_1_2[[1]]$enrichment_df), 
            file = '~/Desktop/transite/transite_CDS_res_1_2_TE_up.csv' , row.names = F )
write.table(as.data.frame(transite_CDS_res_1_2[[2]]$enrichment_df), 
            file = '~/Desktop/transite/transite_CDS_res_1_2_TE_down.csv' , row.names = F )

write.table(as.data.frame(transite_UTR5_res_GV_MII[[1]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR5_res_GV_MII_TE_up_motif.csv' , row.names = F )
write.table(as.data.frame(transite_UTR5_res_GV_MII[[2]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR5_res_GV_MII_TE_down_motif.csv' , row.names = F )
write.table(as.data.frame(transite_UTR5_res_MII_1[[1]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR5_res_MII_1_TE_up_motif.csv' , row.names = F )
write.table(as.data.frame(transite_UTR5_res_MII_1[[2]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR5_res_MII_1_TE_down_motif.csv' , row.names = F )
write.table(as.data.frame(transite_UTR5_res_1_2[[1]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR5_res_1_2_TE_up_motif.csv' , row.names = F )
write.table(as.data.frame(transite_UTR5_res_1_2[[2]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR5_res_1_2_TE_down_motif.csv' , row.names = F )


write.table(as.data.frame(transite_UTR3_res_GV_MII[[1]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR3_res_GV_MII_TE_up_motif.csv' , row.names = F )
write.table(as.data.frame(transite_UTR3_res_GV_MII[[2]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR3_res_GV_MII_TE_down_motif.csv' , row.names = F )
write.table(as.data.frame(transite_UTR3_res_MII_1[[1]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR3_res_MII_1_TE_up_motif.csv' , row.names = F )
write.table(as.data.frame(transite_UTR3_res_MII_1[[2]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR3_res_MII_1_TE_down_motif.csv' , row.names = F )
write.table(as.data.frame(transite_UTR3_res_1_2[[1]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR3_res_1_2_TE_up_motif.csv' , row.names = F )
write.table(as.data.frame(transite_UTR3_res_1_2[[2]]$motif_df), 
            file = '~/Desktop/transite/transite_UTR3_res_1_2_TE_down_motif.csv' , row.names = F )


write.table(as.data.frame(transite_CDS_res_GV_MII[[1]]$motif_df), 
            file = '~/Desktop/transite/transite_CDS_res_GV_MII_TE_up_motif.csv' , row.names = F )
write.table(as.data.frame(transite_CDS_res_GV_MII[[2]]$motif_df), 
            file = '~/Desktop/transite/transite_CDS_res_GV_MII_TE_down_motif.csv' , row.names = F )
write.table(as.data.frame(transite_CDS_res_MII_1[[1]]$motif_df), 
            file = '~/Desktop/transite/transite_CDS_res_MII_1_TE_up_motif.csv' , row.names = F )
write.table(as.data.frame(transite_CDS_res_MII_1[[2]]$motif_df), 
            file = '~/Desktop/transite/transite_CDS_res_MII_1_TE_down_motif.csv' , row.names = F )
write.table(as.data.frame(transite_CDS_res_1_2[[1]]$motif_df), 
            file = '~/Desktop/transite/transite_CDS_res_1_2_TE_up_motif.csv' , row.names = F )
write.table(as.data.frame(transite_CDS_res_1_2[[2]]$motif_df), 
            file = '~/Desktop/transite/transite_CDS_res_1_2_TE_down_motif.csv' , row.names = F )


```


```
